<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>High dimensional statistics with R: Introduction to high-dimensional data</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/01-introduction-to-high-dimensional-data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 0%" class="percentage">
    0%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 0%" aria-valuenow="0" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/01-introduction-to-high-dimensional-data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        1. Introduction to high-dimensional data
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#what-are-high-dimensional-data">What are high-dimensional data?</a></li>
<li><a href="#why-is-dealing-with-high-dimensional-data-challenging">Why is dealing with high-dimensional data challenging?</a></li>
<li><a href="#what-statistical-methods-are-used-to-analyse-high-dimensional-data">What statistical methods are used to analyse high-dimensional
data?</a></li>
<li><a href="#further-reading-and-resources">Further reading and resources</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-high-dimensional-regression.html">2. Regression with many outcomes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-principal-component-analysis.html">4. Principal component analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-k-means.html">6. K-means</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-hierarchical.html">7. Hierarchical clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>

                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-high-dimensional-regression.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="02-high-dimensional-regression.html" rel="next">
          Next: Regression with many...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Introduction to high-dimensional data</h1>
        <p>Last updated on 2025-09-02 |

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/01-introduction-to-high-dimensional-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What are high-dimensional data and what do these data look like in
the biosciences?</li>
<li>What are the challenges when analysing high-dimensional data?</li>
<li>What statistical methods are suitable for analysing these data?</li>
<li>How can Bioconductor be used to access high-dimensional data in the
biosciences?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explore examples of high-dimensional data in the biosciences.</li>
<li>Appreciate challenges involved in analysing high-dimensional
data.</li>
<li>Explore different statistical methods used for analysing
high-dimensional data.</li>
<li>Work with example data created from biological studies.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-are-high-dimensional-data">What are high-dimensional data?<a class="anchor" aria-label="anchor" href="#what-are-high-dimensional-data"></a></h2>
<hr class="half-width"><p><em>High-dimensional data</em> are defined as data with many features
(variables observed). In recent years, advances in information
technology have allowed large amounts of data to be collected and stored
with relative ease. As such, high-dimensional data have become more
common in many scientific fields, including the biological sciences,
where datasets in subjects like genomics and medical sciences often have
a large numbers of features. For example, hospital data may record many
variables, including symptoms, blood test results, behaviours, and
general health. An example of what high-dimensional data might look like
in a biomedical study is shown in the figure below.</p>
<figure style="text-align: center"><img src="fig/intro-table.png" alt="Table displaying a high-dimensional data set with many columns representing features related to health, such as blood pressure, heart rate and respiratory rate. Each row contains the data for an individual patient. This type of high-dimensional data could contain hundreds or thousands of columns (features/variables) and thousands or even millions of rows (observations/samples/patients)." width="5572" class="figure mx-auto d-block"><figcaption>
Example of a high-dimensional data table with features in the columns
and individual observations (patients) in rows.
</figcaption></figure><p>Researchers often want to relate such features to specific patient
outcomes (e.g. survival, length of time spent in hospital). However,
analysing high-dimensional data can be extremely challenging since
standard methods of analysis, such as individual plots of features and
linear regression, are no longer appropriate when we have many features.
In this lesson, we will learn alternative methods for dealing with
high-dimensional data and discover how these can be applied for
practical high-dimensional data analysis in the biological sciences.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1</h3>
<div class="callout-content">
<p>Descriptions of four research questions and their datasets are given
below. Which of these scenarios use high-dimensional data?</p>
<ol style="list-style-type: decimal"><li>Predicting patient blood pressure using: cholesterol level in blood,
age, and BMI measurements, collected from 100 patients.</li>
<li>Predicting patient blood pressure using: cholesterol level in blood,
age, and BMI, as well as information on 200,000 single nucleotide
polymorphisms from 100 patients.</li>
<li>Predicting the length of time patients spend in hospital with
pneumonia infection using: measurements on age, BMI, length of time with
symptoms, number of symptoms, and percentage of neutrophils in blood,
using data from 200 patients.</li>
<li>Predicting probability of a patient’s cancer progressing using gene
expression data from 20,000 genes, as well as data associated with
general patient health (age, weight, BMI, blood pressure) and cancer
growth (tumour size, localised spread, blood test results).</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>No. The number of features is relatively small (4 including the
response variable since this is an observed variable).</li>
<li>Yes, this is an example of high-dimensional data. There are 200,004
features.</li>
<li>No. The number of features is relatively small (6).</li>
<li>Yes. There are 20,008 features.</li>
</ol></div>
</div>
</div>
</div>
<p>Now that we have an idea of what high-dimensional data look like we
can think about the challenges we face in analysing them.</p>
</section><section><h2 class="section-heading" id="why-is-dealing-with-high-dimensional-data-challenging">Why is dealing with high-dimensional data challenging?<a class="anchor" aria-label="anchor" href="#why-is-dealing-with-high-dimensional-data-challenging"></a></h2>
<hr class="half-width"><p>Most classical statistical methods are set up for use on
low-dimensional data (i.e. with a small number of features, <span class="math inline">\(p\)</span>). This is because low-dimensional data
were much more common in the past when data collection was more
difficult and time consuming.</p>
<p>One challenge when analysing high-dimensional data is visualising the
many variables. When exploring low-dimensional datasets, it is possible
to plot the response variable against each of features to get an idea
which of these are important predictors of the response. With
high-dimensional data, the large number of features makes doing this
difficult. In addition, in some high-dimensional datasets it can also be
difficult to identify a single response variable, making standard data
exploration and analysis techniques less useful.</p>
<p>Let’s have a look at a simple dataset with lots of features to
understand some of the challenges we are facing when working with
high-dimensional data. For reference, all data used throughout the
lesson are described in the <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link">data
page</a>.</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2</h3>
<div class="callout-content">
<p>For illustrative purposes, we start with a simple dataset that is not
technically high-dimensional but contains many features. This will
illustrate the general problems encountered when working with many
features in a high-dimensional data set.</p>
<p>First, make sure you have completed <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/setup.html" class="external-link">the
setup instructions</a>. Next, let’s load the <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link"><code>prostate</code></a>
dataset as follows:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prostate</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/prostate.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Examine the dataset (in which each row represents a single patient)
to:</p>
<ol style="list-style-type: decimal"><li>Determine how many observations (<span class="math inline">\(n\)</span>) and features (<span class="math inline">\(p\)</span>) are available (hint: see the
<code>dim()</code> function).</li>
<li>Examine what variables were measured (hint: see the
<code>names()</code> and <code>head()</code> functions).</li>
<li>Plot the relationship between the variables (hint: see the
<code>pairs()</code> function). What problem(s) with high-dimensional
data analysis does this illustrate?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>    <span class="co"># print the number of rows and columns</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>  <span class="co"># examine the variable names</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>   <span class="co"># print the first 6 rows</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>  <span class="co"># examine column names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "lcavol"  "lweight" "age"     "lbph"    "svi"     "lcp"     "gleason"
[8] "pgg45"   "lpsa"   </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pairs</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>  <span class="co"># plot each pair of variables against each other</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/01-introduction-to-high-dimensional-data-rendered-pairs-prostate-1.png" alt="A set of pairwise scatterplots of variables in the 'prostate' dataset, namely lcavol, lweight, age, lbph, svi, lcp, gleason, pgg45, lpsa. The plots are shown in a grid." class="figure mx-auto d-block"><figcaption>
Pairwise plots of the ‘prostate’ dataset.
</figcaption></figure><p>The <code>pairs()</code> function plots relationships between each of
the variables in the <code>prostate</code> dataset. This is possible for
datasets with smaller numbers of variables, but for datasets in which
<span class="math inline">\(p\)</span> is larger it becomes difficult
(and time consuming) to visualise relationships between all variables in
the dataset. Even where visualisation is possible, fitting models to
datasets with many variables is difficult due to the potential for
overfitting and difficulties in identifying a response variable.</p>
</div>
</div>
</div>
</div>
<p>Note that function documentation and information on function
arguments will be useful throughout this lesson. We can access these
easily in R by running <code>?</code> followed by the package name. For
example, the documentation for the <code>dim</code> function can be
accessed by running <code>?dim</code>.</p>
<div id="locating-data-with-r---the-here-package" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="locating-data-with-r---the-here-package" class="callout-inner">
<h3 class="callout-title">Locating data with R - the
<strong><code>here</code></strong> package</h3>
<div class="callout-content">
<p>It is often desirable to access external datasets from inside R and
to write code that does this reliably on different computers. While R
has an inbulit function <code>setwd()</code> that can be used to denote
where external datasets are stored, this usually requires the user to
adjust the code to their specific system and folder structure. The
<strong><code>here</code></strong> package is meant to be used in R
projects. It allows users to specify the data location relative to the R
project directory. This makes R code more portable and can contribute to
improve the reproducibility of an analysis.</p>
</div>
</div>
</div>
<p>As well as many variables causing problems when working with
high-dimensional data, having relatively few observations (<span class="math inline">\(n\)</span>) compared to the number of features
(<span class="math inline">\(p\)</span>) causes additional challenges.
To illustrate these challenges, imagine we are carrying out least
squares regression on a dataset with 25 observations. Fitting a best fit
line through these data produces a plot shown in the left-hand panel of
the figure below.</p>
<p>However, imagine a situation in which the number of observations and
features in a dataset are almost equal. In that situation the effective
number of observations per feature is low. The result of fitting a best
fit line through few observations can be seen in the right-hand panel
below.</p>
<figure style="text-align: center"><img src="fig/intro-scatterplot.png" alt="Two scatter plots side-by-side, each plotting the relationship between two variables. The scatter plot on the left hand side shows 25 observations and a regression line with the points evenly scattered around. The scatter plot on the right hand side shows 2 observations and a regression line that goes through both points." width="5307" class="figure mx-auto d-block"><figcaption>
Scatter plot of two variables (x and y) from a data set with 25
observations (left) and 2 observations (right) with a fitted regression
line (red).
</figcaption></figure><p>In the first situation, the least squares regression line does not
fit the data perfectly and there is some error around the regression
line. But, when there are only two observations the regression line will
fit through the points exactly, resulting in overfitting of the data.
This suggests that carrying out least squares regression on a dataset
with few data points per feature would result in difficulties in
applying the resulting model to further datsets. This is a common
problem when using regression on high-dimensional datasets.</p>
<p>Another problem in carrying out regression on high-dimensional data
is dealing with correlations between explanatory variables. The large
numbers of features in these datasets makes high correlations between
variables more likely. Let’s explore why high correlations might be an
issue in a Challenge.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3</h3>
<div class="callout-content">
<p>Use the <code>cor()</code> function to examine correlations between
all variables in the <code>prostate</code> dataset. Are some pairs of
variables highly correlated using a threshold of 0.75 for the
correlation coefficients?</p>
<p>Use the <code>lm()</code> function to fit univariate regression
models to predict patient age using two variables that are highly
correlated as predictors. Which of these variables are statistically
significant predictors of age? Hint: the <code>summary()</code> function
can help here.</p>
<p>Fit a multiple linear regression model predicting patient age using
both variables. What happened?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Create a correlation matrix of all variables in the
<code>prostate</code> dataset</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cor</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           lcavol      lweight       age         lbph         svi          lcp
lcavol  1.0000000  0.194128286 0.2249999  0.027349703  0.53884500  0.675310484
lweight 0.1941283  1.000000000 0.3075286  0.434934636  0.10877851  0.100237795
age     0.2249999  0.307528614 1.0000000  0.350185896  0.11765804  0.127667752
lbph    0.0273497  0.434934636 0.3501859  1.000000000 -0.08584324 -0.006999431
svi     0.5388450  0.108778505 0.1176580 -0.085843238  1.00000000  0.673111185
lcp     0.6753105  0.100237795 0.1276678 -0.006999431  0.67311118  1.000000000
gleason 0.4324171 -0.001275658 0.2688916  0.077820447  0.32041222  0.514830063
pgg45   0.4336522  0.050846821 0.2761124  0.078460018  0.45764762  0.631528245
lpsa    0.7344603  0.354120390 0.1695928  0.179809410  0.56621822  0.548813169
             gleason      pgg45      lpsa
lcavol   0.432417056 0.43365225 0.7344603
lweight -0.001275658 0.05084682 0.3541204
age      0.268891599 0.27611245 0.1695928
lbph     0.077820447 0.07846002 0.1798094
svi      0.320412221 0.45764762 0.5662182
lcp      0.514830063 0.63152825 0.5488132
gleason  1.000000000 0.75190451 0.3689868
pgg45    0.751904512 1.00000000 0.4223159
lpsa     0.368986803 0.42231586 1.0000000</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span> <span class="co"># rounding helps to visualise the correlations</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        lcavol lweight  age  lbph   svi   lcp gleason pgg45 lpsa
lcavol    1.00    0.19 0.22  0.03  0.54  0.68    0.43  0.43 0.73
lweight   0.19    1.00 0.31  0.43  0.11  0.10    0.00  0.05 0.35
age       0.22    0.31 1.00  0.35  0.12  0.13    0.27  0.28 0.17
lbph      0.03    0.43 0.35  1.00 -0.09 -0.01    0.08  0.08 0.18
svi       0.54    0.11 0.12 -0.09  1.00  0.67    0.32  0.46 0.57
lcp       0.68    0.10 0.13 -0.01  0.67  1.00    0.51  0.63 0.55
gleason   0.43    0.00 0.27  0.08  0.32  0.51    1.00  0.75 0.37
pgg45     0.43    0.05 0.28  0.08  0.46  0.63    0.75  1.00 0.42
lpsa      0.73    0.35 0.17  0.18  0.57  0.55    0.37  0.42 1.00</code></pre>
</div>
<p>As seen above, some variables are highly correlated. In particular,
the correlation between <code>gleason</code> and <code>pgg45</code> is
equal to 0.75.</p>
<p>Fitting univariate regression models to predict age using gleason and
pgg45 as predictors.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_gleason</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">age</span> <span class="op">~</span> <span class="va">gleason</span>, data <span class="op">=</span> <span class="va">prostate</span><span class="op">)</span></span>
<span><span class="va">model_pgg45</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">age</span> <span class="op">~</span> <span class="va">pgg45</span>, data <span class="op">=</span> <span class="va">prostate</span><span class="op">)</span></span></code></pre>
</div>
<p>Check which covariates have a significant efffect</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">model_gleason</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = age ~ gleason, data = prostate)

Residuals:
    Min      1Q  Median      3Q     Max
-20.780  -3.552   1.448   4.220  13.448

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)   45.146      6.918   6.525 3.29e-09 ***
gleason        2.772      1.019   2.721  0.00774 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 7.209 on 95 degrees of freedom
Multiple R-squared:  0.0723,	Adjusted R-squared:  0.06254
F-statistic: 7.404 on 1 and 95 DF,  p-value: 0.007741</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">model_pgg45</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = age ~ pgg45, data = prostate)

Residuals:
     Min       1Q   Median       3Q      Max
-21.0889  -3.4533   0.9111   4.4534  15.1822

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept) 62.08890    0.96758   64.17  &lt; 2e-16 ***
pgg45        0.07289    0.02603    2.80  0.00619 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 7.193 on 95 degrees of freedom
Multiple R-squared:  0.07624,	Adjusted R-squared:  0.06651
F-statistic:  7.84 on 1 and 95 DF,  p-value: 0.006189</code></pre>
</div>
<p>Based on these results we conclude that both <code>gleason</code> and
<code>pgg45</code> have a statistically significant univariate effect
(also referred to as a marginal effect) as predictors of age (5%
significance level).</p>
<p>Fitting a multivariate regression model using both both
<code>gleason</code> and <code>pgg45</code> as predictors</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_multivar</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">age</span> <span class="op">~</span> <span class="va">gleason</span> <span class="op">+</span> <span class="va">pgg45</span>, data <span class="op">=</span> <span class="va">prostate</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">model_multivar</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = age ~ gleason + pgg45, data = prostate)

Residuals:
    Min      1Q  Median      3Q     Max
-20.927  -3.677   1.323   4.323  14.420

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept) 52.95548    9.74316   5.435  4.3e-07 ***
gleason      1.45363    1.54299   0.942    0.349
pgg45        0.04490    0.03951   1.137    0.259
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 7.198 on 94 degrees of freedom
Multiple R-squared:  0.08488,	Adjusted R-squared:  0.06541
F-statistic: 4.359 on 2 and 94 DF,  p-value: 0.01547</code></pre>
</div>
<p>Although <code>gleason</code> and <code>pgg45</code> have
statistically significant univariate effects, this is no longer the case
when both variables are simultaneously included as covariates in a
multivariate regression model.</p>
</div>
</div>
</div>
</div>
<p>Including highly correlated variables such as <code>gleason</code>
and <code>pgg45</code> simultaneously the same regression model can lead
to problems in fitting a regression model and interpreting its output.
Although each variable appears to be associated with the response
individually, the model cannot distinguish the contribution of each
variable to the model. This can also increase the risk of over-fitting
since the model may fit redundant variables to noise rather than true
relationships.</p>
<p>To allow the information from variables to be included in the same
model despite high levels of correlation, we can use dimensionality
reduction methods to collapse multiple variables into a single new
variable (we will explore this dataset further in the dimensionality
reduction lesson). We can also use modifications to linear regression
like regularisation, which we will discuss in the lesson on
high-dimensional regression.</p>
</section><section><h2 class="section-heading" id="what-statistical-methods-are-used-to-analyse-high-dimensional-data">What statistical methods are used to analyse high-dimensional
data?<a class="anchor" aria-label="anchor" href="#what-statistical-methods-are-used-to-analyse-high-dimensional-data"></a></h2>
<hr class="half-width"><p>We have discussed so far that high-dimensional data analysis can be
challenging since variables are difficult to visualise, leading to
challenges identifying relationships between variables and suitable
response variables; we may have relatively few observations compared to
features, leading to over-fitting; and features may be highly
correlated, leading to challenges interpreting models. We therefore
require alternative approaches to examine whether, for example, groups
of observations show similar characteristics and whether these groups
may relate to other features in the data (e.g. phenotype in genetics
data).</p>
<p>In this course, we will cover four methods that help in dealing with
high-dimensional data: (1) regression with numerous outcome variables,
(2) regularised regression, (3) dimensionality reduction, and (4)
clustering. Here are some examples of when each of these approaches may
be used:</p>
<ol style="list-style-type: decimal"><li><p>Regression with numerous outcomes refers to situations in which
there are many variables of a similar kind (expression values for many
genes, methylation levels for many sites in the genome) and when one is
interested in assessing whether these variables are associated with a
specific covariate of interest, such as experimental condition or age.
In this case, multiple univariate regression models (one per each
outcome, using the covariate of interest as predictor) could be fitted
independently. In the context of high-dimensional molecular data, a
typical example are <em>differential gene expression</em> analyses. We
will explore this type of analysis in the <em>Regression with many
outcomes</em> episode.</p></li>
<li><p>Regularisation (also known as <em>regularised regression</em> or
<em>penalised regression</em>) is typically used to fit regression
models when there is a single outcome variable or interest but the
number of potential predictors is large, e.g. there are more predictors
than observations. Regularisation can help to prevent overfitting and
may be used to identify a small subset of predictors that are associated
with the outcome of interest. For example, regularised regression has
been often used when building <em>epigenetic clocks</em>, where
methylation values across several thousands of genomic sites are used to
predict chronological age. We will explore this in more detail in the
<em>Regularised regression</em> episode.</p></li>
<li><p>Dimensionality reduction is commonly used on high-dimensional
datasets for data exploration or as a preprocessing step prior to other
downstream analyses. For instance, a low-dimensional visualisation of a
gene expression dataset may be used to inform <em>quality control</em>
steps (e.g. are there any anomalous samples?). This course contains two
episodes that explore dimensionality reduction techniques: <em>Principal
component analysis</em> and <em>Factor analysis</em>.</p></li>
<li><p>Clustering methods can be used to identify potential grouping
patterns within a dataset. A popular example is the <em>identification
of distinct cell types</em> through clustering cells with similar gene
expression patterns. The <em>K-means</em> episode will explore a
specific method to perform clustering analysis.</p></li>
</ol><div id="using-bioconductor-to-access-high-dimensional-data-in-the-biosciences" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="using-bioconductor-to-access-high-dimensional-data-in-the-biosciences" class="callout-inner">
<h3 class="callout-title">Using Bioconductor to access high-dimensional
data in the biosciences</h3>
<div class="callout-content">
<p>In this workshop, we will look at statistical methods that can be
used to visualise and analyse high-dimensional biological data using
packages available from Bioconductor, open source software for analysing
high throughput genomic data. Bioconductor contains useful packages and
example datasets as shown on the website <a href="https://www.bioconductor.org/" class="external-link uri">https://www.bioconductor.org/</a>.</p>
<p>Bioconductor packages can be installed and used in <code>R</code>
using the <strong><code>BiocManager</code></strong> package. Let’s load
the <strong><code>limma</code></strong> package from Bioconductor (a
package for running linear models).</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">browseVignettes</span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span></code></pre>
</div>
<p>We can explore these packages by browsing the vignettes provided in
Bioconductor. Bioconductor has various packages that can be used to load
and examine datasets in <code>R</code> that have been made available in
Bioconductor, usually along with an associated paper or package.</p>
<p>Next, we load the <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link"><code>methylation</code></a>
dataset which represents data collected using Illumina Infinium
methylation arrays which are used to examine methylation across the
human genome. These data include information collected from the assay as
well as associated metadata from individuals from whom samples were
taken.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methylation</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/methylation.rds"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">methylation</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in colData(methylation): could not find function "colData"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methyl_mat</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">methylation</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in assay(methylation): could not find function "assay"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate correlations between cells in matrix</span></span>
<span><span class="va">cor_mat</span> <span class="op">&lt;-</span> <span class="fu">cor</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="co"># print the top-left corner of the correlation matrix</span></span></code></pre>
</div>
<p>The <code>assay()</code> function creates a matrix-like object where
rows represent probes for genes and columns represent samples. We
calculate correlations between features in the <code>methylation</code>
dataset and examine the first 100 cells of this matrix. The size of the
dataset makes it difficult to examine in full, a common challenge in
analysing high-dimensional genomics data.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="further-reading-and-resources">Further reading and resources<a class="anchor" aria-label="anchor" href="#further-reading-and-resources"></a></h2>
<hr class="half-width"><ul><li>Buhlman, P. &amp; van de Geer, S. (2011) Statistics for
High-Dimensional Data. Springer, London.</li>
<li>
<a href="https://doi.org/10.1146/annurev-statistics-022513-115545" class="external-link">Buhlman,
P., Kalisch, M. &amp; Meier, L. (2014) High-dimensional statistics with
a view toward applications in biology. Annual Review of Statistics and
Its Application</a>.</li>
<li>Johnstone, I.M. &amp; Titterington, D.M. (2009) Statistical
challenges of high-dimensional data. Philosophical Transactions of the
Royal Society A 367:4237-4253.</li>
<li>
<a href="https://www.bioconductor.org/packages/release/workflows/vignettes/methylationArrayAnalysis/inst/doc/methylationArrayAnalysis.html" class="external-link">Bioconductor
ethylation array analysis vignette</a>.</li>
<li>The <em>Introduction to Machine Learning with Python</em> course
covers additional methods that could be used to analyse high-dimensional
data. See <a href="https://carpentries-incubator.github.io/machine-learning-novice-python/" class="external-link">Introduction
to machine learning</a>, <a href="https://carpentries-incubator.github.io/machine-learning-trees-python/" class="external-link">Tree
models</a> and <a href="https://carpentries-incubator.github.io/machine-learning-neural-python/" class="external-link">Neural
networks</a>. Some related (and important!) content is also available in
<a href="https://carpentries-incubator.github.io/machine-learning-responsible-python/" class="external-link">Responsible
machine learning</a>.</li>
<li>
<a href="https://www.youtube.com/c/joshstarmer" class="external-link">Josh Starmer’s</a>
youtube channel.</li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>High-dimensional data are data in which the number of features,
<span class="math inline">\(p\)</span>, are close to or larger than the
number of observations, <span class="math inline">\(n\)</span>.</li>
<li>These data are becoming more common in the biological sciences due
to increases in data storage capabilities and computing power.</li>
<li>Standard statistical methods, such as linear regression, run into
difficulties when analysing high-dimensional data.</li>
<li>In this workshop, we will explore statistical methods used for
analysing high-dimensional data using datasets available on
Bioconductor.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-high-dimensional-regression.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Home
        </a>
        <a class="chapter-link float-end" href="02-high-dimensional-regression.html" rel="next">
          Next: Regression with many...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/01-introduction-to-high-dimensional-data.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/01-introduction-to-high-dimensional-data.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Introduction to high-dimensional data",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/01-introduction-to-high-dimensional-data.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/01-introduction-to-high-dimensional-data.html",
  "dateCreated": "2025-10-02",
  "dateModified": "2025-09-02",
  "datePublished": "2025-10-02"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

