<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>High dimensional statistics with R: Regression with many outcomes</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/02-high-dimensional-regression.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 7%" class="percentage">
    7%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 7%" aria-valuenow="7" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/02-high-dimensional-regression.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction-to-high-dimensional-data.html">1. Introduction to high-dimensional data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Regression with many outcomes
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#dna-methylation-data">DNA methylation data</a></li>
<li><a href="#regression-with-many-outcomes">Regression with many outcomes</a></li>
<li><a href="#fitting-a-linear-model">Fitting a linear model</a></li>
<li><a href="#hypothesis-testing-in-linear-regression">Hypothesis testing in linear regression</a></li>
<li><a href="#fitting-a-lot-of-linear-models">Fitting a lot of linear models</a></li>
<li><a href="#sharing-information-across-outcome-variables">Sharing information across outcome variables</a></li>
<li><a href="#the-problem-of-multiple-tests">The problem of multiple tests</a></li>
<li><a href="#adjusting-for-multiple-tests">Adjusting for multiple tests</a></li>
<li><a href="#further-reading">Further reading</a></li>
<li><a href="#footnotes">Footnotes</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-principal-component-analysis.html">4. Principal component analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-k-means.html">6. K-means</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-hierarchical.html">7. Hierarchical clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>

                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-introduction-to-high-dimensional-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-regression-regularisation.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-introduction-to-high-dimensional-data.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="03-regression-regularisation.html" rel="next">
          Next: Regularised...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Regression with many outcomes</h1>
        <p>Last updated on 2025-04-07 |

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/02-high-dimensional-regression.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we apply linear regression in a high-dimensional
setting?</li>
<li>How can we benefit from the fact that we have many outcomes?</li>
<li>How can we control for the fact that we do many tests?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Perform and critically analyse high-dimensional regression.</li>
<li>Understand methods for shrinkage of noise parameters in
high-dimensional regression.</li>
<li>Perform multiple testing adjustment.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="dna-methylation-data">DNA methylation data<a class="anchor" aria-label="anchor" href="#dna-methylation-data"></a></h2>
<hr class="half-width"><p>For the following few episodes, we will be working with human DNA
methylation data from flow-sorted blood samples, described in <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link">data</a>.
DNA methylation assays measure, for each of many sites in the genome,
the proportion of DNA that carries a methyl mark (a chemical
modification that does not alter the DNA sequence). In this case, the
methylation data come in the form of a matrix of normalised methylation
levels (M-values), where negative values correspond to unmethylated DNA
and positive values correspond to methylated DNA. Along with this, we
have a number of sample phenotypes (eg, age in years, BMI).</p>
<p>Let’s read in the data for this episode:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methylation</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/methylation.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Note: if you want to view the code used to download these data, the
code is available in this repository in the <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/episodes/data/methylation.R" class="external-link"><code>methylation.R</code>
script</a></p>
<p><code>methylation</code> is actually a special Bioconductor
<code>SummarizedExperiment</code> object that summarises lots of
different information about the data. These objects are very useful for
storing all of the information about a dataset in a high-throughput
context. The structure of <code>SummarizedExperiment</code> objects is
described in the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html" class="external-link">vignettes
on Bioconductor</a>. Here, we show how to extract the information for
analysis.</p>
<p>We can extract</p>
<ul><li>the dimensions of the dataset using <code>dim()</code>. Importantly,
in these objects and data structures for computational biology in R
generally, observations are stored as columns and features (in this
case, sites in the genome) are stored as rows. This is in contrast to
usual tabular data, where features or variables are stored as columns
and observations are stored as rows;</li>
<li>assays, (normalised methylation levels here), using
<code>assay()</code>;</li>
<li>sample-level information using <code>colData()</code>.</li>
</ul><div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">methylation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: matrixStats</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: stats4</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: IRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: Biobase</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<p>You can see in this output that this object has a <code>dim()</code>
of <span class="math inline">\(5000 \\times 37\)</span>, meaning it has
5000 <em>features</em> and 37 <em>observations</em>. To extract the
matrix of methylation M-values, we can use the <code>assay()</code>
function.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methyl_mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">methylation</span><span class="op">)</span></span></code></pre>
</div>
<p>The distribution of these M-values looks like this:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">methyl_mat</span>, xlab <span class="op">=</span> <span class="st">"M-value"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-histx-1.png" alt="Histogram of M-values for all features. The distribution appears to be bimodal, with a large number of unmethylated features as well as many methylated features, and many intermediate features." class="figure mx-auto d-block"><figcaption>
Methylation levels are generally bimodally distributed.
</figcaption></figure><p>You can see that there are two peaks in this distribution,
corresponding to features which are largely unmethylated and methylated,
respectively.</p>
<p>Similarly, we can examine the <code>colData()</code>, which
represents the sample-level metadata we have relating to these data. In
this case, the metadata, phenotypes, and groupings in the
<code>colData</code> look like this for the first 6 samples:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">methylation</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<table class="table"><colgroup><col width="8%"><col width="8%"><col width="5%"><col width="2%"><col width="2%"><col width="7%"><col width="6%"><col width="6%"><col width="8%"><col width="11%"><col width="11%"><col width="5%"><col width="5%"><col width="9%"></colgroup><thead><tr class="header"><th align="left">Sample_Well</th>
<th align="left">Sample_Name</th>
<th align="right">purity</th>
<th align="left">Sex</th>
<th align="right">Age</th>
<th align="right">weight_kg</th>
<th align="right">height_m</th>
<th align="right">bmi</th>
<th align="left">bmi_clas</th>
<th align="left">Ethnicity_wide</th>
<th align="left">Ethnic_self</th>
<th align="left">smoker</th>
<th align="left">Array</th>
<th align="right">Slide</th>
</tr></thead><tbody><tr class="odd"><td align="left">A07</td>
<td align="left">PCA0612</td>
<td align="right">94</td>
<td align="left">M</td>
<td align="right">39</td>
<td align="right">88.45051</td>
<td align="right">1.8542</td>
<td align="right">25.72688</td>
<td align="left">Overweight</td>
<td align="left">Mixed</td>
<td align="left">Hispanic</td>
<td align="left">No</td>
<td align="left">R01C01</td>
<td align="right">201868500150</td>
</tr><tr class="even"><td align="left">C07</td>
<td align="left">NKpan2510</td>
<td align="right">95</td>
<td align="left">M</td>
<td align="right">49</td>
<td align="right">81.19303</td>
<td align="right">1.6764</td>
<td align="right">28.89106</td>
<td align="left">Overweight</td>
<td align="left">Indo-European</td>
<td align="left">Caucasian</td>
<td align="left">No</td>
<td align="left">R03C01</td>
<td align="right">201868500150</td>
</tr><tr class="odd"><td align="left">E07</td>
<td align="left">WB1148</td>
<td align="right">95</td>
<td align="left">M</td>
<td align="right">20</td>
<td align="right">80.28585</td>
<td align="right">1.7526</td>
<td align="right">26.13806</td>
<td align="left">Overweight</td>
<td align="left">Indo-European</td>
<td align="left">Persian</td>
<td align="left">No</td>
<td align="left">R05C01</td>
<td align="right">201868500150</td>
</tr><tr class="even"><td align="left">G07</td>
<td align="left">B0044</td>
<td align="right">97</td>
<td align="left">M</td>
<td align="right">49</td>
<td align="right">82.55381</td>
<td align="right">1.7272</td>
<td align="right">27.67272</td>
<td align="left">Overweight</td>
<td align="left">Indo-European</td>
<td align="left">Caucasian</td>
<td align="left">No</td>
<td align="left">R07C01</td>
<td align="right">201868500150</td>
</tr><tr class="odd"><td align="left">H07</td>
<td align="left">NKpan1869</td>
<td align="right">95</td>
<td align="left">F</td>
<td align="right">33</td>
<td align="right">87.54333</td>
<td align="right">1.7272</td>
<td align="right">29.34525</td>
<td align="left">Overweight</td>
<td align="left">Indo-European</td>
<td align="left">Caucasian</td>
<td align="left">No</td>
<td align="left">R08C01</td>
<td align="right">201868500150</td>
</tr><tr class="even"><td align="left">B03</td>
<td align="left">NKpan1850</td>
<td align="right">93</td>
<td align="left">F</td>
<td align="right">21</td>
<td align="right">87.54333</td>
<td align="right">1.6764</td>
<td align="right">31.15070</td>
<td align="left">Obese</td>
<td align="left">Mixed</td>
<td align="left">Finnish/Creole</td>
<td align="left">No</td>
<td align="left">R02C01</td>
<td align="right">201868590193</td>
</tr></tbody></table><p>In this episode, we will focus on the association between age and
methylation. The following heatmap summarises age and methylation levels
available in the methylation dataset:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"pheatmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">age</span> <span class="op">&lt;-</span> <span class="va">methylation</span><span class="op">$</span><span class="va">Age</span></span>
<span></span>
<span><span class="co"># sort methylation values by age </span></span>
<span><span class="va">order</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">age</span><span class="op">)</span></span>
<span><span class="va">age_ord</span> <span class="op">&lt;-</span> <span class="va">age</span><span class="op">[</span><span class="va">order</span><span class="op">]</span></span>
<span><span class="va">methyl_mat_ord</span> <span class="op">&lt;-</span> <span class="va">methyl_mat</span><span class="op">[</span>, <span class="va">order</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># plot heatmap</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">methyl_mat_ord</span>, </span>
<span>         cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_colnames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         legend_title <span class="op">=</span> <span class="st">"M-value"</span>, </span>
<span>         main <span class="op">=</span> <span class="st">"Feature vs Sample"</span>, </span>
<span>         annotation_col <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span>age <span class="op">=</span> <span class="va">age_ord</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in min(x): no non-missing arguments to min; returning Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in max(x): no non-missing arguments to max; returning -Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in seq.int(rx[1L], rx[2L], length.out = nb): 'from' must be a finite number</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1</h3>
<div class="callout-content">
<p>Why can we not just fit many linear regression models relating every
combination of feature (<code>colData</code> and assays) and draw
conclusions by associating all variables with significant model
p-values?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>There are a number of problems that this kind of approach presents.
For example:</p>
<ol style="list-style-type: decimal"><li>If we perform 5000 tests for each of 14 variables, even if there
were no true associations in the data, we’d be likely to observe some
strong spurious associations that arise just from random noise.</li>
<li>We may not have a representative sample for each of these
covariates. For example, we may have very small sample sizes for some
ethnicities, leading to spurious findings.</li>
<li>Without a research question in mind when creating a model, it’s not
clear how we can interpret each model, and rationalising the results
after the fact can be dangerous; it’s easy to make up a “story” that
isn’t grounded in anything but the fact that we have significant
findings.</li>
</ol></div>
</div>
</div>
</div>
<p>In general, it is scientifically interesting to explore two modelling
problems using the three types of data:</p>
<ol style="list-style-type: decimal"><li><p>Predicting methylation levels using age as a predictor. In this
case, we would have 5000 outcomes (methylation levels across the genome)
and a single covariate (age).</p></li>
<li><p>Predicting age using methylation levels as predictors. In this
case, we would have a single outcome (age) which will be predicted using
5000 covariates (methylation levels across the genome).</p></li>
</ol><p>The examples in this episode will focus on the first type of problem,
whilst the next episode will focus on the second.</p>
<div id="measuring-dna-methylation" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="measuring-dna-methylation" class="callout-inner">
<h3 class="callout-title">Measuring DNA Methylation</h3>
<div class="callout-content">
<p>DNA methylation is an epigenetic modification of DNA. Generally, we
are interested in the proportion of methylation at many sites or regions
in the genome. DNA methylation microarrays, as we are using here,
measure DNA methylation using two-channel microarrays, where one channel
captures signal from methylated DNA and the other captures unmethylated
signal. These data can be summarised as “Beta values” (<span class="math inline">\(\\beta\)</span> values), which is the ratio of the
methylated signal to the total signal (methylated plus unmethylated).
The <span class="math inline">\(\\beta\)</span> value for site <span class="math inline">\(i\)</span> is calculated as</p>
<p><span class="math display">\[
\beta_i = \frac{
m_i
} {
u_{i} + m_{i}
}
\]</span></p>
<p>where <span class="math inline">\(m\_i\)</span> is the methylated
signal for site <span class="math inline">\(i\)</span> and <span class="math inline">\(u\_i\)</span> is the unmethylated signal for site
<span class="math inline">\(i\)</span>. <span class="math inline">\(\\beta\)</span> values take on a value in the
range <span class="math inline">\([0, 1]\)</span>, with 0 representing a
completely unmethylated site and 1 representing a completely methylated
site.</p>
<p>The M-values we use here are the <span class="math inline">\(\\log\_2\)</span> ratio of methylated versus
unmethylated signal:</p>
<p><span class="math display">\[
M_i = \log_2\left(\frac{m_i}{u_i}\right)
\]</span></p>
<p>M-values are not bounded to an interval as Beta values are, and
therefore can be easier to work with in statistical models.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="regression-with-many-outcomes">Regression with many outcomes<a class="anchor" aria-label="anchor" href="#regression-with-many-outcomes"></a></h2>
<hr class="half-width"><p>In high-throughput studies, it is common to have one or more
phenotypes or groupings that we want to relate to features of interest
(eg, gene expression, DNA methylation levels). In general, we want to
identify differences in the features of interest that are related to a
phenotype or grouping of our samples. Identifying features of interest
that vary along with phenotypes or groupings can allow us to understand
how phenotypes arise or manifest. Analysis of this type is sometimes
referred to using the term <em>differential analysis</em>.</p>
<p>For example, we might want to identify genes that are expressed at a
higher level in mutant mice relative to wild-type mice to understand the
effect of a mutation on cellular phenotypes. Alternatively, we might
have samples from a set of patients, and wish to identify epigenetic
features that are different in young patients relative to old patients,
to help us understand how ageing manifests.</p>
<p>Using linear regression, it is possible to identify differences like
these. However, high-dimensional data like the ones we’re working with
require some special considerations. A first consideration, as we saw
above, is that there are far too many features to fit each one-by-one as
we might do when analysing low-dimensional datasets (for example using
<code>lm()</code> on each feature and checking the linear model
assumptions). A second consideration is that statistical approaches may
behave slightly differently when applied to very high-dimensional data,
compared to low-dimensional data. A third consideration is the speed at
which we can actually compute statistics for data this large – methods
optimised for low-dimensional data may be very slow when applied to
high-dimensional data.</p>
<p>Ideally when performing regression, we want to identify cases like
this, where there is a clear association, and we probably “don’t need”
statistics:</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example1-1.png" alt="An example of a strong linear association between a continuous phenotype (age) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. A strong linear relationship with a positive slope exists between the two." class="figure mx-auto d-block"><figcaption>
A scatter plot of age and a feature of interest.
</figcaption></figure><p>or equivalently for a discrete covariate:</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example2-1.png" alt="An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups clearly differ with respect to DNA methylation." class="figure mx-auto d-block"><figcaption>
A scatter plot of a grouping and a feature of interest.
</figcaption></figure><p>However, often due to small differences and small sample sizes, the
problem is more difficult:</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example3-1.png" alt="An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups seem to differ with respect to DNA methylation, but the relationship is weak." class="figure mx-auto d-block"><figcaption>
A scatter plot of a grouping and a feature of interest.
</figcaption></figure><p>And, of course, we often have an awful lot of features and need to
prioritise a subset of them! We need a rigorous way to prioritise genes
for further analysis.</p>
</section><section><h2 class="section-heading" id="fitting-a-linear-model">Fitting a linear model<a class="anchor" aria-label="anchor" href="#fitting-a-linear-model"></a></h2>
<hr class="half-width"><p>So, in the data we have read in, we have a matrix of methylation
values <span class="math inline">\(X\)</span> and a vector of ages,
<span class="math inline">\(y\)</span>. One way to model this is to see
if we can use age to predict the expected (average) methylation value
for sample <span class="math inline">\(j\)</span> at a given locus <span class="math inline">\(i\)</span>, which we can write as <span class="math inline">\(X\_{ij}\)</span>. We can write that model as:</p>
<p><span class="math display">\[
\mathbf{E}(X_{ij}) = \beta_0 + \beta_1 \text{Age}_j
\]</span></p>
<p>where <span class="math inline">\(\\text{Age}\_j\)</span> is the age
of sample <span class="math inline">\(j\)</span>. In this model, <span class="math inline">\(\\beta\_1\)</span> represents the unit change in
mean methylation level for each unit (year) change in age. For a
specific CpG, we can fit this model and get more information from the
model object. For illustration purposes, here we arbitrarily select the
first CpG in the <code>methyl_mat</code> matrix (the one on its first
row).</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">age</span> <span class="op">&lt;-</span> <span class="va">methylation</span><span class="op">$</span><span class="va">Age</span></span>
<span><span class="co"># methyl_mat[1, ] indicates that the 1st CpG will be used as outcome variable</span></span>
<span><span class="va">lm_age_methyl1</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span> <span class="op">~</span> <span class="va">age</span><span class="op">)</span></span>
<span><span class="va">lm_age_methyl1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = methyl_mat[1, ] ~ age)

Coefficients:
(Intercept)          age
   0.902334     0.008911  </code></pre>
</div>
<p>We now have estimates for the expected methylation level when age
equals 0 (the intercept) and the change in methylation level for a unit
change in age (the slope). We could plot this linear model:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">age</span>, <span class="va">methyl_mat</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, xlab <span class="op">=</span> <span class="st">"Age"</span>, ylab <span class="op">=</span> <span class="st">"Methylation level"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-lm-methyl1-1.png" alt="An example of the relationship between age (x-axis) and methylation levels (y-axis) for an arbitrarily selected CpG. In this case, the y-axis shows methylation levels for the first CpG in our data. The black line shows the fitted regression line (based on the intercept and slope estimates shown above). For this feature, we can see that there is no strong relationship between methylation and age." class="figure mx-auto d-block"><figcaption>
A scatter plot of age versus the methylation level for an arbitrarily
selected CpG side (the one stored as the first column of methyl_mat).
Each dot represents an individual. The black line represents the
estimated linear model.
</figcaption></figure><p>For this feature, we can see that there is no strong relationship
between methylation and age. We could try to repeat this for every
feature in our dataset; however, we have a lot of features! We need an
approach that allows us to assess associations between all of these
features and our outcome while addressing the three considerations we
outlined previously. Before we introduce this approach, let’s go into
detail about how we generally check whether the results of a linear
model are statistically significant.</p>
</section><section><h2 class="section-heading" id="hypothesis-testing-in-linear-regression">Hypothesis testing in linear regression<a class="anchor" aria-label="anchor" href="#hypothesis-testing-in-linear-regression"></a></h2>
<hr class="half-width"><p>Using the linear model we defined above, we can ask questions based
on the estimated value for the regression coefficients. For example, do
individuals with different age have different methylation values for a
given CpG? We usually do this via <em>hypothesis testing</em>. This
framework compares the results that we observed (here, estimated linear
model coefficients) to the results you would expect under a <em>null
hypothesis</em> associated to our question. In the example above, a
suitable null hypothesis would test whether the regression coefficient
associated to age (<span class="math inline">\(\\beta\_1\)</span>) is
equal to zero or not. If <span class="math inline">\(\\beta\_1\)</span>
is equal to zero, the linear model indicates that there is no linear
relationship between age and the methylation level for the CpG
(remember: as its name suggests, linear regression can only be used to
model linear relationships between predictors and outcomes!). In other
words, the answer to our question would be: no!</p>
<p>The output of a linear model typically returns the results associated
with the null hypothesis described above (this may not always be the
most realistic or useful null hypothesis, but it is the one we have by
default!). To be more specific, the test compares our observed results
with a set of hypothetical counter-examples of what we would expect to
observe if we repeated the same experiment and analysis over and over
again under the null hypothesis.</p>
<p>For this linear model, we can use <code>tidy()</code> from the
<strong><code>broom</code></strong> package to extract detailed
information about the coefficients and the associated hypothesis tests
in this model:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"broom"</span><span class="op">)</span></span>
<span><span class="fu">tidy</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 5
  term        estimate std.error statistic p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
1 (Intercept)  0.902      0.344      2.62   0.0129
2 age          0.00891    0.0100     0.888  0.381 </code></pre>
</div>
<p>The standard errors (<code>std.error</code>) represent the
statistical uncertainty in our regression coefficient estimates (often
referred to as <em>effect size</em>). The test statistics and p-values
(<code>statistic</code> and <code>p.value</code>) represent measures of
how (un)likely it would be to observe results like this under the “null
hypothesis”. For coefficient <span class="math inline">\(k\)</span> in a
linear model (in our case, it would be the slope), the test statistic
calculated in <code>statistic</code> above is a t-statistic given
by:</p>
<p><span class="math display">\[
t_{k} = \frac{\hat{\beta}_{k}}{SE\left(\hat{\beta}_{k}\right)}
\]</span></p>
<p><span class="math inline">\(SE\\left(\\hat{\\beta}\_{k}\\right)\)</span>
measures the uncertainty we have in our effect size estimate. Knowing
what distribution these t-statistics follow under the null hypothesis
allows us to determine how unlikely it would be for us to observe what
we have under those circumstances, if we repeated the experiment and
analysis over and over again. To demonstrate how the t-statistics are
calculated, we can compute them “by hand”:</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">table_age_methyl1</span> <span class="op">&lt;-</span> <span class="fu">tidy</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span></span>
<span><span class="va">tvals</span> <span class="op">&lt;-</span> <span class="va">table_age_methyl1</span><span class="op">$</span><span class="va">estimate</span> <span class="op">/</span> <span class="va">table_age_methyl1</span><span class="op">$</span><span class="va">std.error</span></span>
<span><span class="fu">all.equal</span><span class="op">(</span><span class="va">tvals</span>, <span class="va">table_age_methyl1</span><span class="op">$</span><span class="va">statistic</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>We can see that the t-statistic is just the ratio between the
coefficient estimate and the standard error.</p>
<p>Calculating the p-values is a bit more tricky. Specifically, it is
the proportion of the distribution of the test statistic under the null
hypothesis that is <em>as extreme or more extreme</em> than the observed
value of the test statistic. This is easy to observe visually, by
plotting the theoretical distribution of the test statistic under the
null hypothesis (see next call-out box for more details about it):</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-tdist-1.png" alt="Density plot of a t-distribution showing the observed test statistics (here, t-statistics). The p-values, visualised here with shaded regions, represent the portion of the null distribution that is as extreme or more extreme as the observed test statistics, which are shown as dashed lines." class="figure mx-auto d-block"><figcaption>
The p-value for a regression coefficient represents how often it’d be
observed under the null.
</figcaption></figure><p>The red-ish shaded region represents the portion of the distribution
of the test statistic under the null hypothesis that is equal or greater
to the value we observe for the intercept term. As our null hypothesis
relates to a 2-tailed test (as the null hypothesis states that the
regression coefficient is equal to zero, we would reject it if the
regression coefficient is substantially larger <strong>or</strong>
smaller than zero), the p-value for the test is twice the value of the
shaded region. In this case, the shaded region is small relative to the
total area of the null distribution; therefore, the p-value is small
(<span class="math inline">\(p=0.013\)</span>). The blue-ish shaded
region represents the same measure for the slope term; this is larger,
relative to the total area of the distribution, therefore the p-value is
larger than the one for the intercept term (<span class="math inline">\(p=0.381\)</span>). The p-value is a function of
the test statistic: the ratio between the effect size we’re estimating
and the uncertainty we have in that effect. A large effect with large
uncertainty may not lead to a small p-value, and a small effect with
small uncertainty may lead to a small p-value.</p>
<div id="calculating-p-values-from-a-linear-model" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="calculating-p-values-from-a-linear-model" class="callout-inner">
<h3 class="callout-title">Calculating p-values from a linear model</h3>
<div class="callout-content">
<p>Manually calculating the p-value for a linear model is a little bit
more complex than calculating the t-statistic. The intuition posted
above is definitely sufficient for most cases, but for completeness,
here is how we do it:</p>
<p>Since the statistic in a linear model is a t-statistic, it follows a
student t distribution under the null hypothesis, with degrees of
freedom (a parameter of the student t-distribution) given by the number
of observations minus the number of coefficients fitted, in this case
<span class="math inline">\(37 - 2 = 35\)</span>. We want to know what
portion of the distribution function of the test statistic is as extreme
as, or more extreme than, the value we observed. The function
<code>pt()</code>(similar to<code>pnorm()</code>, etc) can give us this
information.</p>
<p>Since we’re not sure if the coefficient will be larger or smaller
than zero, we want to do a 2-tailed test. Therefore we take the absolute
value of the t-statistic, and look at the upper rather than lower tail.
In the figure above the shaded areas are only looking at “half” of the
t-distribution (which is symmetric around zero), therefore we multiply
the shaded area by 2 in order to calculate the p-value.</p>
<p>Combining all of this gives us:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">*</span> <span class="fu">pt</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">tvals</span><span class="op">)</span>, df <span class="op">=</span> <span class="va">lm_age_methyl1</span><span class="op">$</span><span class="va">df</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">all.equal</span><span class="op">(</span><span class="va">table_age_methyl1</span><span class="op">$</span><span class="va">p.value</span>, <span class="va">pvals</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
</div>
</div>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2</h3>
<div class="callout-content">
<p>In the model we fitted, the estimate for the intercept is 0.902 and
its associated p-value is 0.0129. What does this mean?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>The first coefficient in a linear model like this is the intercept,
which measures the mean of the outcome (in this case, the methylation
value for the first CpG) when age is zero. In this case, the intercept
estimate is 0.902. However, this is not a particularly noteworthy
finding as we do not have any observations with age zero (nor even any
with age &lt; 20!).</p>
<p>The reported p-value is associated to the following null hypothesis:
the intercept (<span class="math inline">\(\\beta\_0\)</span> above) is
equal to zero. Using the usual significance threshold of 0.05, we reject
the null hypothesis as the p-value is smaller than 0.05. However, it is
not really interesting if this intercept is zero or not, since we
probably do not care what the methylation level is when age is zero. In
fact, this question does not even make much sense! In this example, we
are more interested in the regression coefficient associated to age, as
that can tell us whether there is a linear relationship between age and
methylation for the CpG.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="fitting-a-lot-of-linear-models">Fitting a lot of linear models<a class="anchor" aria-label="anchor" href="#fitting-a-lot-of-linear-models"></a></h2>
<hr class="half-width"><p>In the linear model above, we are generally interested in the second
regression coefficient (often referred to as <em>slope</em>) which
measures the linear relationship between age and methylation levels. For
the first CpG, here is its estimate:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coef_age_methyl1</span> <span class="op">&lt;-</span> <span class="fu">tidy</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="va">coef_age_methyl1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 5
  term  estimate std.error statistic p.value
  &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
1 age    0.00891    0.0100     0.888   0.381</code></pre>
</div>
<p>In this case, the p-value is equal to 0.381 and therefore we cannot
reject the null hypothesis: there is no statistical evidence to suggest
that the regression coefficient associated to age is not equal to
zero.</p>
<p>Now, we could do this for every feature (CpG) in the dataset and rank
the results based on their test statistic or associated p-value.
However, fitting models in this way to 5000 features is not very
computationally efficient, and it would also be laborious to do
programmatically. There are ways to get around this, but first let us
talk about what exactly we are doing when we look at significance tests
in this context.</p>
</section><section><h2 class="section-heading" id="sharing-information-across-outcome-variables">Sharing information across outcome variables<a class="anchor" aria-label="anchor" href="#sharing-information-across-outcome-variables"></a></h2>
<hr class="half-width"><p>We are going to introduce an idea that allows us to take advantage of
the fact that we carry out many tests at once on structured data. We can
leverage this fact to <em>share information</em> between model
parameters. The insight that we use to perform <em>information
pooling</em> or sharing is derived from our knowledge about the
structure of the data. For example, in a high-throughput experiment like
a DNA methylation assay, we know that all of the features were measured
simultaneously, using the same technique. This means that generally, we
expect the base-level variability for each feature to be broadly
similar.</p>
<p>This can enable us to get a better estimate of the uncertainty of
model parameters than we could get if we consider each feature in
isolation. So, to share information between features allows us to get
more robust estimators. Remember that the t-statistic for coefficient
<span class="math inline">\(\\beta\_k\)</span> in a linear model is the
ratio between the coefficient estimate and its standard error:</p>
<p><span class="math display">\[
t_{k} = \frac{\hat{\beta}_{k}}{SE\left(\hat{\beta}_{k}\right)}
\]</span></p>
<p>It is clear that large effect sizes will likely lead to small
p-values, as long as the standard error for the coefficent is not large.
However, the standard error is affected by the amount of noise, as we
saw earlier. If we have a small number of observations, it is common for
the noise for some features to be extremely small simply by chance.
This, in turn, causes small p-values for these features, which may give
us unwarranted confidence in the level of certainty we have in the
results (false positives).</p>
<p>There are many statistical methods in genomics that use this type of
approach to get better estimates by pooling information between features
that were measured simultaneously using the same techniques. Here we
will focus on the package <strong><code>limma</code></strong>, which is
an established software package used to fit linear models, originally
for the gene expression micro-arrays that were common in the 2000s, but
which is still in use in RNAseq experiments, among others. The authors
of <strong><code>limma</code></strong> made some assumptions about the
distributions that these follow, and pool information across genes to
get a better estimate of the uncertainty in effect size estimates. It
uses the idea that noise levels should be similar between features to
<em>moderate</em> the estimates of the test statistic by shrinking the
estimates of standard errors towards a common value. This results in a
<em>moderated t-statistic</em>.</p>
<p>The process of running a model in <strong><code>limma</code></strong>
is somewhat different to what you may have seen when running linear
models. Here, we define a <em>model matrix</em> or <em>design
matrix</em>, which is a way of representing the coefficients that should
be fit in each linear model. These are used in similar ways in many
different modelling libraries.</p>
<div id="what-is-a-model-matrix" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="what-is-a-model-matrix" class="callout-inner">
<h3 class="callout-title">What is a model matrix?</h3>
<div class="callout-content">
<p>R fits a regression model by choosing the vector of regression
coefficients that minimises the differences between outcome values and
predicted values using the covariates (or predictor variables). To get
predicted values, we multiply the matrix of predictors by the
coefficients. The latter matrix is called the <strong>model
matrix</strong> (or design matrix). It has one row for each observation
and one column for each predictor, plus one additional column of ones
(the intercept column). Many R libraries contruct the model matrix
behind the scenes, but <strong><code>limma</code></strong> does not.
Usually, the model matrix can be extracted from a model fit using the
function <code>model.matrix()</code>. Here is an example of a model
matrix for the methylation model:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_age</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span> <span class="co"># model matrix</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">design_age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                    (Intercept) age
201868500150_R01C01           1  39
201868500150_R03C01           1  49
201868500150_R05C01           1  20
201868500150_R07C01           1  49
201868500150_R08C01           1  33
201868590193_R02C01           1  21</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">design_age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 37  2</code></pre>
</div>
<p>As you can see, the model matrix has the same number of rows as our
methylation data has samples. It also has two columns - one for the
intercept (similar to the linear model we fit above) and one for age.
This happens “under the hood” when fitting a linear model with
<code>lm()</code>, but here we have to specify it directly. The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf" class="external-link">limma
user manual</a> has more detail on how to make model matrices for
different types of experimental design, but here we are going to stick
with this simple two-variable case.</p>
</div>
</div>
</div>
<p>We pass our methylation data to <code>lmFit()</code>, specifying the
model matrix. Internally, this function runs <code>lm()</code> on each
row of the data in an efficient way. The function <code>eBayes()</code>,
when applied to the output of <code>lmFit()</code>, performs the pooled
estimation of standard errors that results in the moderated t-statistics
and resulting p-values.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'limma'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:BiocGenerics':

    plotMA</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_age</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="va">lm_age_methyl1</span><span class="op">)</span> <span class="co"># model matrix</span></span>
<span><span class="va">fit_age</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">methyl_mat</span>, design <span class="op">=</span> <span class="va">design_age</span><span class="op">)</span></span>
<span><span class="va">fit_age</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit_age</span><span class="op">)</span></span></code></pre>
</div>
<p>To obtain the results of the linear models, we can use the
<code>topTable()</code> function. By default, this returns results for
the first coefficient in the model. As we saw above when using
<code>lm()</code>, and when we defined <code>design_age</code> above,
the first coefficient relates to the intercept term, which we are not
particularly interested in here; therefore we specify
<code>coef = 2</code>. Further, <code>topTable()</code> by default only
returns the top 10 results. To see all of the results in the data, we
specify <code>number = nrow(fit_age)</code> to ensure that it returns a
row for every row of the input matrix.</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toptab_age</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit_age</span>, coef <span class="op">=</span> <span class="fl">2</span>, number <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">fit_age</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">toptab_age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 logFC    AveExpr         t      P.Value   adj.P.Val        B
cg08446924 -0.02571353 -0.4185868 -6.039068 5.595675e-07 0.002797837 5.131574
cg06493994  0.01550941 -2.1057877  5.593988 2.239813e-06 0.005599533 3.747986
cg17661642  0.02266668 -2.0527722  5.358739 4.658336e-06 0.006048733 3.019698
cg05168977  0.02276336 -2.2918472  5.346500 4.838987e-06 0.006048733 2.981904
cg24549277  0.01975577 -1.7466088  4.939242 1.708355e-05 0.011508818 1.731821
cg04436528 -0.01943612  0.7033503 -4.917179 1.828563e-05 0.011508818 1.664608</code></pre>
</div>
<p>The output of <code>topTable</code> includes the coefficient, here
termed a log fold change <code>logFC</code>, the average level
(<code>aveExpr</code>), the t-statistic <code>t</code>, the p-value
(<code>P.Value</code>), and the <em>adjusted</em> p-value
(<code>adj.P.Val</code>). We’ll cover what an adjusted p-value is very
shortly. The table also includes <code>B</code>, which represents the
log-odds that a feature is signficantly different, which we won’t cover
here, but which will generally be a 1-1 transformation of the p-value.
The coefficient estimates here are termed <code>logFC</code> for legacy
reasons relating to how microarray experiments were traditionally
performed. There are more details on this topic in many places, for
example <a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html" class="external-link">this
tutorial by Kasper D. Hansen</a></p>
<p>Now we have estimates of effect sizes and p-values for the
association between methylation level at each locus and age for our 37
samples. It’s useful to create a plot of effect size estimates (model
coefficients) against p-values for each of these linear models, to
visualise the magnitude of effects and the statistical significance of
each. These plots are often called “volcano plots”, because they
resemble an eruption.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">toptab_age</span><span class="op">$</span><span class="va">logFC</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">toptab_age</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Effect size"</span>, ylab <span class="op">=</span> <span class="fu">bquote</span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">-</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-limmavolc1-1.png" alt="A plot of -log10(p) against effect size estimates for a regression of age against methylation using limma." class="figure mx-auto d-block"><figcaption>
Plotting p-values against effect sizes using limma; the results are
similar to a standard linear model.
</figcaption></figure><p>In this figure, every point represents a feature of interest. The
x-axis represents the effect size observed for that feature in a linear
model, while the y-axis is the <span class="math inline">\(-\\log\_{10}(\\text{p-value})\)</span>, where
larger values indicate increasing statistical evidence of a non-zero
effect size. A positive effect size represents increasing methylation
with increasing age, and a negative effect size represents decreasing
methylation with increasing age. Points higher on the y-axis represent
features for which we think the results we observed would be very
unlikely under the null hypothesis.</p>
<p>Since we want to identify features that have different methylation
levels in different age groups, in an ideal case there would be clear
separation between “null” and “non-null” features. However, usually we
observe results as we do here: there is a continuum of effect sizes and
p-values, with no clear separation between these two classes of
features. While statistical methods exist to derive insights from
continuous measures like these, it is often convenient to obtain a list
of features which we are confident have non-zero effect sizes. This is
made more difficult by the number of tests we perform.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3</h3>
<div class="callout-content">
<p>The effect size estimates are very small, and yet many of the
p-values are well below a usual significance level of p &lt; 0.05. Why
is this?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>Because age has a much larger range than methylation levels, the unit
change in methylation level even for a strong relationship is very
small!</p>
<p>As we mentioned, the p-value is a function of both the effect size
estimate and the uncertainty (standard error) of that estimate. Because
the uncertainty in our estimates is much smaller than the estimates
themselves, the p-values are also small.</p>
<p>If we predicted age using methylation level, it is likely we would
see much larger coefficients, though broadly similar p-values!</p>
</div>
</div>
</div>
</div>
<p>It is worthwhile considering what exactly the effect of the
<em>moderation</em> or information sharing that
<strong><code>limma</code></strong> performs has on our results. To do
this, let us compare the effect sizes estimates and p-values from the
two approaches.</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-limma-lm-effect-1.png" alt="A scatter plot of the effect size using limmma vs. those using lm. The plot also shows a straight line through all points showing that the effect sizes are the same." class="figure mx-auto d-block"><figcaption>
Plot of effect sizes using limma vs. those using lm.
</figcaption></figure><p>These are exactly identical! This is because
<strong><code>limma</code></strong> does not perform any sharing of
information when estimating effect sizes. This is in contrast to similar
packages that apply shrinkage to the effect size estimates, like
<strong><code>DESeq2</code></strong>. These often use information
sharing to shrink or moderate the effect size estimates, in the case of
<strong><code>DESeq2</code></strong> by again sharing information
between features about sample-to-sample variability. In contrast, let us
look at the p-values from <strong><code>limma</code></strong> and R’s
built-in <code>lm()</code> function:</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-limma-lm-pval-1.png" alt="A scatter plot of the p-values using limma vs. those using lm. A straight line is also displayed, showing that the p-values for limma tend to be smaller than those using lm towards the left of the plot and higher towards the right of the plot." class="figure mx-auto d-block"><figcaption>
Plot of p-values using limma vs. those using lm.
</figcaption></figure><p>We can see that for the vast majority of features, the results are
broadly similar. There seems to be a minor general tendency for
<strong><code>limma</code></strong> to produce smaller p-values, but for
several features, the p-values from limma are considerably larger than
the p-values from <code>lm()</code>. This is because the information
sharing tends to shrink large standard error estimates downwards and
small estimates upwards. When the degree of statistical significance is
due to an abnormally small standard error rather than a large effect,
this effect results in this prominent reduction in statistical
significance, which has been shown to perform well in case studies. The
degree of shrinkage generally depends on the amount of pooled
information and the strength of the evidence independent of pooling. For
example, with very few samples and many features, information sharing
has a larger effect, because there are a lot of genes that can be used
to provide pooled estimates, and the evidence from the data that this is
weighed against is relatively sparse. In contrast, when there are many
samples and few features, there is not much opportunity to generate
pooled estimates, and the evidence of the data can easily outweigh the
pooling.</p>
<p>Shrinkage methods like these ones can be complex to implement and
understand, but it is useful to develop an intuition about why these
approaches may be more precise and sensitive than the naive approach of
fitting a model to each feature separately.</p>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Try to run the same kind of linear model with smoking status as
covariate instead of age, and making a volcano plot. <em>Note: smoking
status is stored as</em> <code>methylation$smoker</code>.</li>
<li>We saw in the example in the lesson that this information sharing
can lead to larger p-values. Why might this be preferable?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>The following code runs the same type of model with smoking
status:</li>
</ol><div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_smoke</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="va">methylation</span><span class="op">$</span><span class="va">smoker</span><span class="op">)</span></span>
<span><span class="va">fit_smoke</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">methyl_mat</span>, design <span class="op">=</span> <span class="va">design_smoke</span><span class="op">)</span></span>
<span><span class="va">fit_smoke</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit_smoke</span><span class="op">)</span></span>
<span><span class="va">toptab_smoke</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit_smoke</span>, coef <span class="op">=</span> <span class="fl">2</span>, number <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">fit_smoke</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">toptab_smoke</span><span class="op">$</span><span class="va">logFC</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">toptab_smoke</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Effect size"</span>, ylab <span class="op">=</span> <span class="fu">bquote</span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-limmavolc2-1.png" alt="A plot of -log10(p) against effect size estimates for a regression of smoking status against methylation using limma." class="figure mx-auto d-block"><figcaption>
A plot of significance against effect size for a regression of smoking
against methylation.
</figcaption></figure><ol start="2" style="list-style-type: decimal"><li>Being a bit more conservative when identifying features can help to
avoid false discoveries. Furthermore, when rejecting the null hypothesis
is based more on a small standard error resulting from abnormally low
levels of variability for a given feature, we might want to be a bit
more conservative in our expectations.</li>
</ol></div>
</div>
</div>
</div>
<div id="shrinkage" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="shrinkage" class="callout-inner">
<h3 class="callout-title">Shrinkage</h3>
<div class="callout-content">
<p>Shrinkage is an intuitive term for an effect of information sharing,
and is something observed in a broad range of statistical models. Often,
shrinkage is induced by a <em>multilevel</em> modelling approach or by
<em>Bayesian</em> methods.</p>
<p>The general idea is that these models incorporate information about
the structure of the data into account when fitting the parameters. We
can share information between features because of our knowledge about
the data structure; this generally requires careful consideration about
how the data were generated and the relationships within.</p>
<p>An example people often use is estimating the effect of attendance on
grades in several schools. We can assume that this effect is similar in
different schools (but maybe not identical), so we can <em>share
information</em> about the effect size between schools and shrink our
estimates towards a common value.</p>
<p>For example in <strong><code>DESeq2</code></strong>, the authors used
the observation that genes with similar expression counts in RNAseq data
have similar <em>dispersion</em>, and a better estimate of these
dispersion parameters makes estimates of fold changes much more stable.
Similarly, in <strong><code>limma</code></strong> the authors made the
assumption that in the absence of biological effects, we can often
expect the technical variation in the measurement of the expression of
each of the genes to be broadly similar. Again, better estimates of
variability allow us to prioritise genes in a more reliable way.</p>
<p>There are many good resources to learn about this type of approach,
including:</p>
<ul><li><a href="https://www.tjmahr.com/plotting-partial-pooling-in-mixed-effects-models/" class="external-link">a
blog post by TJ Mahr</a></li>
<li><a href="https://gumroad.com/l/empirical-bayes" class="external-link">a book by David
Robinson</a></li>
<li><a href="https://www.stat.columbia.edu/~gelman/arm/" class="external-link">a (relatively
technical) book by Gelman and Hill</a></li>
</ul></div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-problem-of-multiple-tests">The problem of multiple tests<a class="anchor" aria-label="anchor" href="#the-problem-of-multiple-tests"></a></h2>
<hr class="half-width"><p>With such a large number of features, it would be useful to decide
which features are “interesting” or “significant” for further study.
However, if we were to apply a normal significance threshold of 0.05, it
would be likely we end up with a lot of false positives. This is because
a p-value threshold like this represents a <span class="math inline">\(\\frac{1}{20}\)</span> chance that we observe
results as extreme or more extreme under the null hypothesis (that there
is no assocation between age and methylation level). If we carry out
many more than 20 such tests, we can expect to see situations where,
despite the null hypothesis being true, we observe observe signifiant
p-values due to random chance. To demonstrate this, it is useful to see
what happens if we permute (scramble) the age values and run the same
test again:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span> </span>
<span><span class="va">age_perm</span> <span class="op">&lt;-</span> <span class="va">age</span><span class="op">[</span><span class="fu">sample</span><span class="op">(</span><span class="fu">ncol</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span>, <span class="fu">ncol</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">design_age_perm</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="va">age_perm</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_age_perm</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">methyl_mat</span>, design <span class="op">=</span> <span class="va">design_age_perm</span><span class="op">)</span></span>
<span><span class="va">fit_age_perm</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit_age_perm</span><span class="op">)</span></span>
<span><span class="va">toptab_age_perm</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit_age_perm</span>, coef <span class="op">=</span> <span class="fl">2</span>, number <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">fit_age_perm</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">toptab_age_perm</span><span class="op">$</span><span class="va">logFC</span>, <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">toptab_age_perm</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Effect size"</span>, ylab <span class="op">=</span> <span class="fu">bquote</span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-volcplotfake-1.png" alt="Plot of -log10(p) against effect size estimates for a regression of a made-up feature against methylation level for each feature in the data. A dashed line represents a 0.05 significance level." class="figure mx-auto d-block"><figcaption>
Plotting p-values against effect sizes for a randomised outcome shows we
still observe ‘significant’ results.
</figcaption></figure><p>Since we have generated a random sequence of ages, we have no reason
to suspect that there is a true association between methylation levels
and this sequence of random numbers. However, you can see that the
p-value for many features is still lower than a traditional significance
level of <span class="math inline">\(p=0.05\)</span>. In fact, here 226
features are significant at p &lt; 0.05. If we were to use this fixed
threshold in a real experiment, it is likely that we would identify many
features as associated with age, when the results we are observing are
simply due to chance.</p>
<div id="challenge-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-5" class="callout-inner">
<h3 class="callout-title">Challenge 5</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>If we run 5000 tests, even if there are no true differences, how
many of them (on average) will be statistically significant at a
threshold of <span class="math inline">\(p \&lt; 0.05\)</span>?</li>
<li>Why would we want to be conservative in labelling features as
significantly different? By conservative, we mean to err towards
labelling true differences as “not significant” rather than vice
versa.</li>
<li>How could we account for a varying number of tests to ensure
“significant” changes are truly different?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>By default we expect <span class="math inline">\(5000 \\times 0.05 =
250\)</span> features to be statistically significant under the null
hypothesis, because p-values should always be uniformly distributed
under the null hypothesis.</li>
<li>Features that we label as “significantly different” will often be
reported in manuscripts. We may also spend time and money investigating
them further, computationally or in the lab. Therefore, spurious results
have a real cost for ourselves and for others.</li>
<li>One approach to controlling for the number of tests is to divide our
significance threshold by the number of tests performed. This is termed
“Bonferroni correction” and we’ll discuss this further now.</li>
</ol></div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="adjusting-for-multiple-tests">Adjusting for multiple tests<a class="anchor" aria-label="anchor" href="#adjusting-for-multiple-tests"></a></h2>
<hr class="half-width"><p>When performing many statistical tests to categorise features, we are
effectively classifying features as “non-significant” or “significant”,
that latter meaning those for which we reject the null hypothesis. We
also generally hope that there is a subset of features for which the
null hypothesis is truly false, as well as many for which the null truly
does hold. We hope that for all features for which the null hypothesis
is true, we accept it, and for all features for which the null
hypothesis is not true, we reject it. As we showed in the example with
permuted age, with a large number of tests it is inevitable that we will
get some of these wrong.</p>
<p>We can think of these features as being “truly different” or “not
truly different”<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Using this idea, we can see that each
categorisation we make falls into four categories:</p>
<table class="table"><colgroup><col width="43%"><col width="38%"><col width="17%"></colgroup><thead><tr class="header"><th align="right">True outcome</th>
<th align="right">Label as different</th>
<th align="right">Label as not different</th>
</tr></thead><tbody><tr class="odd"><td align="right">Truly different</td>
<td align="right">True positive</td>
<td align="right">False negative</td>
</tr><tr class="even"><td align="right">Truly not different</td>
<td align="right">False positive</td>
<td align="right">True negative</td>
</tr></tbody></table><p>If the null hypothesis was true for every feature, then as we perform
more and more tests we’d tend to correctly categorise most results as
negative. However, since p-values are uniformly distributed under the
null, at a significance level of 5%, 5% of all results will be
“significant” even though we would expect to see these results, given
the null hypothesis is true, simply by chance. These would fall under
the label “false positives” in the table above, and are also termed
“false discoveries.”</p>
<p>There are two common ways of controlling these false discoveries. The
first is to say, when we’re doing <span class="math inline">\(n\)</span>
tests, that we want to have the same certainty of making one false
discovery with <span class="math inline">\(n\)</span> tests as we have
if we’re only doing one test. This is “Bonferroni” correction,<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> which
divides the significance level by the number of tests performed, <span class="math inline">\(n\)</span>. Equivalently, we can use the
non-transformed p-value threshold but multiply our p-values by the
number of tests. This is often very conservative, especially with a lot
of features!</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_raw</span> <span class="op">&lt;-</span> <span class="va">toptab_age</span><span class="op">$</span><span class="va">P.Value</span></span>
<span><span class="va">p_fwer</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">p_raw</span>, method <span class="op">=</span> <span class="st">"bonferroni"</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">p_raw</span>, <span class="va">p_fwer</span>, pch <span class="op">=</span> <span class="fl">16</span>, log<span class="op">=</span><span class="st">"xy"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-p-fwer-1.png" alt="Plot of Bonferroni-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." class="figure mx-auto d-block"><figcaption>
Bonferroni correction often produces very large p-values, especially
with low sample sizes.
</figcaption></figure><p>You can see that the p-values are exactly one for the vast majority
of tests we performed! This is not ideal sometimes, because
unfortunately we usually don’t have very large sample sizes in health
sciences.</p>
<p>The second main way of controlling for multiple tests is to control
the <em>false discovery rate (FDR)</em>.<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> This is the proportion
of false positives, or false discoveries, we’d expect to get each time
if we repeated the experiment over and over.</p>
<ol style="list-style-type: decimal"><li>Rank the p-values</li>
<li>Assign each a rank (1 is smallest)</li>
<li>Calculate the critical value <span class="math display">\[
  q = \left(\frac{i}{m}\right)Q
  \]</span>, where <span class="math inline">\(i\)</span> is rank, <span class="math inline">\(m\)</span> is the number of tests, and <span class="math inline">\(Q\)</span> is the false discovery rate we want to
target.<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>
</li>
<li>Find the largest p-value less than the critical value. All smaller
than this are significant.</li>
</ol><table class="table"><colgroup><col width="53%"><col width="46%"></colgroup><thead><tr class="header"><th align="left">FWER</th>
<th align="left">FDR</th>
</tr></thead><tbody><tr class="odd"><td align="left">+ Controls probability of identifying a false
positive</td>
<td align="left">+ Controls rate of false discoveries</td>
</tr><tr class="even"><td align="left">+ Strict error rate control</td>
<td align="left">+ Allows error control with less stringency</td>
</tr><tr class="odd"><td align="left">- Very conservative</td>
<td align="left">- Does not control probability of making errors</td>
</tr><tr class="even"><td align="left">- Requires larger statistical power</td>
<td align="left">- May result in false discoveries</td>
</tr></tbody></table><div id="challenge-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-6" class="callout-inner">
<h3 class="callout-title">Challenge 6</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>At a significance level of 0.05, with 100 tests performed, what is
the Bonferroni significance threshold?</li>
<li>In a gene expression experiment, after FDR correction with an
FDR-adjusted p-value threshold of 0.05, we observe 500 significant
genes. What proportion of these genes are truly different?</li>
<li>Try running FDR correction on the <code>p_raw</code> vector.
<em>Hint: check <code>help("p.adjust")</code> to see what the method is
called</em>.<br>
Compare these values to the raw p-values and the Bonferroni
p-values.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li><p>The Bonferroni threshold for this significance threshold is <span class="math display">\[
  \frac{0.05}{100} = 0.0005
  \]</span></p></li>
<li><p>We can’t say what proportion of these genes are truly different.
However, if we repeated this experiment and statistical test over and
over, on average 5% of the results from each run would be false
discoveries.</p></li>
<li><p>The following code runs FDR correction and compares it to
non-corrected values and to Bonferroni:</p></li>
</ol><div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_fdr</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">p_raw</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">p_raw</span>, <span class="va">p_fdr</span>, pch <span class="op">=</span> <span class="fl">16</span>, log<span class="op">=</span><span class="st">"xy"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-p-fdr-1.png" alt="Plot of Benjamini-Hochberg-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." class="figure mx-auto d-block"><figcaption>
Benjamini-Hochberg correction is less conservative than Bonferroni
</figcaption></figure><div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">p_fwer</span>, <span class="va">p_fdr</span>, pch <span class="op">=</span> <span class="fl">16</span>, log<span class="op">=</span><span class="st">"xy"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0.05</span>, lty <span class="op">=</span> <span class="st">"dashed"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/02-high-dimensional-regression-rendered-plot-fdr-fwer-1.png" alt="Plot of Benjamini-Hochberg-adjusted p-values (y) against Bonferroni-adjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div id="feature-selection" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="feature-selection" class="callout-inner">
<h3 class="callout-title">Feature selection</h3>
<div class="callout-content">
<p>In this episode, we have focussed on regression in a setting where
there are more features than observations. This approach is relevant if
we are interested in the association of each feature with some outcome
or if we want to screen for features that have a strong association with
an outcome. If, however, we are interested in predicting an outcome or
if we want to know which features explain the variation in the outcome,
we may want to restrict ourselves to a subset of relevant features. One
way of doing this is called <em>regularisation</em>, and this is the
topic of the next episode. An alternative is called <em>feature
selection</em>. This is covered in the subsequent (optional)
episode.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a></h2>
<hr class="half-width"><ul><li><a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html" class="external-link"><strong><code>limma</code></strong>
tutorial by Kasper D. Hansen</a></li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf" class="external-link"><strong><code>limma</code></strong>
user manual</a>.</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/vignettes/variancePartition/inst/doc/dream.html" class="external-link">The
<strong><code>VariancePartition</code></strong> package</a> has similar
functionality to <strong><code>limma</code></strong> but allows the
inclusion of random effects.</li>
</ul></section><section><h2 class="section-heading" id="footnotes">Footnotes<a class="anchor" aria-label="anchor" href="#footnotes"></a></h2>
<hr class="half-width"><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Performing linear regression in a high-dimensional setting requires
us to perform hypothesis testing in a way that low-dimensional
regression may not.</li>
<li>Sharing information between features can increase power and reduce
false positives.</li>
<li>When running a lot of null hypothesis tests for high-dimensional
data, multiple testing correction allows retain power and avoid making
costly false discoveries.</li>
<li>Multiple testing methods can be more conservative or more liberal,
depending on our goals.</li>
</ul></div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr><ol><li id="fn1"><p>“True difference” is a hard category to rigidly define.
As we’ve seen, with a lot of data, we can detect tiny differences, and
with little data, we can’t detect large differences. However, both can
be argued to be “true”.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Bonferroni correction is also termed “family-wise” error
rate control.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>This is often called “Benjamini-Hochberg” adjustment.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>People often perform extra controls on FDR-adjusted
p-values, ensuring that ranks don’t change and the critical value is
never smaller than the original p-value.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol></div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="01-introduction-to-high-dimensional-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="03-regression-regularisation.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="01-introduction-to-high-dimensional-data.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="03-regression-regularisation.html" rel="next">
          Next: Regularised...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/02-high-dimensional-regression.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/02-high-dimensional-regression.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Regression with many outcomes",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/02-high-dimensional-regression.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/02-high-dimensional-regression.html",
  "dateCreated": "2025-04-07",
  "dateModified": "2025-04-07",
  "datePublished": "2025-04-08"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

