<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>High dimensional statistics with R: Hierarchical clustering</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/07-hierarchical.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 87%" class="percentage">
    87%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 87%" aria-valuenow="87" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/07-hierarchical.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction-to-high-dimensional-data.html">1. Introduction to high-dimensional data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-high-dimensional-regression.html">2. Regression with many outcomes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-principal-component-analysis.html">4. Principal component analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-k-means.html">6. K-means</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. Hierarchical clustering
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#why-use-hierarchical-clustering-on-high-dimensional-data">Why use hierarchical clustering on high-dimensional data?</a></li>
<li><a href="#the-agglomerative-hierarchical-clustering-algorithm">The agglomerative hierarchical clustering algorithm</a></li>
<li><a href="#a-motivating-example">A motivating example</a></li>
<li><a href="#hierarchical-clustering">Hierarchical clustering</a></li>
<li><a href="#defining-the-dissimilarity-between-observations-creating-the-distance-matrix">Defining the dissimilarity between observations: creating the
distance matrix</a></li>
<li><a href="#defining-the-dissimilarity-between-clusters-linkage-methods">Defining the dissimilarity between clusters: Linkage methods</a></li>
<li><a href="#computing-a-dendrogram">Computing a dendrogram</a></li>
<li><a href="#identifying-the-number-of-clusters">Identifying the number of clusters</a></li>
<li><a href="#dendrogram-visualisation">Dendrogram visualisation</a></li>
<li><a href="#numerical-visualisation">Numerical visualisation</a></li>
<li><a href="#the-effect-of-different-linkage-methods">The effect of different linkage methods</a></li>
<li><a href="#using-different-distance-methods">Using different distance methods</a></li>
<li><a href="#validating-clusters">Validating clusters</a></li>
<li><a href="#further-reading">Further reading</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>

                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="06-k-means.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="06-k-means.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: K-means
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Hierarchical clustering</h1>
        <p>Last updated on 2025-09-02 |

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/07-hierarchical.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is hierarchical clustering and how does it differ from other
clustering methods?</li>
<li>How do we carry out hierarchical clustering in R?</li>
<li>What distance matrix and linkage methods should we use?</li>
<li>How can we validate identified clusters?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand when to use hierarchical clustering on high-dimensional
data.</li>
<li>Perform hierarchical clustering on high-dimensional data and
evaluate dendrograms.</li>
<li>Understand and explore different distance matrix and linkage
methods.</li>
<li>Use the Dunn index to validate clustering methods.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="why-use-hierarchical-clustering-on-high-dimensional-data">Why use hierarchical clustering on high-dimensional data?<a class="anchor" aria-label="anchor" href="#why-use-hierarchical-clustering-on-high-dimensional-data"></a></h2>
<hr class="half-width"><p>When analysing high-dimensional data in the life sciences, it is
often useful to identify groups of similar data points to understand
more about the relationships within the dataset. General clustering
algorithms group similar data points (or observations) into groups (or
clusters). This results in a set of clusters, where each cluster is
distinct, and the data points within each cluster have similar
characteristics. The clustering algorithm works by iteratively grouping
data points so that different clusters may exist at different stages of
the algorithm’s progression.</p>
<p>Here, we describe <em>hierarchical clustering</em>. Unlike K-means
clustering, hierarchical clustering does not require the number of
clusters <span class="math inline">\(k\)</span> to be specified by the
user before the analysis is carried out.</p>
<p>Hierarchical clustering also provides an attractive
<em>dendrogram</em>, a tree-like diagram showing the degree of
similarity between clusters. The dendrogram is a key feature of
hierarchical clustering. This tree-shaped graph allows the similarity
between data points in a dataset to be visualised and the arrangement of
clusters produced by the analysis to be illustrated. Dendrograms are
created using a distance (or dissimilarity) that quantify how different
are pairs of observations, and a clustering algorithm to fuse groups of
similar data points together.</p>
<p>In this episode we will explore hierarchical clustering for
identifying clusters in high-dimensional data. We will use
<em>agglomerative</em> hierarchical clustering (see box) in this
episode.</p>
<div id="agglomerative-and-divisive-hierarchical-clustering" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="agglomerative-and-divisive-hierarchical-clustering" class="callout-inner">
<h3 class="callout-title">Agglomerative and Divisive hierarchical
clustering</h3>
<div class="callout-content">
<p>There are two main methods of carrying out hierarchical clustering:
agglomerative clustering and divisive clustering. The former is a
‘bottom-up’ approach to clustering whereby the clustering approach
begins with each data point (or observation) being regarded as being in
its own separate cluster. Pairs of data points are merged as we move up
the tree. Divisive clustering is a ‘top-down’ approach in which all data
points start in a single cluster and an algorithm is used to split
groups of data points from this main group.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-agglomerative-hierarchical-clustering-algorithm">The agglomerative hierarchical clustering algorithm<a class="anchor" aria-label="anchor" href="#the-agglomerative-hierarchical-clustering-algorithm"></a></h2>
<hr class="half-width"><p>To start with, we measure distance (or dissimilarity) between pairs
of observations. Initially, and at the bottom of the dendrogram, each
observation is considered to be in its own individual cluster. We start
the clustering procedure by fusing the two observations that are most
similar according to a distance matrix. Next, the next-most similar
observations are fused so that the total number of clusters is
<em>number of observations</em> minus 2 (see panel below). Groups of
observations may then be merged into a larger cluster (see next panel
below, green box). This process continues until all the observations are
included in a single cluster.</p>
<figure style="text-align: center"><img src="fig/hierarchical_clustering_1.png" alt="Scatter plot of observations x2 versus x1. Two clusters of pairs of observations are shown by blue and red boxes, each grouping two observations that are close in their x and y distance." width="500px" class="figure mx-auto d-block"><figcaption>
Example data showing two clusters of observation pairs.
</figcaption></figure><figure style="text-align: center"><img src="fig/hierarchical_clustering_2.png" alt="Scatter plot of observations x2 versus x1. Three boxes are shown this time. Blue and red boxes contain two observations each. The two boxes contain points that are relatively far apart. A third green box is shown encompassing the blue box and an additional data point." width="500px" class="figure mx-auto d-block"><figcaption>
Example data showing fusing of one observation into larger cluster.
</figcaption></figure></section><section><h2 class="section-heading" id="a-motivating-example">A motivating example<a class="anchor" aria-label="anchor" href="#a-motivating-example"></a></h2>
<hr class="half-width"><p>To motivate this lesson, let’s first look at an example where
hierarchical clustering is really useful, and then we can understand how
to apply it in more detail. To do this, we’ll return to the large <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link"><code>methylation</code></a>
dataset we worked with in the regression lessons. Let’s load and
visually inspect the data:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">methyl</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/methylation.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># transpose this Bioconductor dataset to show features in columns</span></span>
<span><span class="va">methyl_mat</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">methyl</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in assay(methyl): could not find function "assay"</code></pre>
</div>
<p>Looking at a heatmap of these data, we may spot some patterns – many
columns appear to have a similar methylation levels across all rows.
However, they are all quite jumbled at the moment, so it’s hard to tell
how many line up exactly.</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pheatmap(methyl_mat, legend_title = "Methylation level", cluster_rows = FALSE, : could not find function "pheatmap"</code></pre>
</div>
<p>Looking at a heatmap of these data, we may spot some patterns –
looking at the vertical stripes, many columns appear to have a similar
methylation levels across all rows. However, they are all quite jumbled
at the moment, so it’s hard to tell how many line up exactly. In
addition, it is challenging to tell how many groups containing similar
methylation levels we may have or what the similarities and differences
are between groups.</p>
<p>We can order these data to make the patterns more clear using
hierarchical clustering. To do this, we can change the arguments we pass
to <code>pheatmap()</code> from the
<strong><code>pheatmap</code></strong> package. <code>pheatmap()</code>
groups features based on dissimilarity (here, Euclidean distance) and
orders rows and columns to show clustering of features and
observations.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">methyl_mat</span>,</span>
<span>         legend_title <span class="op">=</span> <span class="st">"Methylation level"</span>,</span>
<span>         cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>         cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>         main <span class="op">=</span> <span class="st">"Methylation Sites vs Individuals"</span>,</span>
<span>         annotation_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pheatmap(methyl_mat, legend_title = "Methylation level", cluster_rows = TRUE, : could not find function "pheatmap"</code></pre>
</div>
<p>We can see that clustering the features (CpG sites) results in an
overall gradient of high to low methylation levels from left to right.
Maybe more interesting is the fact that the rows (corresponding to
individuals) are now grouped according to methylation patterns. For
example, 12 samples seem to have lower methylation levels for a small
subset of CpG sites in the middle, relative to all the other samples.
It’s not clear without investigating further what the cause of this is –
it could be a batch effect, or a known grouping (e.g., old vs young
samples). However, clustering like this can be a useful part of
exploratory analysis of data to build hypotheses.</p>
</section><section><h2 class="section-heading" id="hierarchical-clustering">Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a></h2>
<hr class="half-width"><p>Now, let’s cover the inner workings of hierarchical clustering in
more detail. Hierarchical clustering is a type of clustering that also
allows us to estimate the number of clusters. There are two things to
consider before carrying out clustering:</p>
<ul><li>how to define dissimilarity between observations using a distance
matrix, and</li>
<li>how to define dissimilarity between clusters and when to fuse
separate clusters.</li>
</ul></section><section><h2 class="section-heading" id="defining-the-dissimilarity-between-observations-creating-the-distance-matrix">Defining the dissimilarity between observations: creating the
distance matrix<a class="anchor" aria-label="anchor" href="#defining-the-dissimilarity-between-observations-creating-the-distance-matrix"></a></h2>
<hr class="half-width"><p>Agglomerative hierarchical clustering is performed in two steps:
calculating the distance matrix (containing distances between pairs of
observations) and iteratively grouping observations into clusters using
this matrix.</p>
<p>There are different ways to specify a distance matrix for
clustering:</p>
<ul><li>Specify distance as a pre-defined option using the
<code>method</code> argument in <code>dist()</code>. Methods include
<code>euclidean</code> (default), <code>maximum</code> and
<code>manhattan</code>.</li>
<li>Create a self-defined function which calculates distance from a
matrix or from two vectors. The function should only contain one
argument.</li>
</ul><p>Of pre-defined methods of calculating the distance matrix, Euclidean
is one of the most commonly used. This method calculates the shortest
straight-line distances between pairs of observations.</p>
<p>Another option is to use a correlation matrix as the input matrix to
the clustering algorithm. The type of distance matrix used in
hierarchical clustering can have a big effect on the resulting tree. The
decision of which distance matrix to use before carrying out
hierarchical clustering depends on the type of data and question to be
addressed.</p>
</section><section><h2 class="section-heading" id="defining-the-dissimilarity-between-clusters-linkage-methods">Defining the dissimilarity between clusters: Linkage methods<a class="anchor" aria-label="anchor" href="#defining-the-dissimilarity-between-clusters-linkage-methods"></a></h2>
<hr class="half-width"><p>The second step in performing hierarchical clustering after defining
the distance matrix (or another function defining similarity between
data points) is determining how to fuse different clusters.</p>
<p><em>Linkage</em> is used to define dissimilarity between groups of
observations (or clusters) and is used to create the hierarchical
structure in the dendrogram. Different linkage methods of creating a
dendrogram are discussed below.</p>
<p><code>hclust()</code> supports various linkage methods (e.g
<code>complete</code>, <code>single</code>, <code>ward D</code>,
<code>ward D2</code>, <code>average</code>, <code>median</code>) and
these are also supported within the <code>pheatmap()</code> function.
The method used to perform hierarchical clustering in
<code>pheatmap()</code> can be specified by the argument
<code>clustering_method</code>. Each linkage method uses a slightly
different algorithm to calculate how clusters are fused together and
therefore different clustering decisions are made depending on the
linkage method used.</p>
<p>Complete linkage (the default in <code>hclust()</code>) works by
computing all pairwise dissimilarities between data points in different
clusters. For each pair of two clusters, it sets their dissimilarity to
the maximum dissimilarity value observed between any of these clusters’
constituent points. The two clusters with smallest dissimilarity value
are then fused.</p>
</section><section><h2 class="section-heading" id="computing-a-dendrogram">Computing a dendrogram<a class="anchor" aria-label="anchor" href="#computing-a-dendrogram"></a></h2>
<hr class="half-width"><p>Dendrograms are useful tools that plot the grouping of points and
clusters into bigger clusters. We can create and plot dendrograms in R
using <code>hclust()</code> which takes a distance matrix as input and
creates a tree using hierarchical clustering. Here we create some
example data to carry out hierarchical clustering.</p>
<p>Let’s generate 20 data points in 2D space. Each point is generated to
belong to one of three classes/groups. Suppose we did not know which
class data points belonged to and we want to identify these via cluster
analysis. Let’s first generate and plot our data:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#First, create some example data with two variables x1 and x2</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">450</span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    x1 <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">20</span>, <span class="fl">8</span>, <span class="fl">4.5</span><span class="op">)</span>,</span>
<span>    x2 <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">20</span>, <span class="fl">6</span>, <span class="fl">3.4</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plot the data and name data points by row numbers</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">example_data</span><span class="op">$</span><span class="va">x1</span>, <span class="va">example_data</span><span class="op">$</span><span class="va">x2</span>, type <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span><span class="fu">text</span><span class="op">(</span></span>
<span>    <span class="va">example_data</span><span class="op">$</span><span class="va">x1</span>,</span>
<span>    <span class="va">example_data</span><span class="op">$</span><span class="va">x2</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">example_data</span><span class="op">)</span>,</span>
<span>    cex <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plotexample-1.png" alt="A scatter plot of randomly-generated data x2 versus x1. The points appear fairly randomly scattered, arguably centered towards the bottom of the plot." class="figure mx-auto d-block"><figcaption>
Scatter plot of randomly-generated data x2 versus x1.
</figcaption></figure><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate distance matrix using euclidean distance</span></span>
<span><span class="va">dist_m</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">example_data</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre>
</div>
<p>Hierarchical clustering carried out on the data can be used to
produce a dendrogram showing how the data is partitioned into clusters.
But how do we interpret this dendrogram? Let’s explore this using our
example data in a Challenge.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1</h3>
<div class="callout-content">
<p>Use <code>hclust()</code> to implement hierarchical clustering using
the distance matrix <code>dist_m</code> and the <code>complete</code>
linkage method and plot the results as a dendrogram using
<code>plot()</code>. Why is hierarchical clustering and the resulting
dendrogram useful for performing clustering this case?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solution:
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">dist_m</span>, method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plotclustex-1.png" alt="A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other)." class="figure mx-auto d-block"><figcaption>
A dendrogram of the randomly-generated data.
</figcaption></figure><p>Hierarchical clustering is particularly useful (compared to K-means)
when we do not know the number of clusters before we perform clustering.
It is useful in this case since we have assumed we do not already know
what a suitable number of clusters may be.</p>
</div>
</div>
</div>
</div>
<p>A dendrogram, such as the one generated in Challenge 1, shows
similarities/differences in distances between data points. Each vertical
line at the bottom of the dendrogram (‘leaf’) represents one of the 20
data points. These leaves fuse into fewer vertical lines (‘branches’) as
the height increases. Observations that are similar fuse into the same
branches. The height at which any two data points fuse indicates how
different these two points are. Points that fuse at the top of the tree
are very different from each other compared with two points that fuse at
the bottom of the tree, which are quite similar. You can see this by
comparing the position of similar/dissimilar points according to the
scatterplot with their position on the tree.</p>
</section><section><h2 class="section-heading" id="identifying-the-number-of-clusters">Identifying the number of clusters<a class="anchor" aria-label="anchor" href="#identifying-the-number-of-clusters"></a></h2>
<hr class="half-width"><p>To identify the number of clusters, we can make a horizontal cut
through the dendrogram at a user-defined height. The sets of
observations beneath this cut and single points representing clusters
above can be thought of as distinct clusters. Equivalently, we can count
the vertical lines we encounter crossing the horizontal cut and the
number of single points above the cut. For example, a cut at height 10
produces 3 clusters for the dendrogram in Challenge 1, while a cut at
height 4 produces 8 clusters.</p>
</section><section><h2 class="section-heading" id="dendrogram-visualisation">Dendrogram visualisation<a class="anchor" aria-label="anchor" href="#dendrogram-visualisation"></a></h2>
<hr class="half-width"><p>We can first visualise cluster membership by highlight branches in
dendrograms. In this example, we calculate a distance matrix between
samples in the <code>methyl_mat</code> dataset. We then draw boxes round
clusters obtained with <code>cutree()</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create a distance matrix using euclidean distance</span></span>
<span><span class="va">distmat</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## hierarchical clustering using complete method</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">distmat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot resulting dendrogram</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## draw border around three clusters</span></span>
<span><span class="fu">rect.hclust</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">3</span>, border <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="co">#border argument specifies the colours</span></span>
<span><span class="co">## draw border around two clusters</span></span>
<span><span class="fu">rect.hclust</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">2</span>, border <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-method-1.png" alt="A dendrogram for the methyl_mat data with boxes overlaid on clusters. There are 5 boxes in total, each indicating separate clusters." class="figure mx-auto d-block"><figcaption>
Dendrogram with boxes around clusters.
</figcaption></figure><p>We can also colour clusters downstream of a specified cut using
<code>color_branches()</code> from the
<strong><code>dendextend</code></strong> package.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## cut tree at height = 4</span></span>
<span><span class="va">cut</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"dendextend"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
---------------------
Welcome to dendextend version 1.19.1
Type citation('dendextend') for how to cite the package.

Type browseVignettes(package = 'dendextend') for the package vignette.
The github page is: https://github.com/talgalili/dendextend/

Suggestions and bug-reports can be submitted at: https://github.com/talgalili/dendextend/issues
You may ask questions at stackoverflow, use the r and dendextend tags:
	 https://stackoverflow.com/questions/tagged/dendextend

	To suppress this message use:  suppressPackageStartupMessages(library(dendextend))
---------------------</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dendextend'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:stats':

    cutree</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">avg_dend_obj</span> <span class="op">&lt;-</span> <span class="fu">as.dendrogram</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>      </span>
<span><span class="co">## colour branches of dendrogram depending on clusters</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">color_branches</span><span class="op">(</span><span class="va">avg_dend_obj</span>, h <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required namespace: colorspace</code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-coloured-branches-1.png" alt="A dendrogram with the different clusters in 4 different colours." class="figure mx-auto d-block"><figcaption>
Dendrogram with coloured branches delineating different clusters.
</figcaption></figure></section><section><h2 class="section-heading" id="numerical-visualisation">Numerical visualisation<a class="anchor" aria-label="anchor" href="#numerical-visualisation"></a></h2>
<hr class="half-width"><p>In addition to visualising clusters directly on the dendrogram, we
can cut the dendrogram to determine number of clusters at different
heights using <code>cutree()</code>. This function cuts a dendrogram
into several groups (or clusters) where the number of desired groups is
controlled by the user, by defining either <code>k</code> (number of
groups) or <code>h</code> (height at which tree is cut). The function
outputs the cluster labels of each data point in order.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## k is a user defined parameter determining</span></span>
<span><span class="co">## the desired number of clusters at which to cut the treee</span></span>
<span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 2 2 3 4 5 6 5 1 7 8 6 5 9 4 4 7 8 5 8</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## h is a user defined parameter determining</span></span>
<span><span class="co">## the numeric height at which to cut the tree</span></span>
<span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">36</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## both give same results </span></span>
<span></span>
<span><span class="va">four_cut</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## we can produce the cluster each observation belongs to</span></span>
<span><span class="co">## using the mutate and count functions</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_cl</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span>, cluster <span class="op">=</span> <span class="va">four_cut</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">count</span><span class="op">(</span><span class="va">example_cl</span>, <span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'example_cl' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#plot cluster each point belongs to on original scatterplot</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">example_cl</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cg01644850</span>, y <span class="op">=</span> <span class="va">cg01656216</span>, color <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'example_cl' not found</code></pre>
</div>
<p>Note that this cut produces 1 clusters (zero before the cut and
another four downstream of the cut).</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2:</h3>
<div class="callout-content">
<p>Identify the value of <code>k</code> in <code>cutree()</code> that
gives the same output as <code>h = 36</code></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Solution:
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span>
<span><span class="co">## create horizontal line at height = 45</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h <span class="op">=</span> <span class="fl">45</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-h-k-ex-plot-1.png" alt="A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other). A dashed horizontal line at a height of 5 indicates the cut point used to divide the data into clusters." class="figure mx-auto d-block"><figcaption>
A dendrogram of the randomly-generated data.
</figcaption></figure><div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 2 2 3 2 1 2 1 1 4 4 2 1 5 2 2 4 4 1 4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">five_cut</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">example_cl</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span>, cluster <span class="op">=</span> <span class="va">five_cut</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">count</span><span class="op">(</span><span class="va">example_cl</span>, <span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'example_cl' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">example_cl</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">cg01644850</span>, y <span class="op">=</span> <span class="va">cg01656216</span>, color <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'example_cl' not found</code></pre>
</div>
<p>Five clusters (<code>k = 5</code>) gives similar results to
<code>h = 45</code>. You can plot a horizontal line on the dendrogram at
<code>h = 45</code> to help identify corresponding value of
<code>k</code>.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-effect-of-different-linkage-methods">The effect of different linkage methods<a class="anchor" aria-label="anchor" href="#the-effect-of-different-linkage-methods"></a></h2>
<hr class="half-width"><p>Now let us look into changing the default behaviour of
<code>hclust()</code>. First, we’ll load a synthetic dataset, comprised
of two variables representing two crescent-shaped point clouds:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cres</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/cres.rds"</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-crescents-1.png" alt="A scatter plot of data simulated to form two crescent shapes. The crescents are horizontally orientated with a a rough line of vertical symmetry." class="figure mx-auto d-block"><figcaption>
Scatter plot of data simulated according to two crescent-shaped point
clouds.
</figcaption></figure><p>We might expect that the crescents are resolved into separate
clusters. But if we run hierarchical clustering with the default
arguments, we get this:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClass</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="co"># save partition for colouring</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClass</span><span class="op">)</span> <span class="co"># colour scatterplot by partition</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-cresClustDefault-1.png" alt="A scatter plot of the crescent-shaped simulated data calculated using Euclidean distance. The points are coloured in black or red according to their membership to 2 clusters. The points in the tails of each crescent have inherited the colour of the opposite crescent." class="figure mx-auto d-block"><figcaption>
Scatter plot of crescent-shaped simulated data, coloured according to
clusters calculated using Euclidean distance.
</figcaption></figure><div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3</h3>
<div class="callout-content">
<p>Carry out hierarchical clustering on the <code>cres</code> data that
we generated above. Try out different linkage methods and use
<code>cutree()</code> to split each resulting dendrogram into two
clusters. Plot the results colouring the dots according to their
inferred cluster identity.</p>
<p>Which method(s) give you the expected clustering outcome?</p>
<p>Hint: Check <code>?hclust</code> to see the possible values of the
argument <code>method</code> (the linkage method used).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Solution:
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#?hclust</span></span>
<span><span class="co"># "complete", "single", "ward.D", "ward.D2", "average", "mcquitty", "median" or "centroid"</span></span>
<span><span class="va">cresClassSingle</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>,method <span class="op">=</span> <span class="st">"single"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassSingle</span>, main<span class="op">=</span><span class="st">"single"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-comp-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with single linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassWard.D</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"ward.D"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassWard.D</span>, main<span class="op">=</span><span class="st">"ward.D"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-wardD-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassWard.D2</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"ward.D2"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassWard.D2</span>, main<span class="op">=</span><span class="st">"ward.D2"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-wardD2-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D2 linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassAverage</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"average"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassAverage</span>, main<span class="op">=</span><span class="st">"average"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-average-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with average linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassMcquitty</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"mcquitty"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassMcquitty</span>, main<span class="op">=</span><span class="st">"mcquitty"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-mcq-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with mcquitty linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassMedian</span><span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"median"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span> </span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassMedian</span>, main<span class="op">=</span><span class="st">"median"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-median-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with median linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cresClassCentroid</span><span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">cres</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"centroid"</span><span class="op">)</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">cres</span>, col<span class="op">=</span><span class="va">cresClassCentroid</span>, main<span class="op">=</span><span class="st">"centroid"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-centroid-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with centroid linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><p>The linkage methods <code>single</code>, <code>ward.D</code>, and
<code>average</code> resolve each crescent as a separate cluster.</p>
</div>
</div>
</div>
</div>
<p>The help page of <code>hclust()</code> gives some intuition on
linkage methods. It describes <code>complete</code> (the default) and
<code>single</code> as opposite ends of a spectrum with all other
methods in between. When using complete linkage, the distance between
two clusters is assumed to be the distance between both clusters’ most
distant points. This opposite it true for single linkage, where the
minimum distance between any two points, one from each of two clusters
is used. Single linkage is described as friends-of-friends appporach -
and really, it groups all close-together points into the same cluster
(thus resolving one cluster per crescent). Complete linkage on the other
hand recognises that some points a the tip of a crescent are much closer
to points in the other crescent and so it splits both crescents.</p>
</section><section><h2 class="section-heading" id="using-different-distance-methods">Using different distance methods<a class="anchor" aria-label="anchor" href="#using-different-distance-methods"></a></h2>
<hr class="half-width"><p>So far, we’ve been using Euclidean distance to define the
dissimilarity or distance between observations. However, this isn’t
always the best metric for how dissimilar the observations are. Let’s
make an example to demonstrate. Here, we’re creating two samples each
with ten observations of random noise:</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">cor_example</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>  sample_a <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  sample_b <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span></span>
<span>  <span class="st">"Feature"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Now, let’s create a new sample that has exactly the same pattern
across all our features as <code>sample_a</code>, just offset by 5:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_example</span><span class="op">$</span><span class="va">sample_c</span> <span class="op">&lt;-</span> <span class="va">cor_example</span><span class="op">$</span><span class="va">sample_a</span> <span class="op">+</span> <span class="fl">5</span></span></code></pre>
</div>
<p>You can see that this is a lot like the <code>assay()</code> of our
methylation object from earlier, where columns are observations or
samples, and rows are features:</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            sample_a    sample_b sample_c
Feature 1  1.1626853 -0.02013537 6.162685
Feature 2 -0.5859245 -0.15038222 4.414076
Feature 3  1.7854650 -0.62812676 6.785465
Feature 4 -1.3325937  1.32322085 3.667406
Feature 5 -0.4465668 -1.52135057 4.553433
Feature 6  0.5696061 -0.43742787 5.569606</code></pre>
</div>
<p>If we plot a heatmap of this, we can see that <code>sample_a</code>
and <code>sample_b</code> are grouped together because they have a small
distance from each other, despite being quite different in their pattern
across the different features. In contrast, <code>sample_a</code> and
<code>sample_c</code> are very distant, despite having <em>exactly</em>
the same pattern across the different features.</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pheatmap(as.matrix(cor_example)): could not find function "pheatmap"</code></pre>
</div>
<p>We can see that more clearly if we do a line plot:</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create a blank plot (type = "n" means don't draw anything)</span></span>
<span><span class="co">## with an x range to hold the number of features we have.</span></span>
<span><span class="co">## the range of y needs to be enough to show all the values for every feature</span></span>
<span><span class="fu">plot</span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span>,</span>
<span>  <span class="fu">rep</span><span class="op">(</span><span class="fu">range</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"n"</span>, </span>
<span>  xlab <span class="op">=</span> <span class="st">"Feature number"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Value"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## draw a red line for sample_a</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">cor_example</span><span class="op">$</span><span class="va">sample_a</span>, col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span></span>
<span><span class="co">## draw a blue line for sample_b</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">cor_example</span><span class="op">$</span><span class="va">sample_b</span>, col <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span></span>
<span><span class="co">## draw a green line for sample_c</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">cor_example</span><span class="op">$</span><span class="va">sample_c</span>, col <span class="op">=</span> <span class="st">"forestgreen"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-lineplot-cor-example-1.png" alt="A line plot of simulated value versus observation number, coloured by sample. Samples a and b are concentrated at the bottom of the plot, while sample c is concentrated at the top of the plot. Samples a and c have exactly the same vertical pattern." class="figure mx-auto d-block"><figcaption>
Line plot of simulated value versus observation number, coloured by
sample.
</figcaption></figure><p>We can see that <code>sample_a</code> and <code>sample_c</code> have
exactly the same pattern across all of the different features. However,
due to the overall difference between the values, they have a high
distance to each other. We can see that if we cluster and plot the data
ourselves using Euclidean distance:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clust_dist</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">clust_dist</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-clust-euc-cor-example-1.png" alt="A dendrogram of the example simulated data clustered according to Euclidean distance. The dendrogram shows that sample c definitively forms its own cluster for any cut height and samples a and b merge into a cluster at a height of around 6." class="figure mx-auto d-block"><figcaption>
Dendrogram of the example simulated data clustered according to
Euclidean distance.
</figcaption></figure><p>In some cases, we might want to ensure that samples that have similar
patterns, whether that be of gene expression, or DNA methylation, have
small distances to each other. Correlation is a measure of this kind of
similarity in pattern. However, high correlations indicate similarity,
while for a distance measure we know that high distances indicate
dissimilarity. Therefore, if we wanted to cluster observations based on
the correlation, or the similarity of patterns, we can use
<code>1 - cor(x)</code> as the distance metric. The input to
<code>hclust()</code> must be a <code>dist</code> object, so we also
need to call <code>as.dist()</code> on it before passing it in.</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_as_dist</span> <span class="op">&lt;-</span> <span class="fu">as.dist</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu">cor</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clust_cor</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">cor_as_dist</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">clust_cor</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-clust-cor-cor-example-1.png" alt="A dendrogram of the example simulated data clustered according to correlation. The dendrogram shows that sample b definitively forms its own cluster and samples a and c form definitively form their own cluster for any cut height." class="figure mx-auto d-block"><figcaption>
Dendrogram of the example simulated data clustered according to
correlation.
</figcaption></figure><p>Now, <code>sample_a</code> and <code>sample_c</code> that have
identical patterns across the features are grouped together, while
<code>sample_b</code> is seen as distant because it has a different
pattern, even though its values are closer to <code>sample_a</code>.
Using your own distance function is often useful, especially if you have
missing or unusual data. It’s often possible to use correlation and
other custom dissimilarity measures in functions that perform
hierarchical clustering, such as <code>pheatmap()</code> and
<code>stats::heatmap()</code>:</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## pheatmap allows you to select correlation directly</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span>, clustering_distance_cols <span class="op">=</span> <span class="st">"correlation"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pheatmap(as.matrix(cor_example), clustering_distance_cols = "correlation"): could not find function "pheatmap"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Using the built-in stats::heatmap </span></span>
<span><span class="fu">heatmap</span><span class="op">(</span></span>
<span>  <span class="fu">as.matrix</span><span class="op">(</span><span class="va">cor_example</span><span class="op">)</span>,</span>
<span>  distfun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">as.dist</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu">cor</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-heatmap-cor-cor-example-1.png" alt="Heatmaps of features versus samples, coloured by simulated value. The columns (samples) are clustered according to the correlation. Samples a and b have mostly low values, delineated by blue in the first plot and yellow in the second plot. Sample c has mostly high values, delineated by red in the first plot and brown in the second plot. Samples A and C form a cluster separate from sample B in the column dendrogram." class="figure mx-auto d-block"><figcaption>
Heatmaps of features versus samples clustered in the samples according
to correlation.
</figcaption></figure></section><section><h2 class="section-heading" id="validating-clusters">Validating clusters<a class="anchor" aria-label="anchor" href="#validating-clusters"></a></h2>
<hr class="half-width"><p>Now that we know how to carry out hierarchical clustering, how do we
know how many clusters are optimal for the dataset?</p>
<p>Hierarchical clustering carried out on any dataset will produce
clusters, even when there are no ‘real’ clusters in the data! We need to
be able to determine whether identified clusters represent true groups
in the data, or whether clusters have been identified just due to
chance. In the last episode, we have introduced silhouette scores as a
measure of cluster compactness and bootstrapping to assess cluster
robustness. Such tests can be used to compare different clustering
algorithms, for example, those fitted using different linkage
methods.</p>
<p>Here, we introduce the Dunn index, which is a measure of cluster
compactness. The Dunn index is the ratio of the smallest distance
between any two clusters and to the largest intra-cluster distance found
within any cluster. This can be seen as a family of indices which differ
depending on the method used to compute distances. The Dunn index is a
metric that penalises clusters that have larger intra-cluster variance
and smaller inter-cluster variance. The higher the Dunn index, the
better defined the clusters.</p>
<p>Let’s calculate the Dunn index for clustering carried out on the
<code>methyl_mat</code> dataset using the
<strong><code>clValid</code></strong> package.</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate dunn index</span></span>
<span><span class="co">## (ratio of the smallest distance between obs not in the same cluster</span></span>
<span><span class="co">## to the largest intra-cluster distance)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"clValid"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: cluster</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate euclidean distance between points </span></span>
<span><span class="va">distmat</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span>  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">distmat</span>, method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-dunn-1.png" alt="A dendrogram for clustering of methylation data. Identical to that in the section Highlighting dendrogram branches, without the colour overlay to show clusters." class="figure mx-auto d-block"><figcaption>
Dendrogram for clustering of methylation data.
</figcaption></figure><div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cut</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## retrieve Dunn's index for given matrix and clusters</span></span>
<span><span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="va">cut</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<p>The value of the Dunn index has no meaning in itself, but is used to
compare between sets of clusters with larger values being preferred.</p>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4</h3>
<div class="callout-content">
<p>Examine how changing the <code>h</code> or <code>k</code> arguments
in <code>cutree()</code> affects the value of the Dunn index.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Solution:
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"clValid"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">distmat</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">methyl_mat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'methyl_mat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">distmat</span>, method <span class="op">=</span> <span class="st">"complete"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Varying h</span></span>
<span><span class="co">## Obtaining the clusters</span></span>
<span><span class="va">cut_h_20</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">cut_h_30</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## How many clusters?</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">cut_h_20</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">cut_h_30</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="va">cut_h_20</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="va">cut_h_30</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb87">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Varying k</span></span>
<span><span class="co">## Obtaining the clusters</span></span>
<span><span class="va">cut_k_10</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">cut_k_5</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## How many clusters?</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">cut_k_5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb89">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">cut_k_10</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb91">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="va">cut_k_5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb93">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="va">cut_k_10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'distmat' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The figures below show in a more systematic way how changing the
values of <code>k</code> and <code>h</code> using <code>cutree()</code>
affect the Dunn index.</p>
<div class="codewrapper sourceCode" id="cb95">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h_seq</span> <span class="op">&lt;-</span> <span class="fl">70</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="va">h_dunn</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">h_seq</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, h <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in FUN(X[[i]], ...): object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb97">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k_seq</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">k_dunn</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">k_seq</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">dunn</span><span class="op">(</span>distance <span class="op">=</span> <span class="va">distmat</span>, <span class="fu">cutree</span><span class="op">(</span><span class="va">clust</span>, k <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in FUN(X[[i]], ...): object 'distmat' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb99">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">h_seq</span>, <span class="va">h_dunn</span>, xlab <span class="op">=</span> <span class="st">"Height (h)"</span>, ylab <span class="op">=</span> <span class="st">"Dunn index"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'y' in selecting a method for function 'plot': object 'h_dunn' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">grid</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in int_abline(a = a, b = b, h = h, v = v, untf = untf, ...): plot.new has not been called yet</code></pre>
</div>
<p>You can see that at low values of <code>h</code>, the Dunn index can
be high. But this is not very useful - cutting the given tree at a low
<code>h</code> value like 15 leads to allmost all observations ending up
each in its own cluster. More relevant is the second maximum in the
plot, around <code>h=55</code>. Looking at the dendrogram, this
corresponds to <code>k=4</code>.</p>
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">k_seq</span>, <span class="va">k_dunn</span>, xlab <span class="op">=</span> <span class="st">"Number of clusters (k)"</span>, ylab <span class="op">=</span> <span class="st">"Dunn index"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'y' in selecting a method for function 'plot': object 'k_dunn' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">grid</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in int_abline(a = a, b = b, h = h, v = v, untf = untf, ...): plot.new has not been called yet</code></pre>
</div>
<p>For the given range of <code>k</code> values explored, we obtain the
highest Dunn index with <code>k=4</code>. This is in agreement with the
previous plot.</p>
<p>There have been criticisms of the use of the Dunn index in validating
clustering results, due to its high sensitivity to noise in the dataset.
An alternative is to use silhouette scores (see the k-means clustering
episode).</p>
<p>As we said before (see previous episode), clustering is a non-trivial
task. It is important to think about the nature of your data and your
expectations rather than blindly using a some algorithm for clustering
or cluster validation.</p>
</section><section><h2 class="section-heading" id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a></h2>
<hr class="half-width"><ul><li>Dunn, J. C. (1974) Well-separated clusters and optimal fuzzy
partitions. Journal of Cybernetics 4(1):95–104.</li>
<li>Halkidi, M., Batistakis, Y. &amp; Vazirgiannis, M. (2001) On
clustering validation techniques. Journal of Intelligent Information
Systems 17(2/3):107-145.</li>
<li>James, G., Witten, D., Hastie, T. &amp; Tibshirani, R. (2013) An
Introduction to Statistical Learning with Applications in R. Section
10.3.2 (Hierarchical Clustering).</li>
<li>
<a href="https://towardsdatascience.com/understanding-the-concept-of-hierarchical-clustering-technique-c6e8243758ec" class="external-link">Understanding
the concept of Hierarchical clustering Technique. towards data science
blog</a>.</li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Hierarchical clustering uses an algorithm to group similar data
points into clusters. A dendrogram is used to plot relationships between
clusters (using the <code>hclust()</code> function in R).</li>
<li>Hierarchical clustering differs from k-means clustering as it does
not require the user to specify expected number of clusters</li>
<li>The distance (dissimilarity) matrix can be calculated in various
ways, and different clustering algorithms (linkage methods) can affect
the resulting dendrogram.</li>
<li>The Dunn index can be used to validate clusters using the original
dataset.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="06-k-means.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="06-k-means.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: K-means
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/07-hierarchical.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/07-hierarchical.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Hierarchical clustering",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/07-hierarchical.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/07-hierarchical.html",
  "dateCreated": "2025-10-10",
  "dateModified": "2025-09-02",
  "datePublished": "2025-11-04"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

