<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>High dimensional statistics with R: All Images</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/images.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/images.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction-to-high-dimensional-data.html">1. Introduction to high-dimensional data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-high-dimensional-regression.html">2. Regression with many outcomes</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-principal-component-analysis.html">4. Principal component analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-k-means.html">6. K-means</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-hierarchical.html">7. Hierarchical clustering</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>

                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="01-introduction-to-high-dimensional-data"><h2 class="section-heading"><a href="01-introduction-to-high-dimensional-data.html">Introduction to high-dimensional data</a></h2>
<hr class="half-width">
<h3 id="01-introduction-to-high-dimensional-data-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Table displaying a high-dimensional data set with many columns representing features related to health, such as blood pressure, heart rate and respiratory rate. Each row contains the data for an individual patient. This type of high-dimensional data could contain hundreds or thousands of columns (features/variables) and thousands or even millions of rows (observations/samples/patients).’</p>
<figure style="text-align: center"><img src="fig/intro-table.png" alt="Table displaying a high-dimensional data set with many columns representing features related to health, such as blood pressure, heart rate and respiratory rate. Each row contains the data for an individual patient. This type of high-dimensional data could contain hundreds or thousands of columns (features/variables) and thousands or even millions of rows (observations/samples/patients)." width="5572" class="figure mx-auto d-block"><figcaption>
Example of a high-dimensional data table with features in the columns
and individual observations (patients) in rows.
</figcaption></figure><hr>
<h3 id="01-introduction-to-high-dimensional-data-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘A set of pairwise scatterplots of variables in the 'prostate' dataset, namely lcavol, lweight, age, lbph, svi, lcp, gleason, pgg45, lpsa. The plots are shown in a grid.’</p>
<figure style="text-align: center"><img src="fig/01-introduction-to-high-dimensional-data-rendered-pairs-prostate-1.png" alt="A set of pairwise scatterplots of variables in the 'prostate' dataset, namely lcavol, lweight, age, lbph, svi, lcp, gleason, pgg45, lpsa. The plots are shown in a grid." class="figure mx-auto d-block"><figcaption>
Pairwise plots of the ‘prostate’ dataset.
</figcaption></figure><hr>
<h3 id="01-introduction-to-high-dimensional-data-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two scatter plots side-by-side, each plotting the relationship between two variables. The scatter plot on the left hand side shows 25 observations and a regression line with the points evenly scattered around. The scatter plot on the right hand side shows 2 observations and a regression line that goes through both points.’</p>
<figure style="text-align: center"><img src="fig/intro-scatterplot.png" alt="Two scatter plots side-by-side, each plotting the relationship between two variables. The scatter plot on the left hand side shows 25 observations and a regression line with the points evenly scattered around. The scatter plot on the right hand side shows 2 observations and a regression line that goes through both points." width="5307" class="figure mx-auto d-block"><figcaption>
Scatter plot of two variables (x and y) from a data set with 25
observations (left) and 2 observations (right) with a fitted regression
line (red).
</figcaption></figure><hr></section><section id="02-high-dimensional-regression"><h2 class="section-heading"><a href="02-high-dimensional-regression.html">Regression with many outcomes</a></h2>
<hr class="half-width">
<h3 id="02-high-dimensional-regression-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Histogram of M-values for all features. The distribution appears to be bimodal, with a large number of unmethylated features as well as many methylated features, and many intermediate features.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-histx-1.png" alt="Histogram of M-values for all features. The distribution appears to be bimodal, with a large number of unmethylated features as well as many methylated features, and many intermediate features." class="figure mx-auto d-block"><figcaption>
Methylation levels are generally bimodally distributed.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘An example of a strong linear association between a continuous phenotype (age) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. A strong linear relationship with a positive slope exists between the two.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example1-1.png" alt="An example of a strong linear association between a continuous phenotype (age) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. A strong linear relationship with a positive slope exists between the two." class="figure mx-auto d-block"><figcaption>
A scatter plot of age and a feature of interest.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups clearly differ with respect to DNA methylation.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example2-1.png" alt="An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups clearly differ with respect to DNA methylation." class="figure mx-auto d-block"><figcaption>
A scatter plot of a grouping and a feature of interest.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups seem to differ with respect to DNA methylation, but the relationship is weak.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-example3-1.png" alt="An example of a strong linear association between a discrete phenotype (group) on the x-axis and a feature of interest (DNA methylation at a given locus) on the y-axis. The two groups seem to differ with respect to DNA methylation, but the relationship is weak." class="figure mx-auto d-block"><figcaption>
A scatter plot of a grouping and a feature of interest.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘An example of the relationship between age (x-axis) and methylation levels (y-axis) for an arbitrarily selected CpG. In this case, the y-axis shows methylation levels for the first CpG in our data. The black line shows the fitted regression line (based on the intercept and slope estimates shown above). For this feature, we can see that there is no strong relationship between methylation and age.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-lm-methyl1-1.png" alt="An example of the relationship between age (x-axis) and methylation levels (y-axis) for an arbitrarily selected CpG. In this case, the y-axis shows methylation levels for the first CpG in our data. The black line shows the fitted regression line (based on the intercept and slope estimates shown above). For this feature, we can see that there is no strong relationship between methylation and age." class="figure mx-auto d-block"><figcaption>
A scatter plot of age versus the methylation level for an arbitrarily
selected CpG side (the one stored as the first column of methyl_mat).
Each dot represents an individual. The black line represents the
estimated linear model.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Density plot of a t-distribution showing the observed test statistics (here, t-statistics). The p-values, visualised here with shaded regions, represent the portion of the null distribution that is as extreme or more extreme as the observed test statistics, which are shown as dashed lines.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-tdist-1.png" alt="Density plot of a t-distribution showing the observed test statistics (here, t-statistics). The p-values, visualised here with shaded regions, represent the portion of the null distribution that is as extreme or more extreme as the observed test statistics, which are shown as dashed lines." class="figure mx-auto d-block"><figcaption>
The p-value for a regression coefficient represents how often it’d be
observed under the null.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A plot of -log10(p) against effect size estimates for a regression of age against methylation using limma.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-limmavolc1-1.png" alt="A plot of -log10(p) against effect size estimates for a regression of age against methylation using limma." class="figure mx-auto d-block"><figcaption>
Plotting p-values against effect sizes using limma; the results are
similar to a standard linear model.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of the effect size using limmma vs. those using lm. The plot also shows a straight line through all points showing that the effect sizes are the same.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-limma-lm-effect-1.png" alt="A scatter plot of the effect size using limmma vs. those using lm. The plot also shows a straight line through all points showing that the effect sizes are the same." class="figure mx-auto d-block"><figcaption>
Plot of effect sizes using limma vs. those using lm.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of the p-values using limma vs. those using lm. A straight line is also displayed, showing that the p-values for limma tend to be smaller than those using lm towards the left of the plot and higher towards the right of the plot.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-plot-limma-lm-pval-1.png" alt="A scatter plot of the p-values using limma vs. those using lm. A straight line is also displayed, showing that the p-values for limma tend to be smaller than those using lm towards the left of the plot and higher towards the right of the plot." class="figure mx-auto d-block"><figcaption>
Plot of p-values using limma vs. those using lm.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘A plot of -log10(p) against effect size estimates for a regression of smoking status against methylation using limma.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-limmavolc2-1.png" alt="A plot of -log10(p) against effect size estimates for a regression of smoking status against methylation using limma." class="figure mx-auto d-block"><figcaption>
A plot of significance against effect size for a regression of smoking
against methylation.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-11">Figure 11</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot of -log10(p) against effect size estimates for a regression of a made-up feature against methylation level for each feature in the data. A dashed line represents a 0.05 significance level.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-volcplotfake-1.png" alt="Plot of -log10(p) against effect size estimates for a regression of a made-up feature against methylation level for each feature in the data. A dashed line represents a 0.05 significance level." class="figure mx-auto d-block"><figcaption>
Plotting p-values against effect sizes for a randomised outcome shows we
still observe ‘significant’ results.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-12">Figure 12</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot of Bonferroni-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-p-fwer-1.png" alt="Plot of Bonferroni-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." class="figure mx-auto d-block"><figcaption>
Bonferroni correction often produces very large p-values, especially
with low sample sizes.
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-13">Figure 13</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot of Benjamini-Hochberg-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds.’</p>
<figure style="text-align: center"><img src="fig/02-high-dimensional-regression-rendered-p-fdr-1.png" alt="Plot of Benjamini-Hochberg-adjusted p-values (y) against unadjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." class="figure mx-auto d-block"><figcaption>
Benjamini-Hochberg correction is less conservative than Bonferroni
</figcaption></figure><hr>
<h3 id="02-high-dimensional-regression-figure-14">Figure 14</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot of Benjamini-Hochberg-adjusted p-values (y) against Bonferroni-adjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds.’</p>
<figure><img src="fig/02-high-dimensional-regression-rendered-plot-fdr-fwer-1.png" alt="Plot of Benjamini-Hochberg-adjusted p-values (y) against Bonferroni-adjusted p-values (x). A dashed black line represents the identity (where x=y), while dashed red lines represent 0.05 significance thresholds." style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section id="03-regression-regularisation"><h2 class="section-heading"><a href="03-regression-regularisation.html">Regularised regression</a></h2>
<hr class="half-width">
<h3 id="03-regression-regularisation-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘A symmetrical heatmap where rows and columns are features in a DNA methylation dataset. Colour corresponds to correlation, with red being large positive correlations and blue being large negative correlations. There are large blocks of deep red and blue throughout the plot.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-corr-mat-meth-1.png" alt="A symmetrical heatmap where rows and columns are features in a DNA methylation dataset. Colour corresponds to correlation, with red being large positive correlations and blue being large negative correlations. There are large blocks of deep red and blue throughout the plot." class="figure mx-auto d-block"><figcaption>
Heatmap of pairwise feature-feature correlations between CpG sites in
DNA methylation data
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘For each observation, the left panel shows the residuals with respect to the optimal line (the one that minimises the sum of square errors). These are calculated as the difference between the value predicted by the line and the observed outcome. Right panel shows the sum of squared residuals across all possible linear regression models (as defined by different values of the regression coefficients).’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-regplot-1.png" alt="For each observation, the left panel shows the residuals with respect to the optimal line (the one that minimises the sum of square errors). These are calculated as the difference between the value predicted by the line and the observed outcome. Right panel shows the sum of squared residuals across all possible linear regression models (as defined by different values of the regression coefficients)." class="figure mx-auto d-block"><figcaption>
Illustrative example demonstrated how regression coefficients are
inferred under a linear model framework.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Schematic representation of how a dataset can be divided into a training (the portion of the data used to fit a model) and a test set (the portion of the data used to assess external generalisability).’</p>
<figure style="text-align: center"><img src="fig/validation.png" alt="Schematic representation of how a dataset can be divided into a training (the portion of the data used to fit a model) and a test set (the portion of the data used to assess external generalisability)." width="500px" class="figure mx-auto d-block"><figcaption>
Schematic representation of how a dataset can be divided into a training
and a test set.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of observed age versus predicted age for individuals in the test set. Each dot represents one individual. Dashed line is used as a reference to indicate how perfect predictions would look (observed = predicted). In this case we observe high prediction error in the test set.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-test-plot-lm-1.png" alt="A scatter plot of observed age versus predicted age for individuals in the test set. Each dot represents one individual. Dashed line is used as a reference to indicate how perfect predictions would look (observed = predicted). In this case we observe high prediction error in the test set." class="figure mx-auto d-block"><figcaption>
A scatter plot of observed age versus predicted age for individuals in
the test set. Each dot represents one individual. Dashed line is used as
a reference to indicate how perfect predictions would look (observed =
predicted).
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘For each observation, the left panel shows the residuals with respect to the optimal lines obtained with and without regularisation. Right panel shows the sum of squared residuals across all possible linear regression models. Regularisation moves the line away from the optimal (in terms of minimising the sum of squared residuals). ’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-ridgeplot-1.png" alt="For each observation, the left panel shows the residuals with respect to the optimal lines obtained with and without regularisation. Right panel shows the sum of squared residuals across all possible linear regression models. Regularisation moves the line away from the optimal (in terms of minimising the sum of squared residuals). " class="figure mx-auto d-block"><figcaption>
Illustrative example demonstrated how regression coefficients are
inferred under a linear model framework, with (blue line) and without
(red line) regularisation. A ridge penalty is used in this example
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot of coefficient estimates against log lambda for a ridge regression model. Lines are depicted in different colours, with coefficients generally having large values on the left of the plot (small log lambda) and moving smoothly and gradually towards zero to the right of the plot (large log lambda). Some coefficients appear to increase and then decrease in magnitude as lambda increases, or switch signs.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-plot-ridge-1.png" alt="A line plot of coefficient estimates against log lambda for a ridge regression model. Lines are depicted in different colours, with coefficients generally having large values on the left of the plot (small log lambda) and moving smoothly and gradually towards zero to the right of the plot (large log lambda). Some coefficients appear to increase and then decrease in magnitude as lambda increases, or switch signs." class="figure mx-auto d-block"><figcaption>
A line plot of coefficient estimates against log lambda for a ridge
regression model.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot of coefficient estimates against log lambda for a ridge regression model. A dashed vertical line depicts the optimal lambda value towards the left of the plot. Lines are depicted in different colours, with coefficients generally having large values on the left of the plot (small log lambda) and moving smoothly and gradually towards zero to the right of the plot (large log lambda). Some coefficients appear to increase and then decrease in magnitude as lambda increases, or switch signs.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-chooselambda-1.png" alt="A line plot of coefficient estimates against log lambda for a ridge regression model. A dashed vertical line depicts the optimal lambda value towards the left of the plot. Lines are depicted in different colours, with coefficients generally having large values on the left of the plot (small log lambda) and moving smoothly and gradually towards zero to the right of the plot (large log lambda). Some coefficients appear to increase and then decrease in magnitude as lambda increases, or switch signs." class="figure mx-auto d-block"><figcaption>
A line plot of coefficient estimates against log lambda for a ridge
regression model, showing the optimal value based on the minimal test
error.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two plots showing OLS predictions and ridge regression predictions of age (y) against true age (x). A dashed line shows the line y=x. In the OLS plot, predictions are quite extreme, while in the ridge regression plot, they are generally more conservative.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-plot-ridge-prediction-1.png" alt="Two plots showing OLS predictions and ridge regression predictions of age (y) against true age (x). A dashed line shows the line y=x. In the OLS plot, predictions are quite extreme, while in the ridge regression plot, they are generally more conservative." class="figure mx-auto d-block"><figcaption>
Two plots showing OLS predictions (left) and ridge regression
predictions (right) of age (y) against true age (x).
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘For each observation, the left panel shows the residuals with respect to the optimal lines obtained with and without regularisation. Right panel shows the sum of squared residuals across all possible linear regression models. Regularisation moves the line away from the optimal (in terms of minimising the sum of squared residuals)’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-shrink-lasso-1.png" alt="For each observation, the left panel shows the residuals with respect to the optimal lines obtained with and without regularisation. Right panel shows the sum of squared residuals across all possible linear regression models. Regularisation moves the line away from the optimal (in terms of minimising the sum of squared residuals)" class="figure mx-auto d-block"><figcaption>
Illustrative example demonstrated how regression coefficients are
inferred under a linear model framework, with (blue line) and without
(red line) regularisation. A LASSO penalty is used in this example.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘Two line plots side-by-side, showing coefficient values from a LASSO model against log lambda (left) and L1 norm (right). The coefficients, generally, suddenly become zero as log lambda increases or, equivalently, L1 norm decreases. However, some coefficients increase in size before decreasing as log lamdba increases.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-plotlas-1.png" alt="Two line plots side-by-side, showing coefficient values from a LASSO model against log lambda (left) and L1 norm (right). The coefficients, generally, suddenly become zero as log lambda increases or, equivalently, L1 norm decreases. However, some coefficients increase in size before decreasing as log lamdba increases." class="figure mx-auto d-block"><figcaption>
Line plots showing coefficient values from a LASSO model against log
lambda (left) and L1 norm (right).
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-11">Figure 11</h3>
<p aria-hidden="true">Image 1 of 1: ‘The data is divided into $K$ chunks. For each cross-validation iteration, one data chunk is used as the test set. The remaining $K-1$ chunks are combined into a training set.’</p>
<figure style="text-align: center"><img src="fig/cross_validation.png" alt="The data is divided into $K$ chunks. For each cross-validation iteration, one data chunk is used as the test set. The remaining $K-1$ chunks are combined into a training set." width="695" class="figure mx-auto d-block"><figcaption>
Schematic representiation of a <span class="math inline">\(K\)</span>-fold cross-validation procedure.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-12">Figure 12</h3>
<p aria-hidden="true">Image 1 of 1: ‘Alt’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-lasso-cv-1.png" alt="Alt" class="figure mx-auto d-block"><figcaption>
Cross-validated mean squared error for different values of lambda under
a LASSO penalty.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-13">Figure 13</h3>
<p aria-hidden="true">Image 1 of 1: ‘For lower values of alpha, the penalty resembles ridge regression. For higher values of alpha, the penalty resembles LASSO regression.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-elastic-contour-1.png" alt="For lower values of alpha, the penalty resembles ridge regression. For higher values of alpha, the penalty resembles LASSO regression." class="figure mx-auto d-block"><figcaption>
For an elastic net, the panels show the effect of the regularisation
across different values of alpha
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-14">Figure 14</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot showing coefficient values from an elastic net model against L1 norm. The coefficients, generally, suddenly become zero as L1 norm decreases. However, some coefficients increase in size before decreasing as L1 norm decreases.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-elastic-1.png" alt="A line plot showing coefficient values from an elastic net model against L1 norm. The coefficients, generally, suddenly become zero as L1 norm decreases. However, some coefficients increase in size before decreasing as L1 norm decreases." class="figure mx-auto d-block"><figcaption>
Line plot showing coefficient values from an elastic net model against
L1 norm.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-15">Figure 15</h3>
<p aria-hidden="true">Image 1 of 1: ‘A plot of the cross-validation mean squared error of an elastic net model against log lambda.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-elastic-cv-1.png" alt="A plot of the cross-validation mean squared error of an elastic net model against log lambda." class="figure mx-auto d-block"><figcaption>
A plot of the cross-validation mean squared error of an elastic net
model against log lambda.
</figcaption></figure><hr>
<h3 id="03-regression-regularisation-figure-16">Figure 16</h3>
<p aria-hidden="true">Image 1 of 1: ‘A plot of the cross-validation binomial deviance of a logistic regression elastic net model against log lambda.’</p>
<figure style="text-align: center"><img src="fig/03-regression-regularisation-rendered-binomial-1.png" alt="A plot of the cross-validation binomial deviance of a logistic regression elastic net model against log lambda." class="figure mx-auto d-block"><figcaption>
A plot of the cross-validation binomial deviance of a logistic
regression elastic net model against log lambda.
</figcaption></figure><hr></section><section id="04-principal-component-analysis"><h2 class="section-heading"><a href="04-principal-component-analysis.html">Principal component analysis</a></h2>
<hr class="half-width">
<h3 id="04-principal-component-analysis-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Side-by-side scatter plots of the prostate data. The left figure displays a scatter plot of the log prostate specific antigen versus the log cancer volume. The first principal component is shown by a red line and the second principal component is shown by a green line. The right figure displays the same scatter plot rotated so that the first principal component is horizontal and the second principal component is shown perpendicular to this.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-pros-pcs-1.png" alt="Side-by-side scatter plots of the prostate data. The left figure displays a scatter plot of the log prostate specific antigen versus the log cancer volume. The first principal component is shown by a red line and the second principal component is shown by a green line. The right figure displays the same scatter plot rotated so that the first principal component is horizontal and the second principal component is shown perpendicular to this." class="figure mx-auto d-block"><figcaption>
Scatter plots of the prostate data with the first principal component in
red and second principal component in green.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Animated scatter plot displaying the iterative process by which principal components are found. The data points are shown in blue and the principal component is shown by a solid black line. The distances of the points from the line are shown by red dashed lines. The animation initially starts with the principal component far away from the direction of variability in the points and, as time goes on, eventually finds the direction of variability exhibiting a springing motion.’</p>
<figure style="text-align: center"><img src="fig/pendulum.gif" alt="Animated scatter plot displaying the iterative process by which principal components are found. The data points are shown in blue and the principal component is shown by a solid black line. The distances of the points from the line are shown by red dashed lines. The animation initially starts with the principal component far away from the direction of variability in the points and, as time goes on, eventually finds the direction of variability exhibiting a springing motion." class="figure mx-auto d-block"><figcaption>
Animation showing the iterative process by which principal components
are found. Data points are shown in blue, distances from the line as red
dashed lines and the principal component is shown by a black line.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Side-by-side histograms of two variables from the dataset, lweight on the left and lbph on the right. The histogram for the lweight data ranges from around 2 to 6 on the x axis, while the histogram for the lbph ranges from around -2 to 3 on the x axis.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-var-hist-1.png" alt="Side-by-side histograms of two variables from the dataset, lweight on the left and lbph on the right. The histogram for the lweight data ranges from around 2 to 6 on the x axis, while the histogram for the lbph ranges from around -2 to 3 on the x axis." class="figure mx-auto d-block"><figcaption>
Histograms of two variables from the prostate data set.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scree plot showing the percentage of variance explained by each principal component versus the principal component number. The points are joined by lines to indicate where the elbow of the scree plot occurs.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-varexp-1.png" alt="A scree plot showing the percentage of variance explained by each principal component versus the principal component number. The points are joined by lines to indicate where the elbow of the scree plot occurs." class="figure mx-auto d-block"><figcaption>
Scree plot showing the percentage of the variance explained by the
principal components calculated from the prostate data.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘Side-by-side scatter plots of the second versus first principal components calculated by prcomp() (left) and by hand (right). The left scatter plot is the same as the right scatter plot but with swapped axes.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-pca-by-hand-1.png" alt="Side-by-side scatter plots of the second versus first principal components calculated by prcomp() (left) and by hand (right). The left scatter plot is the same as the right scatter plot but with swapped axes." class="figure mx-auto d-block"><figcaption>
Scatter plots of the second versus first principal components calculated
by prcomp() (left) and by hand (right).
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot of the second principal component versus the first principal component. Observations as points, numerically labelled.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-stats-biplot-1.png" alt="Scatter plot of the second principal component versus the first principal component. Observations as points, numerically labelled." class="figure mx-auto d-block"><figcaption>
Scatter plot of the first two principal components with observations
numerically labelled.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A bar and line plot showing the variance explained by principal components (PCs) of gene expression data. Blue bars depict the variance explained by each PC, while a red line depicts the cumulative variance explained by the PCs. The first principal component explains roughly 30% of the variance, while succeeding PCs explain less than 10%.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-scree-ex-1.png" alt="A bar and line plot showing the variance explained by principal components (PCs) of gene expression data. Blue bars depict the variance explained by each PC, while a red line depicts the cumulative variance explained by the PCs. The first principal component explains roughly 30% of the variance, while succeeding PCs explain less than 10%." class="figure mx-auto d-block"><figcaption>
A scree plot of the gene expression data.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘A biplot of PC2 against PC1 in the gene expression data, coloured by Grade. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on Grade across these two groups.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-biplot-ex-1.png" alt="A biplot of PC2 against PC1 in the gene expression data, coloured by Grade. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on Grade across these two groups." class="figure mx-auto d-block"><figcaption>
A biplot of PC2 against PC1 in the gene expression data, coloured by
Grade.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘A biplot of PC2 against PC1 in the gene expression data, coloured by ER status. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on ER across these two groups, although there are more ER- samples in the rightmost cluster. Patient ages are overlaid as text near the points, but there is again no clear pattern.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-pca-biplot-ex2-1.png" alt="A biplot of PC2 against PC1 in the gene expression data, coloured by ER status. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on ER across these two groups, although there are more ER- samples in the rightmost cluster. Patient ages are overlaid as text near the points, but there is again no clear pattern." class="figure mx-auto d-block"><figcaption>
A biplot of PC2 against PC1 in the gene expression data, coloured by ER
status with patient ages overlaid.
</figcaption></figure><hr>
<h3 id="04-principal-component-analysis-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘A triangular grid of scatter plots. The grid is the upper right triangle of a square, where each entry of the grid corresponds to a plot of one principal component against another. For example, the plot in the upper left corner of the plot corresponds to principal component one plotted against principal component 2, and the plot to the right of this plots principal component 1 against principal component 3. Points correspond to samples, and are coloured arbitrarily from light blue to dark blue.’</p>
<figure style="text-align: center"><img src="fig/04-principal-component-analysis-rendered-pairsplot-1.png" alt="A triangular grid of scatter plots. The grid is the upper right triangle of a square, where each entry of the grid corresponds to a plot of one principal component against another. For example, the plot in the upper left corner of the plot corresponds to principal component one plotted against principal component 2, and the plot to the right of this plots principal component 1 against principal component 3. Points correspond to samples, and are coloured arbitrarily from light blue to dark blue." class="figure mx-auto d-block"><figcaption>
Multiple biplots produced by pairsplot().
</figcaption></figure><hr></section><section id="05-factor-analysis"><h2 class="section-heading"><a href="05-factor-analysis.html">Factor analysis</a></h2>
<hr class="half-width">
<h3 id="05-factor-analysis-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘A table displaying data of student scores across several subjects. Each row displays the scores across different subjects for a given individual. The plot is annotated at the top with a curly bracket labelled Factor 1: mathematical ability and encompasses the data for the student scores is Arithmetic, Algebra, Geometry, and Statistics. Similarly, the subjects Creative Writing, Literature, Spelling/Grammar are encompassed by a different curly bracket with label Factor 2: writing ability.’</p>
<figure style="text-align: center"><img src="fig/table_for_fa.png" alt="A table displaying data of student scores across several subjects. Each row displays the scores across different subjects for a given individual. The plot is annotated at the top with a curly bracket labelled Factor 1: mathematical ability and encompasses the data for the student scores is Arithmetic, Algebra, Geometry, and Statistics. Similarly, the subjects Creative Writing, Literature, Spelling/Grammar are encompassed by a different curly bracket with label Factor 2: writing ability." width="5448" class="figure mx-auto d-block"><figcaption>
Student scores data across several subjects with hypothesised factors.
</figcaption></figure><hr>
<h3 id="05-factor-analysis-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of the factor 2 loadings for each feature versus the factor 2 loadings for each feature. The lpsa, lcavol and lcp feature points are located in the east of the plot, indicating a high loading on factor 1 and close to zero loading on factor 2. The lbph and lweight features are located in the north of the plot, indicating a close to zero loading on factor 1 and a high loading on factor 2.’</p>
<figure style="text-align: center"><img src="fig/05-factor-analysis-rendered-biplot-1.png" alt="A scatter plot of the factor 2 loadings for each feature versus the factor 2 loadings for each feature. The lpsa, lcavol and lcp feature points are located in the east of the plot, indicating a high loading on factor 1 and close to zero loading on factor 2. The lbph and lweight features are located in the north of the plot, indicating a close to zero loading on factor 1 and a high loading on factor 2." class="figure mx-auto d-block"><figcaption>
Factor 2 loadings versus factor 1 loadings for each feature.
</figcaption></figure><hr>
<h3 id="05-factor-analysis-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot of random data y versus x. There are many black points on the plot representing the data. Two additional points are shown: the (mean(x), mean(y)) co-ordinate point in red and the (median(x), median(y)) co-ordinate point in blue. The median co-ordinate point in blue has a lower x value and is shown to the left of the red mean co-ordinate point.’</p>
<figure style="text-align: center"><img src="fig/05-factor-analysis-rendered-unnamed-chunk-1-1.png" alt="Scatter plot of random data y versus x. There are many black points on the plot representing the data. Two additional points are shown: the (mean(x), mean(y)) co-ordinate point in red and the (median(x), median(y)) co-ordinate point in blue. The median co-ordinate point in blue has a lower x value and is shown to the left of the red mean co-ordinate point." class="figure mx-auto d-block"><figcaption>
Scatter plot of random data y versus x with the (mean(x), mean(y))
co-ordinate point shown in red and the (median(x), median(y))
co-ordinate point shown in blue.
</figcaption></figure><hr></section><section id="06-k-means"><h2 class="section-heading"><a href="06-k-means.html">K-means</a></h2>
<hr class="half-width">
<h3 id="06-k-means-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘An animated scatter plot of data y versus x. The animation starts by identifying 3 initial points, delineated by different colours. The animations then colour codes all points by an associated cluster colour, delineating three distinct and non-overlapping clusters in the space of the scatter plot.’</p>
<figure style="text-align: center"><img src="fig/kmeans.gif" alt="An animated scatter plot of data y versus x. The animation starts by identifying 3 initial points, delineated by different colours. The animations then colour codes all points by an associated cluster colour, delineating three distinct and non-overlapping clusters in the space of the scatter plot." class="figure mx-auto d-block"><figcaption>
Animation showing the iterative process of K-means clustering on data y
versus x.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of random data y versus x. The points are horizontally partitioned at 2 random groups, forming three colour coded clusters. Circles are drawn around each cluster. The data shown appears to have no clusters but the colours and circles give the appearance of clusters artificially.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-fake-cluster-1.png" alt="A scatter plot of random data y versus x. The points are horizontally partitioned at 2 random groups, forming three colour coded clusters. Circles are drawn around each cluster. The data shown appears to have no clusters but the colours and circles give the appearance of clusters artificially." class="figure mx-auto d-block"><figcaption>
Example of artificial clusters fitted to data points.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of principal component 2 versus principal component 1 of the scrnaseq data. Each point is one of four colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing four distinct clusters in the data.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-kmeans-1.png" alt="A scatter plot of principal component 2 versus principal component 1 of the scrnaseq data. Each point is one of four colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing four distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 versus principal component 1 with
points colour coded according to the cluster to which they belong.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-kmeans-ex-1.png" alt="A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 against principal component 1 in
the scRNAseq data, coloured by clusters produced by k-means clustering.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot of random data y versus x. There are many black points on the plot representing the data. Two additional points are shown: the (mean(x), mean(y)) co-ordinate point in red and the (median(x), median(y)) co-ordinate point in blue. The median co-ordinate point in blue has a lower x value and is shown to the left of the red mean co-ordinate point.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-unnamed-chunk-1-1.png" alt="Scatter plot of random data y versus x. There are many black points on the plot representing the data. Two additional points are shown: the (mean(x), mean(y)) co-ordinate point in red and the (median(x), median(y)) co-ordinate point in blue. The median co-ordinate point in blue has a lower x value and is shown to the left of the red mean co-ordinate point." class="figure mx-auto d-block"><figcaption>
Scatter plot of random data y versus x with the (mean(x), mean(y))
co-ordinate point shown in red and the (median(x), median(y))
co-ordinate point shown in blue.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains over half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have slightly lower silhouette widths. The bars therefore reach further to the left than cluster 4.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-silhouette-1.png" alt="Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains over half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have slightly lower silhouette widths. The bars therefore reach further to the left than cluster 4." class="figure mx-auto d-block"><figcaption>
Silhouette plot for each point according to cluster.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of the random y versus x data. Cluster membership is delineated using different point characters. Data points in the same cluster have the same point character. Each point is coloured by its silhouette width: solid red delineating a silhouette width of 1 and white delineating a silhouette width of 0. Colours in between delineate the intermediate colours. Many points are red and fade to white at the boundaries of each cluster. ’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-plot-silhouette-1.png" alt="A scatter plot of the random y versus x data. Cluster membership is delineated using different point characters. Data points in the same cluster have the same point character. Each point is coloured by its silhouette width: solid red delineating a silhouette width of 1 and white delineating a silhouette width of 0. Colours in between delineate the intermediate colours. Many points are red and fade to white at the boundaries of each cluster. " class="figure mx-auto d-block"><figcaption>
Scatter plot of y versus x coloured according to silhouette width and
point characters grouped according to cluster membership.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-silhouette-ex-1.png" alt="A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 against principal component 1 in
the scRNAseq data, coloured by clusters produced by k-means clustering.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains almost half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have similar silhouette widths. The bars for cluster 5 are much smaller, with a small number extending to the left of the origin, indicating negative silhouette widths.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-unnamed-chunk-2-1.png" alt="Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains almost half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have similar silhouette widths. The bars for cluster 5 are much smaller, with a small number extending to the left of the origin, indicating negative silhouette widths." class="figure mx-auto d-block"><figcaption>
Silhouette plot for each point according to cluster.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘A histogram of the mean of each bootstrapped sample. The histogram appears roughly symmetric around 2.8 on the x axis.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-boots-1.png" alt="A histogram of the mean of each bootstrapped sample. The histogram appears roughly symmetric around 2.8 on the x axis." class="figure mx-auto d-block"><figcaption>
Histogram of mean of bootstapped samples.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-11">Figure 11</h3>
<p aria-hidden="true">Image 1 of 1: ‘Grid of 16 squares labelled 1-4 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-bs-heatmap-1.png" alt="Grid of 16 squares labelled 1-4 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares." class="figure mx-auto d-block"><figcaption>
Grid of empirical cluster swapping behaviour estimated by the bootstrap
samples.
</figcaption></figure><hr>
<h3 id="06-k-means-figure-12">Figure 12</h3>
<p aria-hidden="true">Image 1 of 1: ‘Grid of 25 squares labelled 1-5 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend, with the exception of the square corresponding to (4, 5), which is slightly darker green indicating a lower value. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares.’</p>
<figure style="text-align: center"><img src="fig/06-k-means-rendered-bs-ex-1.png" alt="Grid of 25 squares labelled 1-5 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend, with the exception of the square corresponding to (4, 5), which is slightly darker green indicating a lower value. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares." class="figure mx-auto d-block"><figcaption>
Grid of empirical cluster swapping behaviour estimated by the bootstrap
samples.
</figcaption></figure><hr></section><section id="07-hierarchical"><h2 class="section-heading"><a href="07-hierarchical.html">Hierarchical clustering</a></h2>
<hr class="half-width">
<h3 id="07-hierarchical-figure-1">Figure 1</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot of observations x2 versus x1. Two clusters of pairs of observations are shown by blue and red boxes, each grouping two observations that are close in their x and y distance.’</p>
<figure style="text-align: center"><img src="fig/hierarchical_clustering_1.png" alt="Scatter plot of observations x2 versus x1. Two clusters of pairs of observations are shown by blue and red boxes, each grouping two observations that are close in their x and y distance." width="500px" class="figure mx-auto d-block"><figcaption>
Example data showing two clusters of observation pairs.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-2">Figure 2</h3>
<p aria-hidden="true">Image 1 of 1: ‘Scatter plot of observations x2 versus x1. Three boxes are shown this time. Blue and red boxes contain two observations each. The two boxes contain points that are relatively far apart. A third green box is shown encompassing the blue box and an additional data point.’</p>
<figure style="text-align: center"><img src="fig/hierarchical_clustering_2.png" alt="Scatter plot of observations x2 versus x1. Three boxes are shown this time. Blue and red boxes contain two observations each. The two boxes contain points that are relatively far apart. A third green box is shown encompassing the blue box and an additional data point." width="500px" class="figure mx-auto d-block"><figcaption>
Example data showing fusing of one observation into larger cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-3">Figure 3</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of randomly-generated data x2 versus x1. The points appear fairly randomly scattered, arguably centered towards the bottom of the plot.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plotexample-1.png" alt="A scatter plot of randomly-generated data x2 versus x1. The points appear fairly randomly scattered, arguably centered towards the bottom of the plot." class="figure mx-auto d-block"><figcaption>
Scatter plot of randomly-generated data x2 versus x1.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-4">Figure 4</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other).’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plotclustex-1.png" alt="A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other)." class="figure mx-auto d-block"><figcaption>
A dendrogram of the randomly-generated data.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-5">Figure 5</h3>
<p aria-hidden="true">Image 1 of 1: ‘A dendrogram for the methyl_mat data with boxes overlaid on clusters. There are 5 boxes in total, each indicating separate clusters.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-method-1.png" alt="A dendrogram for the methyl_mat data with boxes overlaid on clusters. There are 5 boxes in total, each indicating separate clusters." class="figure mx-auto d-block"><figcaption>
Dendrogram with boxes around clusters.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-6">Figure 6</h3>
<p aria-hidden="true">Image 1 of 1: ‘A dendrogram with the different clusters in 4 different colours.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-coloured-branches-1.png" alt="A dendrogram with the different clusters in 4 different colours." class="figure mx-auto d-block"><figcaption>
Dendrogram with coloured branches delineating different clusters.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-7">Figure 7</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other). A dashed horizontal line at a height of 5 indicates the cut point used to divide the data into clusters.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-h-k-ex-plot-1.png" alt="A line plot depicting a dendrogram --- a tree structure representing the hierarchical structure of the data. The data broadly fit into three clusters, with one sample (14) being quite dissimilar to all others, and the rest of the data comprising two other clusters (one larger than the other). A dashed horizontal line at a height of 5 indicates the cut point used to divide the data into clusters." class="figure mx-auto d-block"><figcaption>
A dendrogram of the randomly-generated data.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-8">Figure 8</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of data simulated to form two crescent shapes. The crescents are horizontally orientated with a a rough line of vertical symmetry.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-crescents-1.png" alt="A scatter plot of data simulated to form two crescent shapes. The crescents are horizontally orientated with a a rough line of vertical symmetry." class="figure mx-auto d-block"><figcaption>
Scatter plot of data simulated according to two crescent-shaped point
clouds.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-9">Figure 9</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of the crescent-shaped simulated data calculated using Euclidean distance. The points are coloured in black or red according to their membership to 2 clusters. The points in the tails of each crescent have inherited the colour of the opposite crescent.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-cresClustDefault-1.png" alt="A scatter plot of the crescent-shaped simulated data calculated using Euclidean distance. The points are coloured in black or red according to their membership to 2 clusters. The points in the tails of each crescent have inherited the colour of the opposite crescent." class="figure mx-auto d-block"><figcaption>
Scatter plot of crescent-shaped simulated data, coloured according to
clusters calculated using Euclidean distance.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-10">Figure 10</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with single linkage, with two clusters, corresponding to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-comp-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with single linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-11">Figure 11</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D linkage, with two clusters, corresponding to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-wardD-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-12">Figure 12</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D2 linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-wardD2-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with ward.D2 linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-13">Figure 13</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with average linkage, with two clusters, corresponding to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-average-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with average linkage, with two clusters, corresponding to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-14">Figure 14</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with mcquitty linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-mcq-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with mcquitty linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-15">Figure 15</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with median linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-median-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with median linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-16">Figure 16</h3>
<p aria-hidden="true">Image 1 of 1: ‘A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with centroid linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-centroid-1.png" alt="A scatter plot of synthetic data, comprising two variables, with points forming two crescent-shaped clusters. Points are coloured based on hierarchical clustering with centroid linkage, with two clusters, though these do not correspond to the two crescent-shaped clouds." class="figure mx-auto d-block"><figcaption>
A scatter plot of synthetic data coloured by cluster.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-17">Figure 17</h3>
<p aria-hidden="true">Image 1 of 1: ‘A line plot of simulated value versus observation number, coloured by sample. Samples a and b are concentrated at the bottom of the plot, while sample c is concentrated at the top of the plot. Samples a and c have exactly the same vertical pattern.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-lineplot-cor-example-1.png" alt="A line plot of simulated value versus observation number, coloured by sample. Samples a and b are concentrated at the bottom of the plot, while sample c is concentrated at the top of the plot. Samples a and c have exactly the same vertical pattern." class="figure mx-auto d-block"><figcaption>
Line plot of simulated value versus observation number, coloured by
sample.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-18">Figure 18</h3>
<p aria-hidden="true">Image 1 of 1: ‘A dendrogram of the example simulated data clustered according to Euclidean distance. The dendrogram shows that sample c definitively forms its own cluster for any cut height and samples a and b merge into a cluster at a height of around 6.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-clust-euc-cor-example-1.png" alt="A dendrogram of the example simulated data clustered according to Euclidean distance. The dendrogram shows that sample c definitively forms its own cluster for any cut height and samples a and b merge into a cluster at a height of around 6." class="figure mx-auto d-block"><figcaption>
Dendrogram of the example simulated data clustered according to
Euclidean distance.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-19">Figure 19</h3>
<p aria-hidden="true">Image 1 of 1: ‘A dendrogram of the example simulated data clustered according to correlation. The dendrogram shows that sample b definitively forms its own cluster and samples a and c form definitively form their own cluster for any cut height.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-clust-cor-cor-example-1.png" alt="A dendrogram of the example simulated data clustered according to correlation. The dendrogram shows that sample b definitively forms its own cluster and samples a and c form definitively form their own cluster for any cut height." class="figure mx-auto d-block"><figcaption>
Dendrogram of the example simulated data clustered according to
correlation.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-20">Figure 20</h3>
<p aria-hidden="true">Image 1 of 1: ‘Heatmaps of features versus samples, coloured by simulated value. The columns (samples) are clustered according to the correlation. Samples a and b have mostly low values, delineated by blue in the first plot and yellow in the second plot. Sample c has mostly high values, delineated by red in the first plot and brown in the second plot. Samples A and C form a cluster separate from sample B in the column dendrogram.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-heatmap-cor-cor-example-1.png" alt="Heatmaps of features versus samples, coloured by simulated value. The columns (samples) are clustered according to the correlation. Samples a and b have mostly low values, delineated by blue in the first plot and yellow in the second plot. Sample c has mostly high values, delineated by red in the first plot and brown in the second plot. Samples A and C form a cluster separate from sample B in the column dendrogram." class="figure mx-auto d-block"><figcaption>
Heatmaps of features versus samples clustered in the samples according
to correlation.
</figcaption></figure><hr>
<h3 id="07-hierarchical-figure-21">Figure 21</h3>
<p aria-hidden="true">Image 1 of 1: ‘A dendrogram for clustering of methylation data. Identical to that in the section Highlighting dendrogram branches, without the colour overlay to show clusters.’</p>
<figure style="text-align: center"><img src="fig/07-hierarchical-rendered-plot-clust-dunn-1.png" alt="A dendrogram for clustering of methylation data. Identical to that in the section Highlighting dendrogram branches, without the colour overlay to show clusters." class="figure mx-auto d-block"><figcaption>
Dendrogram for clustering of methylation data.
</figcaption></figure><hr></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/images.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All Images",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/images.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/images.html",
  "dateCreated": "2025-02-03",
  "dateModified": "2025-03-04",
  "datePublished": "2025-03-04"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

