<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>High dimensional statistics with R: Principal component analysis</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../04-principal-component-analysis.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="data.html">Data</a></li><hr></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 50%" class="percentage">
    50%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 50%" aria-valuenow="50" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../04-principal-component-analysis.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction-to-high-dimensional-data.html">1. Introduction to high-dimensional data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-high-dimensional-regression.html">2. Regression with many outcomes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Principal component analysis
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#principal-component-analysis">Principal component analysis</a></li>
<li><a href="#how-do-we-perform-a-pca">How do we perform a PCA?</a></li>
<li><a href="#how-many-principal-components-do-we-need">How many principal components do we need?</a></li>
<li><a href="#loadings-and-principal-component-scores">Loadings and principal component scores</a></li>
<li><a href="#advantages-and-disadvantages-of-pca">Advantages and disadvantages of PCA</a></li>
<li><a href="#using-pca-to-analyse-gene-expression-data">Using PCA to analyse gene expression data</a></li>
<li><a href="#choosing-how-many-components-are-important-to-explain-the-variance-in-the-data">Choosing how many components are important to explain the variance
in the data</a></li>
<li><a href="#investigating-the-principal-components">Investigating the principal components</a></li>
<li><a href="#using-pca-output-in-further-analysis">Using PCA output in further analysis</a></li>
<li><a href="#further-reading">Further reading</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-k-means.html">6. K-means</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-hierarchical.html">7. Hierarchical clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="data.html">Data</a></li><hr></ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/03-regression-regularisation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/05-factor-analysis.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/03-regression-regularisation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Regularised
        </a>
        <a class="chapter-link float-end" href="../instructor/05-factor-analysis.html" rel="next">
          Next: Factor analysis...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Principal component analysis</h1>
        <p>Last updated on 2025-02-03 |

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/04-principal-component-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 130 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is principal component analysis (PCA) and when can it be
used?</li>
<li>How can we perform a PCA in R?</li>
<li>How many principal components are needed to explain a significant
amount of variation in the data?</li>
<li>How to interpret the output of PCA using loadings and principal
components?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Identify situations where PCA can be used to answer research
questions using high-dimensional data.</li>
<li>Perform a PCA on high-dimensional data.</li>
<li>Select the appropriate number of principal components.</li>
<li>Interpret the output of PCA.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>If a dataset contains many variables (<span class="math inline">\(p\)</span>), it is likely that some of these
variables will be highly correlated. Variables may even be so highly
correlated that they represent the same overall effect. Such datasets
are challenging to analyse for several reasons, with the main problem
being how to reduce dimensionality in the dataset while retaining the
important features.</p>
<p>In this episode we will explore <em>principal component analysis</em>
(PCA) as a popular method of analysing high-dimensional data. PCA is a
statistical method which allows large datasets of correlated variables
to be summarised into smaller numbers of uncorrelated principal
components that explain most of the variability in the original dataset.
As an example, PCA might reduce several variables representing aspects
of patient health (blood pressure, heart rate, respiratory rate) into a
single feature capturing an overarching “patient health” effect. This is
useful from an exploratory point of view, discovering how variables
might be associated and combined. The resulting principal component
could also be used as an effect in further analysis (e.g. linear
regression).</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1</h3>
<div class="callout-content">
<p>Descriptions of three datasets and research questions are given
below. For which of these might PCA be considered a useful tool for
analysing data so that the research questions may be addressed?</p>
<ol style="list-style-type: decimal"><li>An epidemiologist has data collected from different patients
admitted to hospital with infectious respiratory disease. They would
like to determine whether length of stay in hospital differs in patients
with different respiratory diseases.</li>
<li>A scientist has assayed gene expression levels in 1000 cancer
patients and has data from probes targeting different genes in tumour
samples from patients. She would like to create new variables
representing relative abundance of different groups of genes to i) find
out if genes form subgroups based on biological function and ii) use
these new variables in a linear regression examining how gene expression
varies with disease severity.</li>
<li>Both of the above.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In the first case, a regression model would be more suitable; perhaps
a survival model. In the second example, PCA can help to identify
modules of correlated features that explain a large amount of variation
within the data.</p>
<p>Therefore the answer here is 2.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="principal-component-analysis">Principal component analysis<a class="anchor" aria-label="anchor" href="#principal-component-analysis"></a></h2>
<hr class="half-width"><p>PCA transforms a dataset of continuous variables into a new set of
uncorrelated variables called “principal components”. The first
principal component derived explains the largest amount of the
variability in the underlying dataset. The second principal component
derived explains the second largest amount of variability in the dataset
and so on. Once the dataset has been transformed into principal
components, we can extract a subset of the principal components in order
of the variance they explain (starting with the first principal
component that by definition explains the most variability, and then the
second), giving new variables that explain a lot of the variability in
the original dataset. Thus, PCA helps us to produce a lower dimensional
dataset while keeping most of the information in the original
dataset.</p>
<p>To see what these new principal components may look like, the figure
below shows the log-transformed prostate specific antigen
(<code>lpsa</code>) versus cancer volume (<code>lcavol</code>) data from
the <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link"><code>prostate</code></a>
data set introduced in Episode 1. Principal components are a collection
of new, artificial data points, one for each individual observation
called <em>scores</em>. The red line on the plot represents the line
passing through the scores (points) of the first principal component for
each observation. The angle that the first principal component line
passes through the data points at is set to the direction with the
highest variability. The plotted first principal components can
therefore be thought of reflecting the effect in the data that has the
highest variability. The second principal component explains the next
highest amount of variability in the data and is represented by the line
perpendicular to the first (the green line). The second principal
component can be thought of as capturing the overall effect in the data
that has the second-highest variability.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: ggrepel</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'PCAtools'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    biplot, screeplot</code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-pros-pcs-1.png" alt="Side-by-side scatter plots of the prostate data. The left figure displays a scatter plot of the log prostate specific antigen versus the log cancer volume. The first principal component is shown by a red line and the second principal component is shown by a green line. The right figure displays the same scatter plot rotated so that the first principal component is horizontal and the second principal component is shown perpendicular to this." class="figure mx-auto d-block"><figcaption>
Scatter plots of the prostate data with the first principal component in
red and second principal component in green.
</figcaption></figure><p>The animation below illustrates how principal components are
calculated from data. You can imagine that the black line is a rod and
each red dashed line is a spring. The energy of each spring is
proportional to its squared length. The direction of the first principal
component is the one that minimises the total energy of all of the
springs. In the animation below, the springs pull the rod, finding the
direction of the first principal component when they reach equilibrium.
We then use the length of the springs from the rod as the first
principal component. This is explained in more detail on <a href="https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues" class="external-link">this
Q&amp;A website</a>.</p>
<figure style="text-align: center"><img src="../fig/pendulum.gif" alt="Animated scatter plot displaying the iterative process by which principal components are found. The data points are shown in blue and the principal component is shown by a solid black line. The distances of the points from the line are shown by red dashed lines. The animation initially starts with the principal component far away from the direction of variability in the points and, as time goes on, eventually finds the direction of variability exhibiting a springing motion." class="figure mx-auto d-block"><figcaption>
Animation showing the iterative process by which principal components
are found. Data points are shown in blue, distances from the line as red
dashed lines and the principal component is shown by a black line.
</figcaption></figure><div id="mathematical-description-of-pca" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="mathematical-description-of-pca" class="callout-inner">
<h3 class="callout-title">Mathematical description of PCA</h3>
<div class="callout-content">
<p>Mathematically, each principal component is a linear combination of
the variables in the dataset. That is, the first principal component
values or <em>scores</em>, <span class="math inline">\(Z_1\)</span>, are
a linear combination of variables in the dataset, <span class="math inline">\(X_1...X_p\)</span>, given by</p>
<p><span class="math display">\[ Z_1 = a_{11}X_1 + a_{21}X_2
+....+a_{p1}X_p, \]</span></p>
<p>where <span class="math inline">\(a_{11}...a_{p1}\)</span> represent
principal component <em>loadings</em>.</p>
</div>
</div>
</div>
<p>In summary, the principal components’ values are called
<em>scores</em>. The loadings can be thought of as the degree to which
each original variable contributes to the principal component scores. In
this episode, we will see how to perform PCA to summarise the
information in high-dimensional datasets.</p>
</section><section><h2 class="section-heading" id="how-do-we-perform-a-pca">How do we perform a PCA?<a class="anchor" aria-label="anchor" href="#how-do-we-perform-a-pca"></a></h2>
<hr class="half-width"><p>To illustrate how to perform PCA initially, we revisit the
low-dimensional <code>prostate</code> dataset. The data come from a
study which examined the correlation between the level of prostate
specific antigen and a number of clinical measures in 97 men who were
about to receive a radical prostatectomy. The data have 97 rows and 9
columns.</p>
<p>The variables in the dataset (columns) are:</p>
<ul><li>
<code>lcavol</code> (log-transformed cancer volume),</li>
<li>
<code>lweight</code> (log-transformed prostate weight),</li>
<li>
<code>lbph</code> (log-transformed amount of benign prostate
enlargement),</li>
<li>
<code>svi</code> (seminal vesicle invasion),</li>
<li>
<code>lcp</code> (log-transformed capsular penetration; amount of
spread of cancer in outer walls of prostate),</li>
<li>
<code>gleason</code> (Gleason score; grade of cancer cells),</li>
<li>
<code>pgg45</code> (percentage Gleason scores 4 or 5),</li>
<li>
<code>lpsa</code> (log-transformed prostate specific antigen; level
of PSA in blood).</li>
<li>
<code>age</code> (patient age in years).</li>
</ul><p>We will perform PCA on the five continuous clinical variables in our
dataset so that we can create fewer variables representing clinical
markers of cancer progression.</p>
<p>First, we will examine the <code>prostate</code> dataset (originally
part of the <strong><code>lasso2</code></strong> package):</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prostate</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/prostate.rds"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">prostate</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight age      lbph svi       lcp gleason pgg45       lpsa
1 -0.5798185 2.769459  50 -1.386294   0 -1.386294       6     0 -0.4307829
2 -0.9942523 3.319626  58 -1.386294   0 -1.386294       6     0 -0.1625189
3 -0.5108256 2.691243  74 -1.386294   0 -1.386294       7    20 -0.1625189
4 -1.2039728 3.282789  58 -1.386294   0 -1.386294       6     0 -0.1625189
5  0.7514161 3.432373  62 -1.386294   0 -1.386294       6     0  0.3715636
6 -1.0498221 3.228826  50 -1.386294   0 -1.386294       6     0  0.7654678</code></pre>
</div>
<p>Note that each row of the dataset represents a single patient.</p>
<p>We will create a subset of the data including only the clinical
variables we want to use in the PCA.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2</span> <span class="op">&lt;-</span> <span class="va">prostate</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"lcavol"</span>, <span class="st">"lweight"</span>, <span class="st">"lbph"</span>, <span class="st">"lcp"</span>, <span class="st">"lpsa"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">pros2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight      lbph       lcp       lpsa
1 -0.5798185 2.769459 -1.386294 -1.386294 -0.4307829
2 -0.9942523 3.319626 -1.386294 -1.386294 -0.1625189
3 -0.5108256 2.691243 -1.386294 -1.386294 -0.1625189
4 -1.2039728 3.282789 -1.386294 -1.386294 -0.1625189
5  0.7514161 3.432373 -1.386294 -1.386294  0.3715636
6 -1.0498221 3.228826 -1.386294 -1.386294  0.7654678</code></pre>
</div>
<div class="section level3">
<h3 id="do-we-need-to-scale-the-data">Do we need to scale the data?<a class="anchor" aria-label="anchor" href="#do-we-need-to-scale-the-data"></a></h3>
<p>PCA derives principal components based on the variance they explain
in the data. Therefore, we may need to apply some pre-processing to
scale variables in our dataset if we want to ensure that each variable
is considered equally by the PCA. Scaling is essential if we want to
avoid the PCA ignoring variables that may be important to our analysis
just because they take low values and have low variance. We do not need
to scale if we want variables with low variance to carry less weight in
the PCA.</p>
<p>In this example, we want each variable to be treated equally by the
PCA since variables with lower values may be just as informative as
variables with higher values. Let’s therefore investigate the variables
in our dataset to see if we need to scale our variables first:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">apply</span><span class="op">(</span><span class="va">pros2</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  lcavol  lweight     lbph      lcp     lpsa
1.389157 0.246642 2.104840 1.955102 1.332476 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lweight</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lbph</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-var-hist-1.png" alt="Side-by-side histograms of two variables from the dataset, lweight on the left and lbph on the right. The histogram for the lweight data ranges from around 2 to 6 on the x axis, while the histogram for the lbph ranges from around -2 to 3 on the x axis." class="figure mx-auto d-block"><figcaption>
Histograms of two variables from the prostate data set.
</figcaption></figure><p>Note that variance is greatest for <code>lbph</code> and lowest for
<code>lweight</code>. Since we want each of the variables to be treated
equally in our PCA, but there are large differences in the variances of
the variables, we need to scale each of the variables before including
them in a PCA to ensure that differences in variances do not drive the
calculation of principal components. In this example, we standardise all
five variables to have a mean of 0 and a standard deviation of 1.</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Why might it be necessary to standardise variables before performing
a PCA?<br>
Some ideas:</li>
</ol><ul><li>To make the results of the PCA interesting.</li>
<li>If you want to ensure that variables with different ranges of values
contribute equally to analysis.</li>
<li>To allow the feature matrix to be calculated faster, especially in
cases where there are a lot of input variables.</li>
<li>To allow both continuous and categorical variables to be included in
the PCA.</li>
<li>All of the above.</li>
</ul><ol start="2" style="list-style-type: decimal"><li>Can you think of datasets where it might not be necessary to
standardise variables?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal"><li>Scaling the data isn’t guaranteed to make the results more
interesting. It also won’t affect how quickly the output will be
calculated, whether continuous and categorical variables are present or
not.</li>
</ol><p>We use scaling to ensure that features with different ranges of
values have equal weighting in the resulting principal components (point
2).</p>
<ol start="2" style="list-style-type: decimal"><li>You may not want to standardise datasets which contain continuous
variables all measured on the same scale (e.g. gene expression data or
RNA sequencing data). In this case, variables with very little
sample-to-sample variability may represent only random noise, and
standardising the data would give these extra weight in the PCA.</li>
</ol></div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="performing-pca">Performing PCA<a class="anchor" aria-label="anchor" href="#performing-pca"></a></h3>
<p>Throughout this episode, we will carry out PCA using the Bioconductor
package <strong><code>PCAtools</code></strong>. This package provides
several functions that are useful for exploring data via PCA and
producing useful figures and analysis tools. The package is made for the
somewhat unusual Bioconductor style of data tables (observations in
columns, features in rows). When using Bioconductor data sets and
<strong><code>PCAtools</code></strong>, it is thus not necessary to
transpose the data. You can use the help files in the package
documentation to find out about the <code>pca()</code> function (type
<code>help("pca")</code> or <code>?pca</code> in R). Although we focus
on PCA implementation using this package, there are several other
options for performing PCA, including a base R function,
<code>prcomp()</code>. Equivalent functions and output labels for the
common implementations are summarised below.</p>
<table class="table"><thead><tr class="header"><th>library::command()</th>
<th>Principal component scores</th>
<th>Loadings</th>
</tr></thead><tbody><tr class="odd"><td>stats::prcomp()</td>
<td>$x</td>
<td>$rotation</td>
</tr><tr class="even"><td>stats::princomp()</td>
<td>$scores</td>
<td>$loadings</td>
</tr><tr class="odd"><td>PCAtools::pca()</td>
<td>$rotated</td>
<td>$loadings</td>
</tr></tbody></table><p>Let’s use the <strong><code>PCAtools</code></strong> package to
perform PCA on the scaled <code>prostate</code> data. First, let’s load
the <strong><code>PCAtools</code></strong> package.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"PCAtools"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: ggplot2</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: ggrepel</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'PCAtools'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    biplot, screeplot</code></pre>
</div>
<p>The input data (<code>pros2</code>) is in the form of a matrix and we
need to take the transpose of this matrix and convert it to a data frame
for use within the <strong><code>PCAtools</code></strong> package. We
set the <code>scale = TRUE</code> argument to standardise the variables
to have a mean 0 and standard deviation of 1 as above.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2t</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">pros2</span><span class="op">)</span><span class="op">)</span>  <span class="co"># create transposed data frame</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">pros2t</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">dim</span><span class="op">(</span><span class="va">pros2t</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># rename columns for use within PCAtools</span></span>
<span><span class="va">pmetadata</span><span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span><span class="st">"M"</span> <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">pros2t</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">pros2t</span><span class="op">)</span><span class="op">)</span> <span class="co"># create fake metadata for use with PCA tools</span></span>
<span></span>
<span><span class="co">## implement PCA</span></span>
<span><span class="va">pca.pros</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">pros2t</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, metadata <span class="op">=</span> <span class="va">pmetadata</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           Length Class      Mode
rotated     5     data.frame list
loadings    5     data.frame list
variance    5     -none-     numeric
sdev        5     -none-     numeric
metadata    1     data.frame list
xvars       5     -none-     character
yvars      97     -none-     character
components  5     -none-     character</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="how-many-principal-components-do-we-need">How many principal components do we need?<a class="anchor" aria-label="anchor" href="#how-many-principal-components-do-we-need"></a></h2>
<hr class="half-width"><p>We have calculated one principal component for each variable in the
original dataset. How do we choose how many of these are necessary to
represent the true variation in the data, without having extra
components that are unnecessary?</p>
<p>Let’s look at the relative importance of (percentage of the variance
explained by) the components:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prop.var</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">variance</span>, digits <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">prop.var</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    PC1     PC2     PC3     PC4     PC5
48.9767 27.3063 11.1094  8.0974  4.5101 </code></pre>
</div>
<p>This returns the proportion of variance in the data explained by each
of the principal components. In this example, the first principal
component, PC1, explains approximately 49% of variance in the data, PC2
27% of variance, PC3 a further 11%, PC4 approximately 8% and PC5 around
5%.</p>
<p>Let’s visualise this. A plot of the amount of variance accounted for
by each principal component is also called a scree plot. Often, scree
plots show a characteristic pattern where initially, the variance drops
rapidly with each additional principal component. But then there is an
“elbow” after which the variance decreases more slowly. The total
variance explained up to the elbow point is sometimes interpreted as
structural variance that is relevant and should be retained versus noise
which may be discarded after the elbow.</p>
<p>We can create a scree plot using the
<strong><code>PCAtools</code></strong> function
<code>screeplot()</code>:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">screeplot</span><span class="op">(</span><span class="va">pca.pros</span>, axisLabSize <span class="op">=</span> <span class="fl">5</span>, titleLabSize <span class="op">=</span> <span class="fl">8</span>, </span>
<span>          drawCumulativeSumLine <span class="op">=</span> <span class="cn">FALSE</span>, drawCumulativeSumPoints <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">components</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">variance</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-varexp-1.png" alt="A scree plot showing the percentage of variance explained by each principal component versus the principal component number. The points are joined by lines to indicate where the elbow of the scree plot occurs." class="figure mx-auto d-block"><figcaption>
Scree plot showing the percentage of the variance explained by the
principal components calculated from the prostate data.
</figcaption></figure><p>The scree plot shows that the first principal component explains most
of the variance in the data (&gt;50%) and each subsequent principal
component explains less and less of the total variance. The first two
principal components explain &gt;70% of variance in the data.</p>
<p>Selecting a subset of principal components results in loss of
information from the original dataset but selecting principal components
to summarise the <em>majority</em> of the information in the original
dataset is useful from a practical dimension reduction perspective.
There are no clear guidelines on how many principal components should be
included in PCA. We often look at the ‘elbow’ on the scree plot as an
indicator that the addition of principal components does not drastically
contribute to explaining the remaining variance. In this case, the
‘elbow’ of the scree plot appears to be at the third principal
component. We may therefore conclude that three principal components are
sufficient to explain the majority of the variability in the original
dataset. We may also choose another criterion, such as an arbitrary
cut-off for the proportion of the variance explained compared to other
principal components. Essentially, the criterion used to select
principal components should be determined based on what is deemed a
sufficient level of information retention for a specific dataset and
question.</p>
</section><section><h2 class="section-heading" id="loadings-and-principal-component-scores">Loadings and principal component scores<a class="anchor" aria-label="anchor" href="#loadings-and-principal-component-scores"></a></h2>
<hr class="half-width"><p>Most PCA functions will produce two main output matrices: the
<em>principal component scores</em> and the <em>loadings</em>. The
matrix of principal component scores has as many rows as there were
observations in the input matrix. These scores are what is usually
visualised or used for down-stream analyses. The matrix of loadings
(also called rotation matrix) has as many rows as there are features in
the original data. It contains information about how the (usually
centered and scaled) original data relate to the principal component
scores.</p>
<p>We can print the loadings for the
<strong><code>PCAtools</code></strong> implementation using</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca.pros</span><span class="op">$</span><span class="va">loadings</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              PC1         PC2         PC3         PC4         PC5
lcavol  0.5616465 -0.23664270  0.01486043  0.22708502 -0.75945046
lweight 0.2985223  0.60174151 -0.66320198 -0.32126853 -0.07577123
lbph    0.1681278  0.69638466  0.69313753  0.04517286 -0.06558369
lcp     0.4962203 -0.31092357  0.26309227 -0.72394666  0.25253840
lpsa    0.5665123 -0.01680231 -0.10141557  0.56487128  0.59111493</code></pre>
</div>
<p>The principal component scores are obtained by carrying out matrix
multiplication of the (usually centered and scaled) original data times
the loadings. The following callout demonstrates this.</p>
<div id="computing-a-pca-by-hand" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="computing-a-pca-by-hand" class="callout-inner">
<h3 class="callout-title">Computing a PCA “by hand”</h3>
<div class="callout-content">
<p>The rotation matrix obtained in a PCA is identical to the
eigenvectors of the covariance matrix of the data. Multiplying these
with the (centered and scaled) data yields the principal component
scores:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2.scaled</span> <span class="op">&lt;-</span> <span class="fu">scale</span><span class="op">(</span><span class="va">pros2</span><span class="op">)</span> <span class="co"># centre and scale the prostate data</span></span>
<span><span class="va">pros2.cov</span> <span class="op">&lt;-</span> <span class="fu">cov</span><span class="op">(</span><span class="va">pros2.scaled</span><span class="op">)</span>   <span class="co"># generate covariance matrix</span></span>
<span><span class="va">pros2.cov</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           lcavol   lweight         lbph          lcp      lpsa
lcavol  1.0000000 0.1941283  0.027349703  0.675310484 0.7344603
lweight 0.1941283 1.0000000  0.434934636  0.100237795 0.3541204
lbph    0.0273497 0.4349346  1.000000000 -0.006999431 0.1798094
lcp     0.6753105 0.1002378 -0.006999431  1.000000000 0.5488132
lpsa    0.7344603 0.3541204  0.179809410  0.548813169 1.0000000</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2.eigen</span> <span class="op">&lt;-</span> <span class="fu">eigen</span><span class="op">(</span><span class="va">pros2.cov</span><span class="op">)</span> <span class="co"># preform eigen decomposition</span></span>
<span><span class="va">pros2.eigen</span> <span class="co"># The slot $vectors = rotation of the PCA</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>eigen() decomposition
$values
[1] 2.4488355 1.3653171 0.5554705 0.4048702 0.2255067

$vectors
           [,1]        [,2]        [,3]        [,4]        [,5]
[1,] -0.5616465  0.23664270  0.01486043  0.22708502  0.75945046
[2,] -0.2985223 -0.60174151 -0.66320198 -0.32126853  0.07577123
[3,] -0.1681278 -0.69638466  0.69313753  0.04517286  0.06558369
[4,] -0.4962203  0.31092357  0.26309227 -0.72394666 -0.25253840
[5,] -0.5665123  0.01680231 -0.10141557  0.56487128 -0.59111493</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate principal component scores by by hand, using matrix multiplication</span></span>
<span><span class="va">my.pros2.pcs</span> <span class="op">&lt;-</span> <span class="va">pros2.scaled</span> <span class="op">%*%</span> <span class="va">pros2.eigen</span><span class="op">$</span><span class="va">vectors</span></span>
<span><span class="co"># compare results</span></span>
<span><span class="fu">par</span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">rotated</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, main<span class="op">=</span><span class="st">"PCAtools"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, v<span class="op">=</span><span class="fl">0</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">my.pros2.pcs</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, main<span class="op">=</span><span class="st">"\"By hand\""</span>, xlab<span class="op">=</span><span class="st">"PC1"</span>, ylab<span class="op">=</span><span class="st">"PC2"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, v<span class="op">=</span><span class="fl">0</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-pca-by-hand-1.png" alt="Side-by-side scatter plots of the second versus first principal components calculated by prcomp() (left) and by hand (right). The left scatter plot is the same as the right scatter plot but with swapped axes." class="figure mx-auto d-block"><figcaption>
Scatter plots of the second versus first principal components calculated
by prcomp() (left) and by hand (right).
</figcaption></figure><div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Note that the axis orientations may be swapped but the relative positions of the dots should be the same in both plots.</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>One way to visualise how principal components relate to the original
variables is by creating a biplot. Biplots usually show two principal
components plotted against each other. Observations are sometimes
labelled with numbers. The contribution of each original variable to the
principal components displayed is then shown by arrows (generated from
those two columns of the rotation matrix that correspond to the
principal components shown). See <code>help("PCAtools::biplot")</code>
for arguments and their meaning. For instance, <code>lab</code> or
<code>colBy</code> may be useful. Note that there are several biplot
implementations in different R libraries. It is thus a good idea to
specify the desired package when calling <code>biplot()</code>. A biplot
of the first two principal components can be generated as follows:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">PCAtools</span><span class="fu">::</span><span class="fu">biplot</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-stats-biplot-1.png" alt="Scatter plot of the second principal component versus the first principal component. Observations as points, numerically labelled." class="figure mx-auto d-block"><figcaption>
Scatter plot of the first two principal components with observations
numerically labelled.
</figcaption></figure><p>This biplot shows the position of each patient on a 2-dimensional
plot where loadings can be observed via the red arrows associated with
each of the variables. The variables <code>lpsa</code>,
<code>lcavol</code> and <code>lcp</code> are associated with positive
values on PC1 while positive values on PC2 are associated with the
variables <code>lbph</code> and <code>lweight</code>. The length of the
arrows indicates how much each variable contributes to the calculation
of each principal component.</p>
<p>The left and bottom axes show normalised principal component scores.
The axes on the top and right of the plot are used to interpret the
loadings, where loadings are scaled by the standard deviation of the
principal components (<code>pca.pros$sdev</code>) times the square root
the number of observations.</p>
</section><section><h2 class="section-heading" id="advantages-and-disadvantages-of-pca">Advantages and disadvantages of PCA<a class="anchor" aria-label="anchor" href="#advantages-and-disadvantages-of-pca"></a></h2>
<hr class="half-width"><p>Advantages:</p>
<ul><li>It is a relatively easy to use and popular method.</li>
<li>Various software/packages are available to run a PCA.</li>
<li>The calculations used in a PCA are simple to understand compared to
other methods for dimension reduction.</li>
</ul><p>Disadvantages:</p>
<ul><li>It assumes that variables in a dataset are correlated.</li>
<li>It is sensitive to the scale at which input variables are measured.
If input variables are measured at different scales, the variables with
large variance relative to the scale of measurement will have greater
impact on the principal components relative to variables with smaller
variance. In many cases, this is not desirable.</li>
<li>It is not robust against outliers, meaning that very large or small
data points can have a large effect on the output of the PCA.</li>
<li>PCA assumes a linear relationship between variables which is not
always a realistic assumption.</li>
<li>It can be difficult to interpret the meaning of the principal
components, especially when including them in further analyses
(e.g. inclusion in a linear regression).</li>
</ul></section><section><h2 class="section-heading" id="using-pca-to-analyse-gene-expression-data">Using PCA to analyse gene expression data<a class="anchor" aria-label="anchor" href="#using-pca-to-analyse-gene-expression-data"></a></h2>
<hr class="half-width"><p>The <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link">dataset</a>
we will be analysing in this lesson includes two subsets of data:</p>
<ul><li>a matrix of gene expression data showing microarray results for
different probes used to examine gene expression profiles in 91
different breast cancer patient samples.</li>
<li>metadata associated with the gene expression results detailing
information from patients from whom samples were taken.</li>
</ul><p>In this section, we will work through Challenges and you will apply
your own PCA to the high-dimensional gene expression data using what we
have learnt so far.</p>
<p>We will first load the microarray breast cancer gene expression data
and associated metadata, downloaded from the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2990" class="external-link">Gene
Expression Omnibus</a>.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: matrixStats</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: stats4</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: IRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: Biobase</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cancer</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/cancer_expression.rds"</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check that column names and row names match, if they do should return TRUE</span></span>
<span><span class="fu">all</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>The <code>mat</code> variable contains a matrix of gene expression
profiles for each sample. Rows represent gene expression measurements
and columns represent samples. The <code>metadata</code> variable
contains the metadata associated with the gene expression data including
the name of the study from which data originate, the age of the patient
from which the sample was taken, whether or not an oestrogen receptor
was involved in their cancer and the grade and size of the cancer for
each sample (represented by rows).</p>
<p>Microarray data are difficult to analyse for several reasons.
Firstly, formulating a research question using microarray data can be
difficult, especially if not much is known a priori about which genes
code for particular phenotypes of interest. Secondly, they are typically
high-dimensional and therefore are subject to the same difficulties
associated with analysing high-dimensional data. Finally, exploratory
analysis, which can be used to help formulate research questions and
display relationships, is difficult using microarray data due to the
number of potentially interesting response variables (i.e. expression
data from probes targeting different genes).</p>
<p>If researchers hypothesise that groups of genes (e.g. biological
pathways) may be associated with different phenotypic characteristics of
cancers (e.g. histologic grade, tumour size), using statistical methods
that reduce the number of columns in the microarray matrix to a smaller
number of dimensions representing groups of genes would help visualise
the data and address research questions regarding the effect different
groups of genes have on disease progression.</p>
<p>Using <strong><code>PCAtools</code></strong>, we will perform PCA on
the cancer gene expression data, plot the amount of variation in the
data explained by each principal component and plot the most important
principal components against each other as well as understanding what
each principal component represents.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3</h3>
<div class="callout-content">
<p>Apply a PCA to the cancer gene expression data using the
<code>pca()</code> function from
<strong><code>PCAtools</code></strong>.</p>
<p>Let us assume we only care about the principal components accounting
for the top 80% of the variance in the dataset. Use the
<code>removeVar</code> argument in <code>pca()</code> to remove the
principal components accounting for the bottom 20%.</p>
<p>As in the example using <code>prostate</code> data above, examine the
first 5 rows and columns of rotated data and loadings from your PCA.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span> <span class="co"># implement a PCA</span></span>
<span><span class="co"># We can check the scree plot to see that many principal components explain a very small amount of</span></span>
<span><span class="co"># the total variance in the data</span></span>
<span></span>
<span><span class="co"># Let's remove the principal components with lowest 20% of the variance</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span>, removeVar <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-- removing the lower 20% of variables based on variance</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Explore other arguments provided in pca</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">rotated</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># obtain the first 5 principal component scores for the first 5 observations</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               PC1        PC2        PC3        PC4        PC5
GSM65752 -29.79105  43.866788   3.255903 -40.663138 15.3427597
GSM65753 -37.33911 -15.244788  -4.948201  -6.182795  9.4725870
GSM65755 -29.41462   7.846858 -22.880525 -16.149669 22.3821009
GSM65757 -33.35286   1.343573 -22.579568   2.200329 15.0082786
GSM65758 -40.51897  -8.491125   5.288498  14.007364  0.8739772</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co"># obtain the first 5 principal component loadings for the first 5 features</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                    PC1          PC2          PC3        PC4          PC5
206378_at -0.0024680993 -0.053253543 -0.004068209 0.04068635  0.015078376
205916_at -0.0051557973  0.001315022 -0.009836545 0.03992371  0.038552048
206799_at  0.0005684075 -0.050657061 -0.009515725 0.02610233  0.006208078
205242_at  0.0130742288  0.028876408  0.007655420 0.04449641 -0.001061205
206509_at  0.0019031245 -0.054698479 -0.004667356 0.01566468  0.001306807</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># index of the maximal loading for the first principal component</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 49</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># principal component loadings for the feature</span></span>
<span><span class="co"># with the maximal loading on the first principal component:</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   PC1          PC2         PC3          PC4          PC5
215281_x_at 0.03752947 -0.007369379 0.006243377 -0.008242589 -0.004783206
                   PC6          PC7         PC8         PC9          PC10
215281_x_at 0.01194012 -0.002822407 -0.01216792 0.001137451 -0.0009056616
                   PC11          PC12       PC13         PC14          PC15
215281_x_at -0.00196034 -0.0001676705 0.00699201 -0.002897995 -0.0005044658
                     PC16        PC17         PC18        PC19         PC20
215281_x_at -0.0004547916 0.002277035 -0.006199078 0.002708574 -0.006217326
                  PC21        PC22        PC23        PC24        PC25
215281_x_at 0.00516745 0.007625912 0.003434534 0.005460017 0.001477415
                   PC26         PC27          PC28         PC29       PC30
215281_x_at 0.002350428 0.0007183107 -0.0006195515 0.0006349803 0.00413627
                    PC31        PC32         PC33         PC34         PC35
215281_x_at 0.0001322301 0.003182956 -0.002123462 -0.001042769 -0.001729869
                    PC36        PC37        PC38          PC39        PC40
215281_x_at -0.006556369 0.005766949 0.002537993 -0.0002846248 -0.00018195
                     PC41        PC42         PC43          PC44         PC45
215281_x_at -0.0007970789 0.003888626 -0.008210075 -0.0009570174 0.0007998935
                     PC46         PC47        PC48        PC49         PC50
215281_x_at -0.0006931441 -0.005717836 0.005189649 0.002591188 0.0007810259
                   PC51        PC52         PC53         PC54        PC55
215281_x_at 0.006610815 0.005371134 -0.001704796 -0.002286475 0.001365417
                   PC56         PC57        PC58         PC59         PC60
215281_x_at 0.003529892 0.0003375981 0.009895923 -0.001564423 -0.006989092
                   PC61        PC62         PC63          PC64        PC65
215281_x_at 0.000971273 0.001345406 -0.003575415 -0.0005588113 0.006516669
                    PC66        PC67       PC68         PC69        PC70
215281_x_at -0.008770186 0.006699641 0.01284606 -0.005041574 0.007845653
                   PC71        PC72         PC73         PC74        PC75
215281_x_at 0.003964697 -0.01104367 -0.001506485 -0.001583824 0.003798343
                   PC76         PC77         PC78         PC79          PC80
215281_x_at 0.004817252 -0.001290033 -0.004402926 -0.003440367 -0.0001646198
                   PC81        PC82          PC83         PC84        PC85
215281_x_at 0.003923775 0.003179556 -0.0004388192 9.664648e-05 0.003501335
                   PC86        PC87          PC88         PC89         PC90
215281_x_at -0.00112973 0.006489667 -0.0005039785 -0.004296355 -0.002751513
                   PC91
215281_x_at 0.009785973</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># index of the maximal loading for the second principal component</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 27</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># principal component loadings for the feature</span></span>
<span><span class="co"># with the maximal loading on the second principal component:</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   PC1        PC2          PC3        PC4          PC5
211122_s_at 0.01649085 0.05090275 -0.003378728 0.05178144 -0.003742393
                    PC6         PC7          PC8        PC9        PC10
211122_s_at -0.00543753 -0.03522848 -0.006333521 0.01575401 0.004732546
                   PC11        PC12        PC13        PC14       PC15
211122_s_at 0.004687599 -0.01349892 0.005207937 -0.01731898 0.02323893
                   PC16       PC17        PC18       PC19        PC20
211122_s_at -0.02069509 0.01477432 0.005658529 0.02667751 -0.01333503
                    PC21        PC22       PC23         PC24        PC25
211122_s_at -0.003254036 0.003572342 0.01416779 -0.005511838 -0.02582847
                  PC26        PC27       PC28        PC29       PC30      PC31
211122_s_at 0.03405417 -0.01797345 0.01826328 0.005123959 0.01300763 0.0127127
                   PC32       PC33       PC34        PC35        PC36
211122_s_at 0.002477672 0.01933214 0.03017661 -0.01935071 -0.01960912
                   PC37        PC38        PC39        PC40       PC41
211122_s_at 0.004411188 -0.01263612 -0.02019279 -0.01441513 -0.0310399
                   PC42         PC43        PC44        PC45        PC46
211122_s_at -0.02540426 0.0007949801 -0.00200195 -0.01748543 0.006881834
                   PC47         PC48        PC49         PC50        PC51
211122_s_at 0.006690698 -0.004000732 -0.02747926 -0.006963189 -0.02232332
                     PC52        PC53        PC54        PC55       PC56
211122_s_at -0.0003089115 -0.01604491 0.005649511 -0.02629501 0.02332997
                   PC57        PC58        PC59        PC60         PC61
211122_s_at -0.01248022 -0.01563245 0.005369433 0.009445262 -0.005209349
                  PC62       PC63       PC64        PC65        PC66
211122_s_at 0.01787645 0.01629425 0.02457665 -0.02384242 0.002814479
                    PC67        PC68         PC69         PC70       PC71
211122_s_at 0.0004584731 0.007939733 -0.009554166 -0.003967123 0.01825668
                   PC72        PC73        PC74        PC75        PC76
211122_s_at -0.00580374 -0.02236727 0.001295688 -0.02264723 0.006855855
                   PC77         PC78       PC79         PC80        PC81
211122_s_at 0.004995447 -0.008404118 0.00442875 -0.001027912 0.006104406
                   PC82        PC83         PC84       PC85       PC86
211122_s_at -0.01988441 0.009667348 -0.008248781 0.01198369 0.01221713
                    PC87        PC88        PC89        PC90       PC91
211122_s_at -0.003864842 -0.02876816 -0.01771452 -0.02164973 0.01124148</code></pre>
</div>
<p>The function <code>pca()</code> is used to perform PCA, and uses as
input a matrix (<code>mat</code>) containing continuous numerical data
in which rows are data variables and columns are samples, and
<code>metadata</code> associated with the matrix in which rows represent
samples and columns represent data variables. It has options to centre
or scale the input data before a PCA is performed, although in this case
gene expression data do not need to be transformed prior to PCA being
carried out as variables are measured on a similar scale (values are
comparable between rows). The output of the <code>pca()</code> function
includes a lot of information such as loading values for each variable
(<code>loadings</code>), principal component scores
(<code>rotated</code>) and the amount of variance in the data explained
by each principal component.</p>
<p>Rotated data shows principal component scores for each sample and
each principal component. Loadings the contribution each variable makes
to each principal component.</p>
</div>
</div>
</div>
</div>
<div id="scaling-variables-for-pca" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="scaling-variables-for-pca" class="callout-inner">
<h3 class="callout-title">Scaling variables for PCA</h3>
<div class="callout-content">
<p>When running <code>pca()</code> above, we kept the default setting,
<code>scale=FALSE</code>. That means genes with higher variation in
their expression levels should have higher loadings, which is what we
are interested in. Whether or not to scale variables for PCA will depend
on your data and research question.</p>
<p>Note that this is different from normalising gene expression data.
Gene expression data have to be normalised before donwstream analyses
can be carried out. This is to reduce technical and other potentially
confounding factors. We assume that the expression data we use had been
normalised previously.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="choosing-how-many-components-are-important-to-explain-the-variance-in-the-data">Choosing how many components are important to explain the variance
in the data<a class="anchor" aria-label="anchor" href="#choosing-how-many-components-are-important-to-explain-the-variance-in-the-data"></a></h2>
<hr class="half-width"><p>As in the example using the <code>prostate</code> dataset, we can use
a scree plot to compare the proportion of variance in the data explained
by each principal component. This allows us to understand how much
information in the microarray dataset is lost by projecting the
observations onto the first few principal components and whether these
principal components represent a reasonable amount of the variation. The
proportion of variance explained should sum to one.</p>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4</h3>
<div class="callout-content">
<p>This time using the <code>screeplot()</code> function in
<strong><code>PCAtools</code></strong>, create a scree plot to show
proportion of variance explained by the first 20 principal component
(hint: <code>components = 1:20</code>). Explain the output of the scree
plot in terms of proportion of the variance in the data explained by
each principal component.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="fu">screeplot</span><span class="op">(</span><span class="va">pc</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">80</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Scale for y is already present.
Adding another scale for y, which will replace the existing scale.</code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-scree-ex-1.png" alt="A bar and line plot showing the variance explained by principal components (PCs) of gene expression data. Blue bars depict the variance explained by each PC, while a red line depicts the cumulative variance explained by the PCs. The first principal component explains roughly 30% of the variance, while succeeding PCs explain less than 10%." class="figure mx-auto d-block"><figcaption>
A scree plot of the gene expression data.
</figcaption></figure><p>The first principal component explains around 33% of the variance in
the microarray data, the first 4 principal components explain around
50%, and 20 principal components explain around 75%. Many principal
components explain very little variation. A first ‘elbow’ appears to be
around 4-5 principal components, indicating that this may be a suitable
number of principal components. However, these principal components
cumulatively explain only 51-55% of the variance in the dataset.
Although the fact we are able to summarise most of the information in
the complex dataset in 4-5 principal components may be a useful result,
we may opt to retain more principal components (for example, 20) to
capture more of the variability in the dataset depending on research
question. A second ‘elbow’ around 12 principal components may provide a
good middleground. Note that first principal component (PC1) explains
more variation than other principal components (which is always the case
in PCA). The scree plot shows that the first principal component only
explains ~33% of the total variation in the microarray data and many
principal components explain very little variation. The red line shows
the cumulative percentage of explained variation with increasing
principal components.</p>
<p>Note that in this case 18 principal components are needed to explain
over 75% of variation in the data. This is not an unusual result for
complex biological datasets including genetic information as clear
relationships between groups are sometimes difficult to observe in the
data. The scree plot shows that using a PCA we have reduced 91
predictors to 18 in order to explain a significant amount of variation
in the data. See additional arguments in scree plot function for
improving the appearance of the plot.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="investigating-the-principal-components">Investigating the principal components<a class="anchor" aria-label="anchor" href="#investigating-the-principal-components"></a></h2>
<hr class="half-width"><p>Once the most important principal components have been identified
using <code>screeplot()</code> and explored, these can be explored in
more detail by plotting principal components against each other and
highlighting points based on variables in the metadata. This will allow
any potential clustering of points according to demographic or
phenotypic variables to be seen.</p>
<p>We can use biplots to look for patterns in the output from the PCA.
Note that there are two functions called <code>biplot()</code>, one in
the package <strong><code>PCAtools</code></strong> and one in
<strong><code>stats</code></strong>. Both functions produce biplots but
their scales are different!</p>
<div id="challenge-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-5" class="callout-inner">
<h3 class="callout-title">Challenge 5</h3>
<div class="callout-content">
<p>Create a biplot of the first two principal components from your PCA
using <code>biplot()</code> function in
<strong><code>PCAtools</code></strong>. See
<code>help("PCAtools::biplot")</code> for arguments and their meaning.
For instance, <code>lab</code> or <code>colby</code> may be useful.</p>
<p>Examine whether the data appear to form clusters. Explain your
results.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">PCAtools</span><span class="fu">::</span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>, lab <span class="op">=</span> <span class="cn">NULL</span>, colby <span class="op">=</span> <span class="st">'Grade'</span>, legendPosition <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-biplot-ex-1.png" alt="A biplot of PC2 against PC1 in the gene expression data, coloured by Grade. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on Grade across these two groups." class="figure mx-auto d-block"><figcaption>
A biplot of PC2 against PC1 in the gene expression data, coloured by
Grade.
</figcaption></figure><p>The biplot shows the position of patient samples relative to PC1 and
PC2 in a 2-dimensional plot. Note that two groups are apparent along the
PC1 axis according to expressions of different genes while no separation
can be seem along the PC2 axis. Labels of patient samples are
automatically added in the biplot. Labels for each sample are added by
default, but can be removed if there is too much overlap in names. Note
that <strong><code>PCAtools</code></strong> does not scale biplot in the
same way as biplot using the <strong><code>stats</code></strong>
package.</p>
</div>
</div>
</div>
</div>
<p>Let’s consider this biplot in more detail, and also display the
loadings:</p>
<div id="challenge-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-6" class="callout-inner">
<h3 class="callout-title">Challenge 6</h3>
<div class="callout-content">
<p>Use <code>colby</code> and <code>lab</code> arguments in
<code>biplot()</code> to explore whether these two groups may cluster by
patient age or by whether or not the sample expresses the oestrogen
receptor gene (ER+ or ER-).</p>
<p>Note: You may see a warning about
<strong><code>ggrepel</code></strong>. This happens when there are many
labels but little space for plotting. This is not usually a serious
problem - not all labels will be shown.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb76">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu">PCAtools</span><span class="fu">::</span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>,</span>
<span>    lab <span class="op">=</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">Age</span>,<span class="st">'years'</span><span class="op">)</span>,</span>
<span>    colby <span class="op">=</span> <span class="st">'ER'</span>,</span>
<span>    legendPosition <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: ggrepel: 23 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-pca-biplot-ex2-1.png" alt="A biplot of PC2 against PC1 in the gene expression data, coloured by ER status. The points on the scatter plot separate clearly on PC1, but there is no clear grouping of samples based on ER across these two groups, although there are more ER- samples in the rightmost cluster. Patient ages are overlaid as text near the points, but there is again no clear pattern." class="figure mx-auto d-block"><figcaption>
A biplot of PC2 against PC1 in the gene expression data, coloured by ER
status with patient ages overlaid.
</figcaption></figure><p>It appears that one cluster has more ER+ samples than the other
group.</p>
</div>
</div>
</div>
</div>
<p>So far, we have only looked at a biplot of PC1 versus PC2 which only
gives part of the picture. The <code>pairplots()</code> function in
<strong><code>PCAtools</code></strong> can be used to create multiple
biplots including different principal components.</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pairsplot</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.
Coordinate system already present. Adding new coordinate system, which will
replace the existing one.</code></pre>
</div>
<figure style="text-align: center"><img src="../fig/04-principal-component-analysis-rendered-pairsplot-1.png" alt="A triangular grid of scatter plots. The grid is the upper right triangle of a square, where each entry of the grid corresponds to a plot of one principal component against another. For example, the plot in the upper left corner of the plot corresponds to principal component one plotted against principal component 2, and the plot to the right of this plots principal component 1 against principal component 3. Points correspond to samples, and are coloured arbitrarily from light blue to dark blue." class="figure mx-auto d-block"><figcaption>
Multiple biplots produced by pairsplot().
</figcaption></figure><p>The plots show two apparent clusters involving the first principal
component only. No other clusters are found involving other principal
components. Each dot is coloured differently along a gradient of blues.
This can potentially help identifying the same observation/individual in
several panels. Here too, the argument <code>colby</code> allows you to
set custom colours.</p>
<p>Finally, it can sometimes be of interest to compare how certain
variables contribute to different principal components. This can be
visualised with <code>plotloadings()</code> from the
<strong><code>PCAtools</code></strong> package. The function checks the
range of loadings for each principal component specified (default: first
five principal components). It then selects the features in the top and
bottom 5% of these ranges and displays their loadings. This behaviour
can be adjusted with the <code>rangeRetain</code> argument, which has
0.1 as the default value (i.e. 5% on each end of the range). NB, if
there are too many labels to be plotted, you will see a warning. This is
not a serious problem.</p>
<p><code>{r loadingsplots, fig.cap=c("Plot of the features in the top and bottom 5 % of the loadings range for the first principal component.", "Plot of the features in the top and bottom 5 % of the loadings range for the second principal component.", fig.cap = "Plot of the features in the top and bottom 5 % of the loadings range for the either the first or second principal components."), fig.alt=c("A plot with component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the first principal component. The component loading values for four features are shown. The features with the highest loadings are shown at the top of the plot and consist of three features, each with a blue dot and feature label. The feature with the lowest loading is shown at the bottom of the plot and delineated by a yellow dot with a feature label.",      "A plot with component loadings versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the second principal component. The component loading values for 9 features are shown. The features with the highest loadings are shown at the top of the plot and consist of 6 features, each with a blue dot and feature label. The 3 features at the lower range are at the bottom of the plot and delineated by a yellow dot with a feature label.",      fig.alt = "A plot with component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range for either principal component. The component loading values for several features are shown with loading score-delineated colours ranging from dark blue for the highest loadings and dark yellow for the lowest loadings. Each point has a feature label. The loadings of features points for the first principal component are concentrated towards the top, while the loadings of features points for the second principal components are concentrated at the top, middle and bottom of the range.")} plotloadings(pc, c("PC1"), rangeRetain = 0.1) plotloadings(pc, c("PC2"), rangeRetain = 0.1) plotloadings(pc, c("PC1", "PC2"), rangeRetain = 0.1)</code></p>
<p>You can see how the third code line produces more dots, some of which
do not have extreme loadings. This is because all loadings selected for
any principal component are shown for all other principal components.
For instance, it is plausible that features which have high loadings on
PC1 may have lower ones on PC2.</p>
</section><section><h2 class="section-heading" id="using-pca-output-in-further-analysis">Using PCA output in further analysis<a class="anchor" aria-label="anchor" href="#using-pca-output-in-further-analysis"></a></h2>
<hr class="half-width"><p>The output of PCA can be used to interpret data or can be used in
further analyses. For example, the PCA outputs new variables (principal
components) which represent several variables in the original dataset.
These new variables are useful for further exploring data, for example,
comparing principal component scores between groups or including the new
variables in linear regressions. Because the principal components are
uncorrelated (and independent) they can be included together in a single
linear regression.</p>
<div id="principal-component-regression" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="principal-component-regression" class="callout-inner">
<h3 class="callout-title">Principal component regression</h3>
<div class="callout-content">
<p>PCA is often used to reduce large numbers of correlated variables
into fewer uncorrelated variables that can then be included in linear
regression or other models. This technique is called principal component
regression (PCR) and it allows researchers to examine the effect of
several correlated explanatory variables on a single response variable
in cases where a high degree of correlation initially prevents them from
being included in the same model. Principal component regression (PCR)
is just one example of how principal components can be used in further
analysis of data. When carrying out PCR, the variable of interest
(response/dependent variable) is regressed against the principal
components calculated using PCA, rather than against each individual
explanatory variable from the original dataset. As there as many
principal components created from PCA as there are variables in the
dataset, we must select which principal components to include in PCR.
This can be done by examining the amount of variation in the data
explained by each principal component (see above).</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a></h2>
<hr class="half-width"><ul><li>James, G., Witten, D., Hastie, T. &amp; Tibshirani, R. (2013) An
Introduction to Statistical Learning with Applications in R. Chapter 6.3
(Dimension Reduction Methods), Chapter 10 (Unsupervised Learning).</li>
<li>
<a href="https://dx.doi.org/10.1098/rsta.2015.0202" class="external-link">Jolliffe, I.T.
&amp; Cadima, J. (2016) Principal component analysis: a review and
recent developments. Phil. Trans. R. Soc A 374.</a>.</li>
<li><a href="https://doi.org/10.1098/rsta.2009.0159" class="external-link">Johnstone, I.M.
&amp; Titterington, D.M. (2009) Statistical challenges of
high-dimensional data. Phil. Trans. R. Soc A 367.</a></li>
<li>
<a href="https://www.analyticsvidhya.com/blog/2016/03/pca-practical-guide-principal-component-analysis-python/" class="external-link">PCA:
A Practical Guide to Principal Component Analysis, Analytics
Vidhya</a>.</li>
<li>
<a href="https://towardsdatascience.com/a-one-stop-shop-for-principal-component-analysis-5582fb7e0a9c" class="external-link">A
One-Stop Shop for Principal Component Analysis, Towards Data
Science</a>.</li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>A principal component analysis is a statistical approach used to
reduce dimensionality in high-dimensional datasets (i.e. where <span class="math inline">\(p\)</span> is equal or greater than <span class="math inline">\(n\)</span>).</li>
<li>PCA may be used to create a low-dimensional set of features from a
larger set of variables. Examples of when a PCA may be useful include
reducing high-dimensional datasets to fewer variables for use in a
linear regression and for identifying groups with similar features.</li>
<li>PCA is a useful dimensionality reduction technique used in the
analysis of complex biological datasets (e.g. high throughput data or
genetics data).</li>
<li>The first principal component represents the dimension along which
there is maximum variation in the data. Subsequent principal components
represent dimensions with progressively less variation.</li>
<li>Scree plots and biplots may be used to show: 1. how much variation
in the data is explained by each principal component and 2. how data
points cluster according to principal component scores and which
variables are associated with these scores.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/03-regression-regularisation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/05-factor-analysis.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/03-regression-regularisation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Regularised
        </a>
        <a class="chapter-link float-end" href="../instructor/05-factor-analysis.html" rel="next">
          Next: Factor analysis...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/04-principal-component-analysis.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/04-principal-component-analysis.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Principal component analysis",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/04-principal-component-analysis.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/04-principal-component-analysis.html",
  "dateCreated": "2025-02-03",
  "dateModified": "2025-02-03",
  "datePublished": "2025-02-11"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

