<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>High dimensional statistics with R: K-means</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../06-k-means.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      High dimensional statistics with R
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            High dimensional statistics with R
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="data.html">Data</a></li><hr></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  High dimensional statistics with R
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 75%" class="percentage">
    75%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 75%" aria-valuenow="75" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../06-k-means.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction-to-high-dimensional-data.html">1. Introduction to high-dimensional data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-high-dimensional-regression.html">2. Regression with many outcomes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-regression-regularisation.html">3. Regularised regression</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-principal-component-analysis.html">4. Principal component analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-factor-analysis.html">5. Factor analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. K-means
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#what-is-k-means-clustering">What is K-means clustering?</a></li>
<li><a href="#believing-in-clusters">Believing in clusters</a></li>
<li><a href="#k-means-clustering-applied-to-the-single-cell-rna-sequencing-data">K-means clustering applied to the single-cell RNA sequencing
data</a></li>
<li><a href="#cluster-separation">Cluster separation</a></li>
<li><a href="#cluster-robustness-bootstrapping">Cluster robustness: bootstrapping</a></li>
<li><a href="#bootstrapping">Bootstrapping</a></li>
<li><a href="#further-reading">Further reading</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-hierarchical.html">7. Hierarchical clustering</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="data.html">Data</a></li><hr></ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/05-factor-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/07-hierarchical.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/05-factor-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Factor analysis
        </a>
        <a class="chapter-link float-end" href="../instructor/07-hierarchical.html" rel="next">
          Next: Hierarchical...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>K-means</h1>
        <p>Last updated on 2025-09-02 |

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/06-k-means.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 80 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do we detect real clusters in high-dimensional data?</li>
<li>How does K-means work and when should it be used?</li>
<li>How can we perform K-means in <code>R</code>?</li>
<li>How can we appraise a clustering and test cluster robustness?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand what clusters look like in in high-dimensional data.</li>
<li>Understand and perform K-means clustering in <code>R</code>.</li>
<li>Assess clustering performance using silhouette scores.</li>
<li>Assess cluster robustness using bootstrapping.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h2>
<hr class="half-width"><p>As we saw in previous episodes, visualising high-dimensional data
with a large amount of features is difficult and can limit our
understanding of the data and associated processes. In some cases, a
known grouping causes this heterogeneity (sex, treatment groups, etc).
In other cases, heterogeneity may arise from the presence of unknown
subgroups in the data.</p>
<p>While PCA can be used to reduce the dimension of the dataset into a
smaller set of uncorrelated variables and factor analysis can be used to
identify underlying factors, clustering is a set of techniques that
allow us to discover unknown groupings.</p>
<p>Cluster analysis involves finding groups of observations that are
more similar to each other (according to some feature) than they are to
observations in other groups and are thus likely to represent the same
source of heterogeneity. Once groups (or clusters) of observations have
been identified using cluster analysis, further analyses or
interpretation can be carried out on the groups, for example, using
metadata to further explore groups.</p>
<p>Cluster analysis is commonly used to discover unknown groupings in
fields such as bioinformatics, genomics, and image processing, in which
large datasets that include many features are often produced.</p>
<p>There are various ways to look for clusters of observations in a
dataset using different <em>clustering algorithms</em>. One way of
clustering data is to minimise distance between observations within a
cluster and maximise distance between proposed clusters. Using this
process, we can also iteratively update clusters so that we become more
confident about the shape and size of the clusters.</p>
</section><section><h2 class="section-heading" id="what-is-k-means-clustering">What is K-means clustering?<a class="anchor" aria-label="anchor" href="#what-is-k-means-clustering"></a></h2>
<hr class="half-width"><p><strong>K-means clustering</strong> groups data points into a
user-defined number of distinct, non-overlapping clusters. To create
clusters of ‘similar’ data points, K-means clustering creates clusters
that minimise the within-cluster variation and thus the amount that data
points within a cluster differ from each other. The distance between
data points within a cluster is used as a measure of within-cluster
variation.</p>
<p>To carry out K-means clustering, we first pick <span class="math inline">\(k\)</span> initial points as centres or
“centroids” of our clusters. There are a few ways to choose these
initial “centroids” and this is discussed below. Once we have picked
intitial points, we then follow these two steps until appropriate
clusters have been formed:</p>
<ol style="list-style-type: decimal"><li>Assign each data point to the cluster with the closest centroid</li>
<li>Update centroid positions as the average of the points in that
cluster.</li>
</ol><p>We can see this process in action in this animation:</p>
<figure style="text-align: center"><img src="../fig/kmeans.gif" alt="An animated scatter plot of data y versus x. The animation starts by identifying 3 initial points, delineated by different colours. The animations then colour codes all points by an associated cluster colour, delineating three distinct and non-overlapping clusters in the space of the scatter plot." class="figure mx-auto d-block"><figcaption>
Animation showing the iterative process of K-means clustering on data y
versus x.
</figcaption></figure><p>While K-means has some advantages over other clustering methods (easy
to implement and to understand), it does have some disadvantages,
particularly difficulties in identifying initial clusters which
observations belong to and the need for the user to specify the number
of clusters that the data should be partitioned into.</p>
<div id="initialisation" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="initialisation" class="callout-inner">
<h3 class="callout-title">Initialisation</h3>
<div class="callout-content">
<p>The algorithm used in K-means clustering finds a <em>local</em>
rather than a <em>global</em> optimum, so that results of clustering are
dependent on the initial cluster that each observation is randomly
assigned to. This initial configuration can have a significant effect on
the final configuration of the clusters, so dealing with this limitation
is an important part of K-means clustering. Some strategies to deal with
this problem are:</p>
<ul><li>Choose <span class="math inline">\(K\)</span> points at random from
the data as the cluster centroids.</li>
<li>Randomly split the data into <span class="math inline">\(K\)</span>
groups, and then average these groups.</li>
<li>Use the K-means++ algorithm to choose initial values.</li>
</ul><p>These each have advantages and disadvantages. In general, it’s good
to be aware of this limitation of K-means clustering and that this
limitation can be addressed by choosing a good initialisation method,
initialising clusters manually, or running the algorithm from multiple
different starting points.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="believing-in-clusters">Believing in clusters<a class="anchor" aria-label="anchor" href="#believing-in-clusters"></a></h2>
<hr class="half-width"><p>When using clustering, it’s important to realise that data may seem
to group together even when these groups are created randomly. It’s
especially important to remember this when making plots that add extra
visual aids to distinguish clusters.</p>
<p>For example, if we cluster data from a single 2D normal distribution
and draw ellipses around the points, these clusters suddenly become
almost visually convincing. This is a somewhat extreme example, since
there is genuinely no heterogeneity in the data, but it does reflect
what can happen if you allow yourself to read too much into faint
signals.</p>
<p>Let’s explore this further using an example. We create two columns of
data (‘x’ and ‘y’) and partition these data into three groups (‘a’, ‘b’,
‘c’) according to data values. We then plot these data and their
allocated clusters and put ellipses around the clusters using the
<code>stat_ellipse()</code> function in
<strong><code>ggplot</code></strong>.</p>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-fake-cluster-1.png" alt="A scatter plot of random data y versus x. The points are horizontally partitioned at 2 random groups, forming three colour coded clusters. Circles are drawn around each cluster. The data shown appears to have no clusters but the colours and circles give the appearance of clusters artificially." class="figure mx-auto d-block"><figcaption>
Example of artificial clusters fitted to data points.
</figcaption></figure><p>The randomly created data used here appear to form three clusters
when we plot the data. Putting ellipses around the clusters can further
convince us that the clusters are ‘real’. But how do we tell if clusters
identified visually are ‘real’?</p>
<div id="initialisation-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="initialisation-1" class="callout-inner">
<h3 class="callout-title">Initialisation</h3>
<div class="callout-content">
<p>The algorithm used in K-means clustering finds a <em>local</em>
rather than a <em>global</em> optimum, so that results of clustering are
dependent on the initial cluster that each observation is randomly
assigned to. This initial configuration can have a significant effect on
the final configuration of the clusters, so dealing with this limitation
is an important part of K-means clustering. Some strategies to deal with
this problem are:</p>
<ul><li>Choose <span class="math inline">\(k\)</span> points at random from
the data as the cluster centroids.</li>
<li>Randomly split the data into <span class="math inline">\(k\)</span>
groups, and then average these groups.</li>
<li>Use the K-means++ algorithm to choose initial values.</li>
</ul><p>These each have advantages and disadvantages. In general, it’s good
to be aware of this limitation of K-means clustering and that this
limitation can be addressed by choosing a good initialisation method,
initialising clusters manually, or running the algorithm from multiple
different starting points.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="k-means-clustering-applied-to-the-single-cell-rna-sequencing-data">K-means clustering applied to the single-cell RNA sequencing
data<a class="anchor" aria-label="anchor" href="#k-means-clustering-applied-to-the-single-cell-rna-sequencing-data"></a></h2>
<hr class="half-width"><p>Let’s carry out K-means clustering in <code>R</code> using some real
high-dimensional data. We’re going to work with single-cell RNA
sequencing data, <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/data/index.html" class="external-link"><em>scRNAseq</em></a>,
in these clustering challenges, which is often <em>very</em>
high-dimensional. Commonly, experiments profile the expression level of
10,000+ genes in thousands of cells. Even after filtering the data to
remove low quality observations, the dataset we’re using in this episode
contains measurements for over 9,000 genes in over 3,000 cells.</p>
<p>One way to get a handle on a dataset of this size is to use something
we covered earlier in the course - dimensionality reduction.
Dimensionality reduction allows us to visualise this incredibly complex
data in a small number of dimensions. In this case, we’ll be using
principal component analysis (PCA) to compress the data by identifying
the major axes of variation in the data, before running our clustering
algorithms on this lower-dimensional data to explore the similarity of
features.</p>
<p>The <strong><code>scater</code></strong> package has some easy-to-use
tools to calculate a PCA for <code>SummarizedExperiment</code> objects.
Let’s load the <em>scRNAseq</em> data and calculate some principal
components.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: matrixStats</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: stats4</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: IRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: Biobase</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: scuttle</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scrnaseq</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/scrnaseq.rds"</span><span class="op">)</span></span>
<span><span class="va">scrnaseq</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">scrnaseq</span>, ncomponents <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">scrnaseq</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<p>The first two principal components capture almost 50% of the
variation within the data. For now, we’ll work with just these two
principal components, since we can visualise those easily, and they’re a
quantitative representation of the underlying data, representing the two
largest axes of variation.</p>
<p>We can now run K-means clustering on the first and second principal
components of the <em>scRNAseq</em> data using the <code>kmeans()</code>
function.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="va">pcs</span>, centers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">scrnaseq</span><span class="op">$</span><span class="va">kmeans</span> <span class="op">&lt;-</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">scrnaseq</span>, <span class="st">"PCA"</span>, colour_by <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-kmeans-1.png" alt="A scatter plot of principal component 2 versus principal component 1 of the scrnaseq data. Each point is one of four colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing four distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 versus principal component 1 with
points colour coded according to the cluster to which they belong.
</figcaption></figure><p>We can see that this produces a sensible-looking partition of the
data. However, is it totally clear whether there might be more or fewer
clusters here?</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1</h3>
<div class="callout-content">
<p>Perform clustering to group the data into <span class="math inline">\(k=5\)</span> clusters, and plot it using
<code>plotReducedDim()</code>. Save this with a variable name that’s
different to what we just used, because we’ll use this again later.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">cluster5</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="va">pcs</span>, centers <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">scrnaseq</span><span class="op">$</span><span class="va">kmeans5</span> <span class="op">&lt;-</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">cluster5</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">scrnaseq</span>, <span class="st">"PCA"</span>, colour_by <span class="op">=</span> <span class="st">"kmeans5"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-kmeans-ex-1.png" alt="A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 against principal component 1 in
the scRNAseq data, coloured by clusters produced by k-means clustering.
</figcaption></figure></div>
</div>
</div>
</div>
<div id="k-medoids-pam" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="k-medoids-pam" class="callout-inner">
<h3 class="callout-title">K-medoids (PAM)</h3>
<div class="callout-content">
<p>One problem with K-means is that using the mean to define cluster
centroids means that clusters can be very sensitive to outlying
observations.</p>
<p>K-medoids, also known as “partitioning around medoids (PAM)” is
similar to K-means, but uses the median rather than the mean as the
method for defining cluster centroids. Using the median rather than the
mean reduces sensitivity of clusters to outliers in the data.</p>
<p>K-medioids has had popular application in genomics, for example the
well-known PAM50 gene set in breast cancer, which has seen some
prognostic applications.</p>
<p>The following example shows how cluster centroids differ when created
using medians rather than means.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span> <span class="op">+</span> <span class="fl">10</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="fu">median</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu">median</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-unnamed-chunk-1-1.png" alt="Scatter plot of random data y versus x. There are many black points on the plot representing the data. Two additional points are shown: the (mean(x), mean(y)) co-ordinate point in red and the (median(x), median(y)) co-ordinate point in blue. The median co-ordinate point in blue has a lower x value and is shown to the left of the red mean co-ordinate point." class="figure mx-auto d-block"><figcaption>
Scatter plot of random data y versus x with the (mean(x), mean(y))
co-ordinate point shown in red and the (median(x), median(y))
co-ordinate point shown in blue.
</figcaption></figure><p>PAM can be carried out using <code>pam()</code> from the
<strong><code>cluster</code></strong> package.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="cluster-separation">Cluster separation<a class="anchor" aria-label="anchor" href="#cluster-separation"></a></h2>
<hr class="half-width"><p>When performing clustering, it is important for us to be able to
measure how well our clusters are separated. One measure to test this is
silhouette width, which measures how similar a data point is to points
in the same cluster compared to other clusters.</p>
<p>The silhouette width is computed for every observation and can range
from -1 to 1. A high silhouette width means an observation is closer to
other observations within the same cluster. For each cluster, the
silhouette widths can then be averaged or an overall average can be
taken.</p>
<div id="more-detail-on-silhouette-widths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="more-detail-on-silhouette-widths" class="callout-inner">
<h3 class="callout-title">More detail on silhouette widths</h3>
<div class="callout-content">
<p>In more detail, each observation’s silhouette width is computed as
follows:</p>
<ol style="list-style-type: decimal"><li>Compute the average distance between the focal observation and all
other observations in the same cluster.</li>
<li>For each of the other clusters, compute the average distance between
focal observation and all observations in the other cluster. Keep the
smallest of these average distances.</li>
<li>Subtract (1.)-(2.) then divivde by whichever is smaller (1.) or
(2).</li>
</ol></div>
</div>
</div>
<p>Ideally, we would have only large positive silhouette widths,
indicating that each data point is much more similar to points within
its cluster than it is to the points in any other cluster. However, this
is rarely the case. Often, clusters are very fuzzy, and even if we are
relatively sure about the existence of discrete groupings in the data,
observations on the boundaries can be difficult to confidently place in
either cluster.</p>
<p>Here we use the <code>silhouette()</code> function from the
<strong><code>cluster</code></strong> package to calculate the
silhouette width of our K-means clustering using a distance matrix of
distances between points in the clusters.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"cluster"</span><span class="op">)</span></span>
<span><span class="va">dist_mat</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu">silhouette</span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span>, dist <span class="op">=</span> <span class="va">dist_mat</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">sil</span>, border <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-silhouette-1.png" alt="Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains over half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have slightly lower silhouette widths. The bars therefore reach further to the left than cluster 4." class="figure mx-auto d-block"><figcaption>
Silhouette plot for each point according to cluster.
</figcaption></figure><p>Let’s plot the silhouette score on the original dimensions used to
cluster the data. Here, we’re mapping cluster membership to point shape,
and silhouette width to colour.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">sil</span> <span class="op">&lt;-</span> <span class="va">sil</span><span class="op">[</span>, <span class="st">"sil_width"</span><span class="op">]</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">sil</span><span class="op">[</span>, <span class="st">"sil_width"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.7065662</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">aes</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, shape <span class="op">=</span> <span class="va">clust</span>, colour <span class="op">=</span> <span class="va">sil</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_colour_gradient2</span><span class="op">(</span></span>
<span>        low <span class="op">=</span> <span class="st">"dodgerblue"</span>, high <span class="op">=</span> <span class="st">"firebrick"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_shape_manual</span><span class="op">(</span></span>
<span>        values <span class="op">=</span> <span class="fu">setNames</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-plot-silhouette-1.png" alt="A scatter plot of the random y versus x data. Cluster membership is delineated using different point characters. Data points in the same cluster have the same point character. Each point is coloured by its silhouette width: solid red delineating a silhouette width of 1 and white delineating a silhouette width of 0. Colours in between delineate the intermediate colours. Many points are red and fade to white at the boundaries of each cluster. " class="figure mx-auto d-block"><figcaption>
Scatter plot of y versus x coloured according to silhouette width and
point characters grouped according to cluster membership.
</figcaption></figure><p>This plot shows that silhouette values for individual observations
tends to be very high in the centre of clusters, but becomes quite low
towards the edges. This makes sense, as points that are “between” two
clusters may be more similar to points in another cluster than they are
to the points in the cluster one they belong to.</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2</h3>
<div class="callout-content">
<p>Calculate the silhouette width for the k of 5 clustering we did
earlier. Are 5 clusters appropriate? Why/why not?</p>
<p>Can you identify where the differences lie?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sil5</span> <span class="op">&lt;-</span> <span class="fu">silhouette</span><span class="op">(</span><span class="va">cluster5</span><span class="op">$</span><span class="va">cluster</span>, dist <span class="op">=</span> <span class="va">dist_mat</span><span class="op">)</span></span>
<span><span class="va">scrnaseq</span><span class="op">$</span><span class="va">kmeans5</span> <span class="op">&lt;-</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">cluster5</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">scrnaseq</span>, <span class="st">"PCA"</span>, colour_by <span class="op">=</span> <span class="st">"kmeans5"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-silhouette-ex-1.png" alt="A scatter plot of principal component 1 versus principal component 2 of the scrnaseq data. Each point is one of five colours, representing cluster membership. Points of the same colour appear in the same areas of the plot, showing five distinct clusters in the data." class="figure mx-auto d-block"><figcaption>
Scatter plot of principal component 2 against principal component 1 in
the scRNAseq data, coloured by clusters produced by k-means clustering.
</figcaption></figure><div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mean</span><span class="op">(</span><span class="va">sil5</span><span class="op">[</span>, <span class="st">"sil_width"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.5849979</code></pre>
</div>
<p>The average silhouette width is lower when k=5.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">sil5</span>, border <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-unnamed-chunk-2-1.png" alt="Plot with horizontal axis silhoutte width. The plot shows the silhouette width for each point in the data set according to cluster. Cluster 4 contains almost half of the points in the data set and largely consists of points with a large silhouette list, leading to a bar that extends to the right side of the graph. The other clusters contain many fewer points and have similar silhouette widths. The bars for cluster 5 are much smaller, with a small number extending to the left of the origin, indicating negative silhouette widths." class="figure mx-auto d-block"><figcaption>
Silhouette plot for each point according to cluster.
</figcaption></figure><p>This seems to be because some observations in clusters 3 and 5 seem
to be more similar to other clusters than the one they have been
assigned to. This may indicate that k is too high.</p>
</div>
</div>
</div>
</div>
<div id="gap-statistic" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="gap-statistic" class="callout-inner">
<h3 class="callout-title">Gap statistic</h3>
<div class="callout-content">
<p>Another measure of how good our clustering is is the “gap statistic”.
This compares the observed squared distance between observations in a
cluster and the centre of the cluster to an “expected” squared
distances. The expected distances are calculated by randomly
distributing cells within the range of the original data. Larger values
represent lower squared distances within clusters, and thus better
clustering. We can see how this is calculated in the following
example.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"cluster"</span><span class="op">)</span></span>
<span><span class="va">gaps</span> <span class="op">&lt;-</span> <span class="fu">clusGap</span><span class="op">(</span><span class="va">pcs</span>, <span class="va">kmeans</span>, K.max <span class="op">=</span> <span class="fl">20</span>, iter.max <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">best_k</span> <span class="op">&lt;-</span> <span class="fu">maxSE</span><span class="op">(</span><span class="va">gaps</span><span class="op">$</span><span class="va">Tab</span><span class="op">[</span>, <span class="st">"gap"</span><span class="op">]</span>, <span class="va">gaps</span><span class="op">$</span><span class="va">Tab</span><span class="op">[</span>, <span class="st">"SE.sim"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">best_k</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">gaps</span><span class="op">$</span><span class="va">Tab</span><span class="op">[</span>,<span class="st">"gap"</span><span class="op">]</span>, xlab <span class="op">=</span> <span class="st">"Number of clusters"</span>, ylab <span class="op">=</span> <span class="st">"Gap statistic"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v <span class="op">=</span> <span class="va">best_k</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="cluster-robustness-bootstrapping">Cluster robustness: bootstrapping<a class="anchor" aria-label="anchor" href="#cluster-robustness-bootstrapping"></a></h2>
<hr class="half-width"><p>When we cluster data, we want to be sure that the clusters we
identify are not a result of the exact properties of the input data.
That is, we want to ensure that the clusters identified do not change
substantially if the observed data change slightly. This makes it more
likely that the clusters can be reproduced.</p>
<p>To assess this, we can <em>bootstrap</em>. We first resample the data
with replacement to reproduce a ‘new’ data set. We can then calculate
new clusters for this data set and compare these to those on the
original data set, thus helping us to see how the clusters may change
for small changes in the data. This is maybe easier to see with an
example. First, we define some data:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span></code></pre>
</div>
<p>Then, we can take a sample from this data without replacement:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="va">data</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4 1 3 5 2</code></pre>
</div>
<p>This sample is a subset of the original data, and points are only
present once. This is the case every time even if we do it many
times:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Each column is a sample</span></span>
<span><span class="fu">replicate</span><span class="op">(</span><span class="fl">10</span>, <span class="fu">sample</span><span class="op">(</span><span class="va">data</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
[1,]    5    2    5    2    3    1    3    5    5     3
[2,]    4    5    4    5    4    4    1    3    1     2
[3,]    2    1    1    3    2    5    2    2    3     4
[4,]    1    4    2    1    5    3    5    1    2     5
[5,]    3    3    3    4    1    2    4    4    4     1</code></pre>
</div>
<p>However, if we sample <em>with replacement</em>, then sometimes
individual data points are present more than once.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">replicate</span><span class="op">(</span><span class="fl">10</span>, <span class="fu">sample</span><span class="op">(</span><span class="va">data</span>, <span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
[1,]    3    1    2    2    1    3    3    2    4     2
[2,]    1    3    2    4    2    5    2    1    2     5
[3,]    5    5    4    4    2    2    1    1    1     3
[4,]    1    1    4    2    1    4    4    5    5     4
[5,]    3    1    2    1    4    2    5    3    3     2</code></pre>
</div>
</section><section><h2 class="section-heading" id="bootstrapping">Bootstrapping<a class="anchor" aria-label="anchor" href="#bootstrapping"></a></h2>
<hr class="half-width"><p>Bootstrapping is a powerful and common statistical technique.</p>
<p>We would like to know about the sampling distribution of a statistic,
but we don’t have any knowledge of its behaviour under the null
hypothesis.</p>
<p>For example, we might want to understand the uncertainty around an
estimate of the mean of our data. To do this, we could resample the data
with replacement and calculate the mean of each average.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boots</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="fu">mean</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">data</span>, <span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">boots</span>,</span>
<span>    breaks <span class="op">=</span> <span class="st">"FD"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"1,000 bootstrap samples"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Mean of sample"</span></span>
<span> <span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-boots-1.png" alt="A histogram of the mean of each bootstrapped sample. The histogram appears roughly symmetric around 2.8 on the x axis." class="figure mx-auto d-block"><figcaption>
Histogram of mean of bootstapped samples.
</figcaption></figure><p>In this case, the example is simple, but it’s possible to devise more
complex statistical tests using this kind of approach.</p>
<p>The bootstrap, along with permutation testing, can be a very flexible
and general solution to many statistical problems.</p>
<p>In applying the bootstrap to clustering, we want to see two
things:</p>
<ol style="list-style-type: decimal"><li>Will observations within a cluster consistently cluster together in
different bootstrap replicates?</li>
<li>Will observations frequently swap between clusters?</li>
</ol><p>In the plot below, the diagonal of the plot shows how often the
clusters are reproduced in boostrap replicates. High scores on the
diagonal mean that the clusters are consistently reproduced in each
boostrap replicate. Similarly, the off-diagonal elements represent how
often observations swap between clusters in bootstrap replicates. High
scores indicate that observations rarely swap between clusters.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"pheatmap"</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"bluster"</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"viridis"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: viridisLite</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">kmeans</span><span class="op">(</span><span class="va">x</span>, centers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ratios</span> <span class="op">&lt;-</span> <span class="fu">bootstrapStability</span><span class="op">(</span><span class="va">pcs</span>, FUN <span class="op">=</span> <span class="va">km_fun</span>, clusters <span class="op">=</span> <span class="va">cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">ratios</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    breaks <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, length.out <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-bs-heatmap-1.png" alt="Grid of 16 squares labelled 1-4 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares." class="figure mx-auto d-block"><figcaption>
Grid of empirical cluster swapping behaviour estimated by the bootstrap
samples.
</figcaption></figure><p>Yellow boxes indicate values slightly greater than 1, which may be
observed. These are “good” (despite missing in the colour bar).</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3</h3>
<div class="callout-content">
<p>Repeat the bootstrapping process with k=5. Do the results appear
better or worse? Can you identify where the differences occur on the
<code>plotReducedDim()</code>?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km_fun5</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">kmeans</span><span class="op">(</span><span class="va">x</span>, centers <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">ratios5</span> <span class="op">&lt;-</span> <span class="fu">bootstrapStability</span><span class="op">(</span><span class="va">pcs</span>, FUN <span class="op">=</span> <span class="va">km_fun5</span>, clusters <span class="op">=</span> <span class="va">cluster5</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">ratios5</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>    breaks <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, length.out <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/06-k-means-rendered-bs-ex-1.png" alt="Grid of 25 squares labelled 1-5 on each of the x and y axes. The diagonal and off-diagonal squares of the grid are coloured in green, indicating the highest scoring value of 1 according to the legend, with the exception of the square corresponding to (4, 5), which is slightly darker green indicating a lower value. The lower triangular squares are coloured in grey, indicating NA values since these would be the same as the upper triangular squares." class="figure mx-auto d-block"><figcaption>
Grid of empirical cluster swapping behaviour estimated by the bootstrap
samples.
</figcaption></figure><p>When k=5, we can see that the values on the diagonal of the matrix
are smaller, indicating that the clusters aren’t exactly reproducible in
the bootstrap samples.</p>
<p>Similarly, the off-diagonal elements are considerably lower for some
elements. This indicates that observations are “swapping” between these
clusters in bootstrap replicates.</p>
</div>
</div>
</div>
</div>
<div id="consensus-clustering" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="consensus-clustering" class="callout-inner">
<h3 class="callout-title">Consensus clustering</h3>
<div class="callout-content">
<p>One useful and generic method of clustering is <em>consensus
clustering</em>. This method can use k-means or other clustering
methods.</p>
<p>The idea behind this is to bootstrap the data repeatedly, and cluster
it each time, perhaps using different numbers of clusters. If a pair of
data points always end up in the same cluster, it’s likely that they
really belong to the same underlying cluster.</p>
<p>This is really computationally demanding but has been shown to
perform very well in some situations. It also allows you to visualise
how cluster membership changes over different values of k.</p>
</div>
</div>
</div>
<div id="speed" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="speed" class="callout-inner">
<h3 class="callout-title">Speed</h3>
<div class="callout-content">
<p>It’s worth noting that a lot of the methods we’ve discussed here are
very computationally demanding. When clustering data, we may have to
compare points to each other many times. This becomes more and more
difficult when we have many observations and many features. This is
especially problematic when we want to do things like bootstrapping that
requires us to cluster the data over and over.</p>
<p>As a result, there are a lot of approximate methods for finding
clusters in the data. For example, the <a href="https://www.bioconductor.org/packages/3.13/bioc/html/mbkmeans.html" class="external-link">mbkmeans</a>
package includes an algorithm for clustering extremely large data. The
idea behind this algorithm is that if the clusters we find are robust,
we don’t need to look at all of the data every time. This is very
helpful because it reduces the amount of data that needs to be held in
memory at once, but also because it minimises the computational
cost.</p>
<p>Similarly, approximate nearest neighbour methods like <a href="https://pypi.org/project/annoy/" class="external-link">Annoy</a> can be used to identify
what the <span class="math inline">\(k\)</span> closest points are in
the data, and this can be used in some clustering methods (for example,
graph-based clustering).</p>
<p>Generally, these methods sacrifice a bit of accuracy for a big gain
in speed.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a></h2>
<hr class="half-width"><ul><li>
<a href="https://doi.org/10.1007/978-3-642-29807-3_1" class="external-link">Wu, J. (2012)
Cluster analysis and K-means clustering: An Introduction. In: Advances
in K-means Clustering. Springer Berlin, Heidelberg.</a>.</li>
<li>
<a href="https://web.stanford.edu/class/bios221/book/Chap-Clustering.html" class="external-link">Modern
statistics for modern biology, Susan Holmes and Wolfgang Huber (Chapter
5)</a>.</li>
<li>
<a href="https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1" class="external-link">Understanding
K-means clustering in machine learning, Towards Data Science</a>.</li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>K-means is an intuitive algorithm for clustering data.</li>
<li>K-means has various advantages but can be computationally
intensive.</li>
<li>Apparent clusters in high-dimensional data should always be treated
with some scepticism.</li>
<li>Silhouette width and bootstrapping can be used to assess how well
our clustering algorithm has worked.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/05-factor-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/07-hierarchical.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/05-factor-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Factor analysis
        </a>
        <a class="chapter-link float-end" href="../instructor/07-hierarchical.html" rel="next">
          Next: Hierarchical...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/edit/main/episodes/06-k-means.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/high-dimensional-stats-r/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:alan.ocallaghan@outlook.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/06-k-means.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "K-means",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/06-k-means.html",
  "identifier": "https://carpentries-incubator.github.io/high-dimensional-stats-r/instructor/06-k-means.html",
  "dateCreated": "2025-10-02",
  "dateModified": "2025-09-02",
  "datePublished": "2025-10-02"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

